// bluetooth/extensions/face++/index.js

exports.blocks = require('./blocks.json');
exports.toolbox = require('./toolbox.json');
const {message} = antd;

function check(test, fn, value){
	if(!test){
		message.warn('请先设置账号信息');
		return value;
	}
	return fn();
}

exports.context = {
	faceplusplus_set_info({key, secret}){
		return request(key, secret, {
			image_url: 'http://bj-mc-prod-asset.oss-cn-beijing.aliyuncs.com/wiki-pic/Gesture1.jpg'
		}).then(
			() => {
				this.key = key;
				this.secret = secret;
				message.success('设置账号成功');
			},
			error => {
				this.key = null;
				this.secret = null;
				message.error(error);
			}
		);
	},
	faceplusplus_check({gesture}){
		const {key, secret} = this;
		return check(key, async function(){
			try{
				let image_file = await getImage();
				let {hands} = await request(key, secret, {image_file});
				if(hands.length <= 0)return 0;
				const ctx = getCtx();
				ctx.strokeStyle = 'red';
				for(let hand of hands){
					const {top,left,width,height} = hand.hand_rectangle;
					ctx.strokeRect(left, top, width, height);
				}
				setTimeout(() => {
					for(let hand of hands){
						const {top,left,width,height} = hand.hand_rectangle;
						ctx.clearRect(left-1, top-1, width+2, height+2);
					}
				}, 500);
				return hands[0].gesture[gesture];
			}catch(error){
				message.error(error);
				return 0;
			}
		}, 0);
	}
}

const getCtx = () => document.querySelector('div.video > canvas').getContext('2d');

const getImage = () => new Promise(resolve => {
	const video = document.querySelector('div.video > video');
	const canvas = video.nextElementSibling;
	const {clientWidth, clientHeight} = video;
	canvas.width = clientWidth;
	canvas.height = clientHeight;
	const ctx = canvas.getContext('2d');
	ctx.drawImage(video, 0, 0, clientWidth, clientHeight);
	canvas.toBlob(v => {
		ctx.clearRect(0, 0, clientWidth, clientHeight);
		resolve(v);
	}, 'image/jpeg', 0.9);
});

function request(api_key, api_secret, args){
	const body = new FormData();
	body.append('api_key', api_key);
	body.append('api_secret', api_secret);
	for(let key in args){
		body.append(key, args[key]);
	}
	return fetch('https://api-cn.faceplusplus.com/humanbodypp/v1/gesture', {method:'POST',body})
	.then(v => v.json())
	.then(v => v.error_message ? Promise.reject(v.error_message) : v);
}
