// renderers/RenderQueue.js
"use strict";

const ObjectGroup = require("../utils/ObjectGroup");

class RenderQueue
{
	constructor(){
		this.drawUnitGroup = new ObjectGroup([
			drawUnit => drawUnit.material.shader,
			drawUnit => drawUnit.material.texture,
			drawUnit => drawUnit.material.constructor.name,
			drawUnit => drawUnit.vao
		],[
			function(drawUnit, gl, camera){
				gl.profileMgr.profile3d.shaderSwitchCount++;
				this.program = gl.programMgr.fetch(drawUnit.material.shader);
				gl.useProgram(this.program);
				drawUnit.material.onActiveUniform(gl, camera, this.program);
			},
			function(drawUnit, gl, camera){
				gl.profileMgr.profile3d.textureSwitchCount++;
				gl.bindTexture(gl.TEXTURE_2D, drawUnit.material.texture);
			},
			function(drawUnit, gl, camera){
				gl.profileMgr.profile3d.stateSwitchCount++;
				this.material = drawUnit.material;
				this.material.onActiveState(gl);
			},
			function(drawUnit, gl, camera){
				gl.profileMgr.profile3d.vaoSwitchCount++;
				this.vao = drawUnit.vao;
				this.vao.active(gl);
				this.material.onActiveVAO(gl, drawUnit);
			},
			function(drawUnitList, gl, camera){
				this.material.draw(gl, this, drawUnitList);
			}
		]);
	}

	draw(gl, camera, drawUnitList){
		if(drawUnitList.length <= 0){
			return;
		}
		let {drawUnitGroup} = this;
		drawUnitGroup.group(drawUnitList);
		drawUnitGroup.update(gl, camera);
	}
}

module.exports = RenderQueue;
