// webgpu/view3d/drawScene3D.js
'use strict';

const Vulkan = require('../Vulkan');

const drawScene3D = function(){
	const renderPassDesc = {
		colorAttachments: [{
			attachment: null,
			loadValue: [0,0,0,1]
		}],
		depthStencilAttachment: {
			attachment: null,
			depthLoadValue: 1.0,
			depthStoreOp: 'store',
			depthReadOnly: true,//default false, optional
			stencilLoadValue: 0,
			stencilStoreOp: 'store',
			stencilReadOnly: true,//default false, optional
		}
	};
	return function(root){
		const {mainColorTexture, mainDepthStencil, camera3DGPUResourceMgr} = Vulkan;
		const camera = root.camera;

		camera._gpuAsset = camera3DGPUResourceMgr.getOut();

		if(camera.cullEnabled){
			Vulkan.directBufferMgr.prepareCull(camera);
		}

		renderPassDesc.colorAttachments[0].attachment = mainColorTexture;
		renderPassDesc.depthStencilAttachment.attachment = mainDepthStencil;
		Vulkan.renderPassEncoder = Vulkan.commandEncoder.beginRenderPass(renderPassDesc);
		
		camera.draw(root);

		Vulkan.renderPassEncoder.endPass();

		if(camera.cullEnabled){
			Vulkan.directBufferMgr.setDispatchValue();
		}
	}
}();

module.exports = drawScene3D;
