// net/tcp/Packet.js

class Packet
{
	static cut(buffer, offset=0){
		let count = buffer.length - offset;
		if(count < 2)return;
		let packetLen = buffer.readUInt16BE(offset);
		if(count < packetLen)return;
		return buffer.slice(offset, offset+packetLen);
	}

	static decode(buffer){
		let reqId = buffer.readUInt16BE(2);
		let usrId = buffer.readUInt16BE(4);
		let msgId = buffer.readUInt16BE(6);
		let payload = buffer.length > 8 ? buffer.slice(8) : null;
		let packet = new this(msgId, payload, usrId);
		packet.reqId = reqId;
		return packet;
	}

	constructor(msgId, payload, usrId=0){
		this.usrId = usrId;
		this.msgId = msgId;
		this.payload = payload;
		this.reqId = 0;
	}

	toBytes(){
		const {payload} = this;
		const packetLen = 8 + (payload ? payload.length : 0);
		const buffer = Buffer.allocUnsafe(packetLen);
		buffer.writeUInt16BE(packetLen, 0);
		buffer.writeUInt16BE(this.reqId, 2);
		buffer.writeUInt16BE(this.usrId, 4);
		buffer.writeUInt16BE(this.msgId, 6);
		payload?.copy(buffer, 8);
		return buffer;
	}
}

module.exports = Packet;
