// bluetooth/extensions/tlink/index.js

exports.blocks = require('./blocks.json');
exports.toolbox = require('./toolbox.json');
const {tokenFactory, wrapTokenFn} = require('utils/token');
const {message} = antd;

/*
Blockly.getMainWorkspace().registerButtonCallback('reg_tlink', ()=>{
			nw.Shell.openExternal('https://www.tlink.io/');
		});
*/

function check(test, fn, value){
	if(!test){
		message.warn('请先设置账号信息');
		return value;
	}
	return fn();
}

exports.buttons = {
	register_tlink(){
		nw.Shell.openExternal('https://www.tlink.io');
	}
}

exports.context = {
	tlink_set_info(info){
		const {userId, clientId} = info;
		const tokenFn = wrapTokenFn(tokenFactory(() => getToken(info)), userId, clientId);
		return tokenFn(token => {
			this.tokenFn = tokenFn;
			message.success('账号设置成功!')
		}).catch(error => {
			this.tokenFn = null;
			message.error('账号信息有误!');
		});
	},
	tlink_set_value({sensorId, deviceNo, value}){
		const {tokenFn} = this;
		return check(tokenFn, () => tokenFn(deviceWrite, deviceNo, sensorId, value).then(
			v => undefined,
			error => void message.error(error)
		));
	},
	tlink_get_value({sensorId}){
		const {tokenFn} = this;
		return check(tokenFn, () => tokenFn(deviceRead, sensorId).then(
			v => v.value,
			error => (message.error(error), 0)
		), 0);
	}
}

function getToken({username, password, clientId, secret}){
	const params = new URLSearchParams();
	params.append('grant_type', 'password');
	params.append('username', username);
	params.append('password', password);
	return fetch(`http://api.tlink.io/oauth/token?${params}`, {
		method:'POST',
		redirect:'error',
		credentials:'omit',
		headers:{
			'authorization': `Basic ` + Buffer.from(clientId+':'+secret).toString('base64'),
			'Content-Type':'text/plain',
			'cache-control':'no-cache',
		}
	})
	.then(response => response.json())
	.then(v => v.error ? Promise.reject(v.error_description) : [v.access_token, v.expires_in]);
}

function deviceDo(api, argsFn){
	return function(access_token, userId, clientId, ...args){
		const body = argsFn(...args);
		body.userId = userId;
		return fetch(`http://api.tlink.io/api/device/${api}`, {
			method:'POST',
			headers:{
				'tlinkAppId':clientId,
				'Authorization':`Bearer ` + access_token,
				'Content-Type':'application/json',
				'cache-control':'no-cache',
			},
			body:JSON.stringify(body)
		})
		.then(response => response.json())
		.then(v => {
			if(v.error)return Promise.reject(v.message);
			if(v.flag != '00')return Promise.reject(v.msg);
			return v;
		});
	}
}

const deviceWrite = deviceDo('sendDataPoint', (deviceNo, sensorId, value) => ({deviceNo,sensorDatas:[{sensorId,value}]}));
const deviceRead = deviceDo('getSingleSensorDatas', sensorId => ({sensorId}));
