// mvc/src/Application.js

import {Injector} from 'ioc';
import Module from './Module';
import Msg from './Msg';
import MsgName from './MsgName';

class Application
{
	_injector = new Injector();
	_moduleDict = Object.create(null);
	_msgDispatched = Object.create(null);
	_msgInterested = Object.create(null);
	_roleMap = new Map();

	constructor(){
		this._injector.mapValue(Application, this, null);
		this._injector.mapValue(Injector, this._injector, null);
	}

	regRole(role, injector){
		this._roleMap.set(role, injector);
	}

	activateRoles(){
		for(let [role, injector] of this._roleMap)
			injector.getInstance(role);
		this._roleMap.clear();
	}

	regModule(meta){
		let module = new Module(meta, this._injector, this);
		console.assert(!(this.hasStartup || (module.name in this._moduleDict)));
		this._moduleDict[module.name] = module;
		return this;
	}

	getModule(name){
		return this._moduleDict[name];
	}

	notifyAll(msgName, msgData, dispatcher){
		if(!dispatcher){
			for(let module of Object.values(this._moduleDict)){
				module.notify(new Msg(msgName, msgData, module));
			}
		}else if(!( msgName  in this._msgDispatched)){//local msgs
		}else if(dispatcher !== this._msgDispatched[msgName]){
			console.warn("%s can't notify global msg %s!", dispatcher.name, msgName.valueOf());
		}else{
			let moduleList = this._msgInterested[msgName];
			if(!moduleList)return;
			for(let module of moduleList){
				module.notify(new Msg(msgName, msgData, this));
			}
		}
	}

	startup(...args){
		if(this.hasStartup)return;
		Object.defineProperty(this, "hasStartup", {value:true});
		this.onStartup(...args);
		this.notifyAll(MsgName.ModuleStartup);
	}

	onStartup(){
		let moduleList = Object.values(this._moduleDict);
		for(let module of moduleList)
			module.regMsgs(this._msgDispatched, this._msgInterested);
		moduleList.forEach(initAllModels);
		this.activateRoles();
		moduleList.forEach(initAllServices);
		this.activateRoles();
		moduleList.forEach(initAllControllers);
	}
}

function initAllModels(module){
	for(let v of module.collectAllModels()){
		if(Array.isArray(v)){
			if(v.length == 2){
				module.regModel(v[1], v[0]);
				continue;
			}else if(v.length == 1){
				v = v[0];
			}else{
				throw new Error(v);
			}
		}
		if(v.prototype){
			module.regModel(new v());
		}else{
			module.regModel(v);
		}
	}
}

function initAllServices(module){
	for(let v of module.collectAllServices()){
		if(!Array.isArray(v)){
			module.regService(v, v);
		}else if(v.length == 3){
			module.regService(...v);
		}else if(v.length == 2){
			if(typeof v[1] == "boolean"){
				module.regService(v[0], ...v);
			}else{
				module.regService(...v);
			}
		}else if(v.length == 1){
			module.regService(v[0], v[0]);
		}else{
			throw new Error(v);
		}
	}
}

function initAllControllers(module){
	for(let v of module.collectAllControllers()){
		module.regController(v);
	}
}

export default Application;
