// webgpu/pipelines/createPipelineDesc3D.js
'use strict';

const Vulkan = require('../Vulkan');
const {attributes, blend} = require('./helper');

const defaultColorBlend = {
	srcFactor: 'src-alpha',
	dstFactor: 'one-minus-src-alpha',
	operation: 'add'
}

module.exports = function(bindGroupLayouts, vertexCode, fragmentCode, primitiveTopology='triangle-list', colorBlend=null, depthWriteEnabled=true){
	const {device} = Vulkan;
	/*
	return {
		layout: device.createPipelineLayout({
			bindGroupLayouts
		}),
		vertex:{
			module: device.createShaderModule({
				code: vertexCode,
			}),
			entryPoint: 'main',
			buffers: [{
				arrayStride: 9 << 2,
				stepMode: 'vertex',
				attributes: attributes('float32x3', 'float32x3', 'float32x2', 'uint32')
			}]
		},
		fragment:{
			module: device.createShaderModule({
				code: fragmentCode,
			}),
			entryPoint: 'main',
			targets:[{
				format: Vulkan.swapChainFormat,
				blend: blend(colorBlend || 'transparent', 'transparent')
			}]
		},
		depthStencil: {
			format: 'depth24plus-stencil8',
			depthWriteEnabled: depthWriteEnabled,
			depthCompare: 'less-equal',
		},
		primitive: {
			topology: primitiveTopology,
			stripIndexFormat: (primitiveTopology == 'line-strip'  || primitiveTopology ==  'triangle-strip') ? 'uint16' : undefined,
			frontFace: 'ccw',
			cullMode: 'none',
		}
	};
	//*/
	return {
		layout: device.createPipelineLayout({
			bindGroupLayouts
		}),
		vertexStage: {
			module: device.createShaderModule({
				code: vertexCode,
			}),
			entryPoint: 'main'
		},
		fragmentStage: {
			module: device.createShaderModule({
				code: fragmentCode,
			}),
			entryPoint: 'main'
		},
		primitiveTopology,
		rasterizationState: {
			frontFace: 'ccw',
			cullMode: 'none',
		},
		colorStates:[{
			format: Vulkan.swapChainFormat,
			colorBlend: colorBlend || defaultColorBlend,
			alphaBlend: {
				srcFactor: 'src-alpha',
				dstFactor: 'one-minus-src-alpha',
				operation: 'add'
			}
		}],
		depthStencilState: {
			format: 'depth24plus-stencil8',
			depthWriteEnabled: depthWriteEnabled,
			depthCompare: 'less-equal',
		},
		vertexState: {
			//indexFormat: 'uint16',
			vertexBuffers:[{
				arrayStride: 9 << 2,
				stepMode: 'vertex',
				attributes:[{
					shaderLocation: 0,
					offset: 0,
					format: 'float32x3'
				},{
					shaderLocation: 1,
					offset: 3 << 2,
					format: 'float32x3'
				},{
					shaderLocation: 2,
					offset: 6 << 2,
					format: 'float32x2'
				},{
					shaderLocation: 3,
					offset: 8 << 2,
					format: 'uint32'
				}]
			}]
		}
	};
}
