// webgpu/pipelines/pipelineFilterBlur.js
'use strict';

const Vulkan = require('../Vulkan');
const {blend} = require('./helper');

module.exports = function(){
	const {device} = Vulkan;
	/*
	return device.createRenderPipeline({
		layout: device.createPipelineLayout({
			bindGroupLayouts: [
				Vulkan.layout_filter2D,
				Vulkan.layout_filter_blur
			]
		}),
		vertex:{
			module: device.createShaderModule({
				code: Vulkan.shader_filter.vertexCode,
			}),
			entryPoint: 'main',
		},
		fragment:{
			module: device.createShaderModule({
				code: Vulkan.shader_filter.fragmentCode,
			}),
			entryPoint: 'main',
			targets:[{
				format: Vulkan.swapChainFormat,
				blend: blend('transparent')
			}]
		},
		primitive: {
			topology: 'triangle-strip',
			stripIndexFormat: 'uint16',
		}
	});
	//*/
	return device.createRenderPipeline({
		layout: Vulkan.createPipelineLayout(['filter2D', 'filter_blur']),
		vertexStage: {
			module: Vulkan.shader_vertex_filter,
			entryPoint: 'main'
		},
		fragmentStage: {
			module: Vulkan.shader_fragment_filter,
			entryPoint: 'main'
		},
		primitiveTopology: 'triangle-strip',
		vertexState: {
			indexFormat: 'uint16'
		},
		colorStates:[{
			format: Vulkan.swapChainFormat,
			colorBlend: {
				srcFactor: 'src-alpha',
				dstFactor: 'one-minus-src-alpha',
				operation: 'add'
			},
			alphaBlend: {
				srcFactor: 'src-alpha',
				dstFactor: 'one-minus-src-alpha',
				operation: 'add'
			}
		}]
	});
}
