// webgpu/pipelines/pipelineTerrain.js
'use strict';

const Vulkan = require('../Vulkan');

module.exports = function(){
	const {device} = Vulkan;
	return device.createRenderPipeline({
		layout: device.createPipelineLayout({
			bindGroupLayouts:[
				Vulkan.bindGroupLayoutDict.bindGroupLayout,
				Vulkan.bindGroupLayoutDict.bindGroupLayout5,
			]
		}),
		vertexStage: {
			module: device.createShaderModule({
				code: Vulkan.shader_terrain.vertexCode,
			}),
			entryPoint: 'main'
		},
		fragmentStage: {
			module: device.createShaderModule({
				code: Vulkan.shader_terrain.fragmentCode,
			}),
			entryPoint: 'main'
		},
		primitiveTopology: 'triangle-list',
		colorStates:[{
			format: Vulkan.swapChainFormat,
			colorBlend: {
				srcFactor: 'src-alpha',
				dstFactor: 'one-minus-src-alpha',
				operation: 'add'
			},
			alphaBlend: {
				srcFactor: 'src-alpha',
				dstFactor: 'one-minus-src-alpha',
				operation: 'add'
			}
		}],
		depthStencilState: {
			format: 'depth24plus-stencil8',
			depthWriteEnabled: true,
			depthCompare: 'less-equal',
		},
		vertexState: {
			//indexFormat: 'uint16',
			vertexBuffers:[{
				arrayStride: 12,
				stepMode: 'vertex',
				attributes:[{
					shaderLocation: 0,
					offset: 0,
					format: 'uchar2'
				},{
					shaderLocation: 1,
					offset: 4,
					format: 'uchar2norm'
				},{
					shaderLocation: 2,
					offset: 8,
					format: 'uchar4norm'
				}]
			}]
		}
	});
}
