// webgpu/pipelines/pipelineBound.js
'use strict';

const Vulkan = require('../Vulkan');
const {blend} = require('./helper');

module.exports = function(){
	const {device} = Vulkan;
	/*
	return device.createRenderPipeline({
		layout: device.createPipelineLayout({
			bindGroupLayouts:[
				Vulkan.layout_0,
				Vulkan.layout_2_dynamic,
			]
		}),
		vertex:{
			module: device.createShaderModule({
				code: Vulkan.shader_bound.vertexCode,
			}),
			entryPoint: 'main',
		},
		fragment:{
			module: device.createShaderModule({
				code: Vulkan.shader_bound.fragmentCode,
			}),
			entryPoint: 'main',
			targets:[{
				format: Vulkan.swapChainFormat,
				blend: blend('transparent')
			}]
		},
		depthStencil: {
			format: 'depth24plus-stencil8',
			depthWriteEnabled: true,
			depthCompare: 'less-equal',
		},
		primitive: {
			topology: 'line-list',
		}
	});
	//*/
	return device.createRenderPipeline({
		layout: device.createPipelineLayout({
			bindGroupLayouts:[
				Vulkan.layout_0,
				Vulkan.layout_2_dynamic,
		]}),
		vertexStage: {
			module: device.createShaderModule({
				code: Vulkan.shader_bound.vertexCode,
			}),
			entryPoint: 'main'
		},
		fragmentStage: {
			module: device.createShaderModule({
				code: Vulkan.shader_bound.fragmentCode,
			}),
			entryPoint: 'main'
		},
		primitiveTopology: 'line-list',
		colorStates:[{
			format: Vulkan.swapChainFormat,
			colorBlend: {
				srcFactor: 'src-alpha',
				dstFactor: 'one-minus-src-alpha',
				operation: 'add'
			},
			alphaBlend: {
				srcFactor: 'src-alpha',
				dstFactor: 'one-minus-src-alpha',
				operation: 'add'
			}
		}],
		depthStencilState: {
			format: 'depth24plus-stencil8',
			depthWriteEnabled: true,
			depthCompare: 'less-equal',
		}
	});
}
