// webgpu/pipelines/pipeline3DBillboard.js
'use strict';

const Vulkan = require('../Vulkan');
const {blend} = require('./helper');

module.exports = function(){
	const {device} = Vulkan;
	/*
	return device.createRenderPipeline({
		layout: device.createPipelineLayout({
			bindGroupLayouts: [
				Vulkan.layout_0,
				Vulkan.layout_2_dynamic,
				Vulkan.layout_3,
			]
		}),
		vertex:{
			module: device.createShaderModule({
				code: Vulkan.shader_billboard.vertexCode,
			}),
			entryPoint: 'main',
		},
		fragment:{
			module: device.createShaderModule({
				code: Vulkan.shader_billboard.fragmentCode,
			}),
			entryPoint: 'main',
			targets:[{
				format: Vulkan.swapChainFormat,
				blend: blend(['one', 'one-minus-src-color'], 'transparent')
			}]
		},
		depthStencil: {
			format: 'depth24plus-stencil8',
			depthWriteEnabled: false,
			depthCompare: 'less-equal',
		},
		primitive: {
			topology: 'triangle-strip',
			stripIndexFormat: 'uint16',
		}
	});
	//*/
	return device.createRenderPipeline({
		layout: device.createPipelineLayout({
			bindGroupLayouts: [
				Vulkan.layout_0,
				Vulkan.layout_2_dynamic,
				Vulkan.layout_3,
			]
		}),
		vertexStage: {
			module: device.createShaderModule({
				code: Vulkan.shader_billboard.vertexCode,
			}),
			entryPoint: 'main'
		},
		fragmentStage: {
			module: device.createShaderModule({
				code: Vulkan.shader_billboard.fragmentCode,
			}),
			entryPoint: 'main'
		},
		primitiveTopology: 'triangle-strip',
		rasterizationState: {
			frontFace: 'ccw',
			cullMode: 'none',
		},
		colorStates:[{
			format: Vulkan.swapChainFormat,
			colorBlend: {
				srcFactor: 'one',
				dstFactor: 'one-minus-src-color',
				operation: 'add'
			},
			alphaBlend: {
				srcFactor: 'src-alpha',
				dstFactor: 'one-minus-src-alpha',
				operation: 'add'
			}
		}],
		depthStencilState: {
			format: 'depth24plus-stencil8',
			depthWriteEnabled: false,
			depthCompare: 'less-equal',
		},
		vertexState: {
			indexFormat: 'uint16',
		}
	});
}
