// bluetooth/extensions/elf-mini/index.js

exports.blocks = require('./blocks.json');
exports.toolbox = require('./toolbox.json');

exports.context = {
	elf_mini_buzzer_pin(){
		return 7;
	}
}

exports.arduino = {
	elf_mini_port:function(){
		const pinMap = {19:'PORT_A', 18:'PORT_B', 16:'PORT_C', 15:'PORT_D'};
		return function(block, {value}){
			this.addInclude('#include<WeELFMini.h>');
			return [pinMap[value], this.ORDER_ATOMIC];
		}
	}(),
	elf_mini_buzzer_pin(){
		this.addInclude('#include<WeELFMini.h>');
		return ['OnBoard_Buzzer', this.ORDER_ATOMIC];
	},
	elf_mini_back_led:function(){
		const pinMap = {4:'MINI_LEFT_YELLOW', 3:'MINI_LEFT_RED', 14:'MINI_RIGHT_RED', 13:'MINI_RIGHT_YELLOW'};
		return function(block, {pin, value}){
			this.addInclude('#include<WeELFMini.h>');
			return `digitalWrite(${pinMap[pin]}, ${value});\n`;
		}
	}(),
	elf_mini_light_pin(){
		this.addInclude('#include<WeELFMini.h>');
		return ['OnBoard_Light', this.ORDER_ATOMIC];
	},
	elf_mini_sound_pin(){
		this.addInclude('#include<WeELFMini.h>');
		return ['OnBoard_Sound', this.ORDER_ATOMIC];
	},
	elf_mini_ir_pin(){
		this.addInclude('#include<WeELFMini.h>');
		return ['OnBoard_IR', this.ORDER_ATOMIC];
	}
}
