// net/load.js
'use strict';

const {URL} = require('url');
const {sendCookie, recvCookie} = require('./cookie');
const readDataAsync = require('./readDataAsync');
const compose = require('utils/function/compose');

const castType = Reflect.get.bind(null, Object.create(null, {
	text: {value: v => v.toString()},
	json: {value: v => JSON.parse(v.toString())},
}));

const castBody = function(){
	const BOUNDARY = '----WebKitFormBoundaryJogJuHTlZmoELETx';
	const newFormBody = function(FORM_BEGIN, FORM_END, FORM_RETURN){
		return function*(form){
			for(let k in form){
				let v = form[k];
				yield FORM_BEGIN;
				if(Buffer.isBuffer(v)){
					yield Buffer.from(`Content-Disposition: form-data; name="${k}"; filename="blob"\nContent-Type: image/jpeg\n\n`);
					yield v;
					yield FORM_RETURN;
				}else{
					yield Buffer.from(`Content-Disposition: form-data; name="${k}"\n\n${v}\n`);
				}
			}
			yield FORM_END;
		}
	}(Buffer.from(`--${BOUNDARY}\n`), Buffer.from(`--${BOUNDARY}--`), Buffer.from('\n'));
	return function(headers, body){
		const contentType = headers['Content-Type'];
		if(contentType == 'application/json')return JSON.stringify(body);
		if(contentType == 'application/x-www-form-urlencoded')return new URLSearchParams(body).toString();
		if(contentType == 'multipart/form-data'){
			headers['Content-Type'] = `multipart/form-data; boundary=${BOUNDARY}`;
			return Buffer.concat(Array.from(newFormBody(body)));
		}
		return body;
	}
}()

function load(url, {method, body, headers, query}={}){
	return new Promise(function(resolve, reject){
		const info = new URL(url);
		for(let k in query)info.searchParams.append(k, query[k]);
		const http = require(info.protocol.slice(0, -1));
		const request = http.request({method, headers,
			path: info.pathname + info.search,
			host: info.hostname,
			port: info.port
		}, response => {
			if(response.statusCode == 200){
				recvCookie(response, info);
				readDataAsync(response, resolve);
			}else{
				reject(response);
			}
		});
		sendCookie(request, info);
		request.on('error', reject);
		request.end(body);
	});
}

const castRequestTypePlugin = next => (url, args) => {
	if(args){
		const {headers, body} = args;
		if(headers && body && !Buffer.isBuffer(body) && typeof body != 'string'){
			args.body = castBody(headers, body);
		}
	}
	return next(url, args);
}

const castResponseTypePlugin = next => (url, args) => {
	const responseType = (args && args.responseType) ? args.responseType : 'text';
	return next(url, args).then(castType(responseType));
}

const redirectPlugin = next => (url, args) => {
	const promise = next(url, args);
	if(args && args.redirect === false)return promise;
	return promise.then(null, response => {
		switch(response.statusCode){
			case 301:case 302:
				return next(response.headers.location, args);
		}
		return Promise.reject(response);
	});
}

module.exports = compose(
	castRequestTypePlugin,
	castResponseTypePlugin,
	redirectPlugin,
)(load);
