// web_services/DLNA.js
'use strict';

const dgram = require('dgram');
const http = require('http');
const {URL} = require('url');
const fs = require('fs');
const {basename} = require('path');
const EventEmitter = require('events');
const networkInterface = require('../net/networkInterface');
const xmlEscape = require('../utils/html/escape');
const load = require('../net/load');

const multicast_ip = '239.255.255.250';
const multicast_port = 1900;

const separator = '\r\n';
const concatHead = list => list.map(v => v + separator).join('') + separator;
const splitHead = text => text.toString().trim().split(separator);

const discoverPacket = concatHead([
	`M-SEARCH * HTTP/1.1`,
	`MX: 5`,
	`ST: upnp:rootdevice`,
	`MAN: "ssdp:discover"`,
	`Host: ${multicast_ip}:${multicast_port}`,
]);

const Packet = function(){
	function createPacket(name, args){
		return `<s:Envelope s:encodingStyle="http://schemas.xmlsoap.org/soap/encoding/" xmlns:s="http://schemas.xmlsoap.org/soap/envelope/"><s:Body><u:${name} xmlns:u="urn:schemas-upnp-org:service:AVTransport:1">${Object.entries(args).map(([k, v]) => `<${k}>${xmlEscape(v.toString())}</${k}>`).join('')}</u:${name}></s:Body></s:Envelope>`;
	}
	return {
		play(CurrentURI){
			return createPacket('SetAVTransportURI', {InstanceID: 0, CurrentURIMetaData: '', CurrentURI});
		},
		seek(Target, Unit='ABS_TIME'){
			return createPacket('Seek', {InstanceID: 0, Unit, Target});
		},
		pause: createPacket('Pause', {InstanceID: 0}),
		resume: createPacket('Play', {InstanceID: 0, Speed: 1}),
		stop: createPacket('Stop', {InstanceID: 0}),
	};
}();

function findScreen(){
	return new Promise(resolve => {
		//let findFlag = false;
		dgram.createSocket('udp4').on('error', function(error){
			console.log(`server error:\n${error.stack}`);
			this.close();
		}).on('message', async function(message){
			//if(findFlag)return;
			const head = splitHead(message);
			head.shift();
			const dict = new Map(head.map(v => v.split(': ')));
			if(!dict.get('SERVER').includes(' DLNADOC/'))return;
			//findFlag = true;
			this.close();
			resolve(await getControlURL(dict.get('Location')));
		}).on('listening', function(){
			this.addMembership(multicast_ip);
			this.setMulticastTTL(255);
			this.send(discoverPacket, multicast_port, multicast_ip);
		}).bind();
	});
}

async function getControlURL(location){
	const {origin} = new URL(location);
	const info = await load(location);
	const controlService = /<serviceType>urn:schemas-upnp-org:service:AVTransport:1<\/serviceType>(.+?)<\/service>/s.exec(info)[1];
	const controlURL = /<controlURL>(.+?)<\/controlURL>/.exec(controlService)[1];
	return origin + controlURL;
}

async function sendPacket(controlURL, packet){
	const result = await load(controlURL, {
		method: 'POST',
		headers: {
			'SOAPAction': `urn:schemas-upnp-org:service:AVTransport:1#SetAVTransportURI"`,
			'Content-Type': `text/xml; charset="utf-8"`,
		},
		body: packet
	});
	console.log(result);
}

function generateLink(fileName, onRequest){
	return new Promise(resolve => {
		const host = networkInterface();
		//const date = new Date().toGMTString();
		//const ETag = JSON.stringify(Buffer.from(url).toString('hex'));
		http.createServer(function(request, response){
			if(decodeURI(request.url).slice(1) != fileName)return;
			onRequest(request, response);
		}).listen(0, '0.0.0.0', function(){
			const {port} = this.address();
			resolve(`http://${host}:${port}/${encodeURI(fileName)}`);
		});
	});
}

function castPath(fileName, totalLength, readFn){
	return generateLink(fileName, function(request, response){
		const {range} = request.headers;
		console.log(request.url, range);
		response.setHeader('Content-Type', 'video/mp4');
		if(range && range != 'bytes=0-'){
			let [start, end] = range.split('=')[1].split('-').map(Number);
			if(!end){
				end = totalLength - 1;
			}
			response.setHeader('Content-Length', end - start + 1);
			response.setHeader('Content-Range', `bytes ${start}-${end}/${totalLength}`);
			response.writeHead(206, 'Partial Content');
			readFn(response, start, end);
		}else{
			response.setHeader('Content-Length', totalLength);
			readFn(response, 0, totalLength - 1);
		}
	});
}

function castFilePath(url){
	return castPath(basename(url), fs.lstatSync(url).size, (response, start, end) => {
		fs.createReadStream(url, {start, end}).pipe(response);
	});
}

function castHttpPath(url){
	const {protocol, hostname, pathname} = new URL(url);
	const http = require(protocol.slice(0, -1));
	let videoBuffer;
	let videoLength = 0;
	const emitter = new EventEmitter();
	return new Promise(resolve => http.get(encodeURI(url), {headers:{Host: hostname}}, response => {
		console.log(response.headers);
		const totalLength = Number(response.headers['content-length']);
		videoBuffer = Buffer.alloc(totalLength);
		response.on('data', data => {
			videoLength += data.copy(videoBuffer, videoLength);
			emitter.emit('data', data);
		});
		resolve(totalLength);
	})).then(totalLength => castPath(decodeURI(pathname.slice(1)), totalLength, (response, start, end) => {
		if(end < videoLength){
			response.end(videoBuffer.slice(start, end + 1));
		}else{
			if(start < videoLength){
				response.write(videoBuffer.slice(start, videoLength));
			}
			function onData(data){
				response.write(videoBuffer.slice(Math.max(start, videoLength - data.length), Math.min(end + 1, videoLength)));
				if(end < videoLength){
					response.end();
					emitter.off('data', onData);
				}
			}
			emitter.on('data', onData);
			response.on('close', () => emitter.off('data', onData));
		}
	}));
}

exports.findScreen = findScreen;

exports.castFilePath = castFilePath;
exports.castHttpPath = castHttpPath;

exports.sendPacket = sendPacket;
exports.Packet = Packet;
