// fileformats/torrent.js
"use strict"

class Reader
{
	constructor(source, offset=0){
		this.source = source;
		this.offset = offset;
	}

	peek(){
		return String.fromCharCode(this.source[this.offset]);
	}

	read(){
		switch(this.peek()){
			case 'd': return this.readDict();
			case 'l': return this.readList();
			case 'i': return this.readNumber();
		}
		return this.readString();
	}

	readDict(){
		let result = {};
		++this.offset;//d
		while(this.peek() !== 'e')
			result[this.read()] = this.read();
		++this.offset;//e
		return result;
	}

	readList(){
		let result = [];
		++this.offset;//l
		while(this.peek() !== 'e')
			result.push(this.read());
		++this.offset;//e
		return result;
	}

	readNumber(){
		let {source} = this;
		let index = source.indexOf('e', this.offset);
		let result = source.toString('utf8', this.offset + 1, index);
		this.offset = index + 1;
		return Number(result);
	}

	readString(){
		let {source} = this;
		let index = source.indexOf(':', this.offset);
		let count = parseInt(source.toString('utf8', this.offset, index));
		this.offset = index + 1 + count;
		return source.toString('utf8', index + 1, this.offset);
	}
}

module.exports = Reader;
