// opengl/3d/BoneObject.js
'use strict';

const {Matrix3D} = require('../math');

class BoneObject
{
	constructor(parent, id){
		this.parent = parent;
		this.id = id;
		this.transform = new Matrix3D();
		this.worldTransform = new Matrix3D();
		this.nextSibling = null;
		this.firstChild = null;
	}

	get lastSibling(){
		let test = this;
		while(test.nextSibling)
			test = test.nextSibling;
		return test;
	}

	addChild(value){
		if(this.firstChild){
			this.firstChild.addSibling(value);
		}else{
			this.firstChild = value;
		}
	}
		
	addSibling(value){
		this.lastSibling.nextSibling = value;
	}

	forEachSibling(self, handler){
		let test = this;
		while(test){
			handler.call(self, test);
			test = test.nextSibling;
		}
	}

	forEachBone(self, handlerBefore, handlerAfter){
		const {nextSibling, firstChild} = this;
		if(handlerBefore){
			handlerBefore.call(self, this);
		}
		if(nextSibling){
			nextSibling.forEachBone(self, handlerBefore, handlerAfter);
		}
		if(firstChild){
			firstChild.forEachBone(self, handlerBefore, handlerAfter);
		}
		/*
		if(handlerAfter){
			handlerAfter.call(self, this);
		}
		//*/
	}
}


module.exports = BoneObject;
