// utils/function/operator.js
/*
优先级	运算符类型	结合性	运算符
19
分组	n/a（不相关）	( … )
18
成员访问	从左到右	… . …
需计算的成员访问	从左到右	… [ … ]
new（带参数列表）	n/a	new … ( … )
函数调用	从左到右	… ( … )
可选链（Optional chaining）	从左到右	?.
17
new（无参数列表）	从右到左	new …
16
后置递增
后置递减
15
逻辑非 (!)	从右到左	! …
按位非 (~)	~ …
一元加法 (+)	+ …
一元减法 (-)	- …
前置递增	++ …
前置递减	-- …
typeof	typeof …
void	void …
delete	delete …
await	await …
14
幂 (**)	从右到左	… ** …
13
乘法 (*)	从左到右	… * …
除法 (/)	… / …
取余 (%)	… % …
12
加法 (+)	从左到右	… + …
减法 (-)	… - …
11
按位左移 (<<)	从左到右	… << …
按位右移 (>>)	… >> …
无符号右移 (>>>)	… >>> …
10
小于 (<)	从左到右	… < …
小于等于 (<=)	… <= …
大于 (>)	… > …
大于等于 (>=)	… >= …
in	… in …
instanceof	… instanceof …
9
相等 (==)	从左到右	… == …
不相等 (!=)	… != …
一致/严格相等 (===)	… === …
不一致/严格不相等 (!==)	… !== …
8
按位与 (&)	从左到右	… & …
7
按位异或 (^)	从左到右	… ^ …
6
按位或 (|)	从左到右	… | …
5
逻辑与 (&&)	从左到右	… && …
4
逻辑或 (||)	从左到右	… || …
空值合并 (??)	从左到右	… ?? …
3
条件（三元）运算符	从右到左	… ? … : …
2
赋值	从右到左
… = …
… += …
… -= …
… **= …
… *= …
… /= …
… %= …
… <<= …
… >>= …
… >>>= …
… &= …
… ^= …
… |= …
… &&= …
… ||= …
… ??= …
1
逗号 / 序列	从左到右	… , …
*/
exports.Add = (a, b) => a + b;
exports.Subtract = (a, b) => a - b;
exports.Multiply = (a, b) => a * b;
exports.Divide = (a, b) => a / b;
exports.Modulus = (a, b) => a % b;
exports.Not = (value) => !value;
exports.And = (a, b) => a && b;
exports.Or = (a, b) => a || b;
exports.Equal = (a, b) => a == b;
exports.NotEqual = (a, b) => a != b;
exports.GreaterEqual = (a, b) => a >= b;
exports.GreaterThan = (a, b) => a > b;
exports.LessEqual = (a, b) => a <= b;
exports.LessThan = (a, b) => a < b;

exports.Double = v => v * 2;
exports.Square = v => v * v;
exports.IsOdd = v => v % 2 == 1;
