// webgpu/core/Events.js

const Signal = require('mvc/node_modules/signal');

class MouseEvents
{
	constructor(){
		window.addEventListener('pointerdown', evt => this.onMouseDown(evt));
		window.addEventListener('pointerup', evt => this.onMouseUp(evt));
		window.addEventListener('pointermove', evt => this.onMouseMove(evt));

		this.mouseDownSignal = new Signal();
		this.mouseUpSignal = new Signal();
		this.mouseMoveSignal = new Signal();

		this.mouseX = 0;
		this.mouseY = 0;
	}

	onMouseDown(evt){
		this.mouseX = evt.x;
		this.mouseY = evt.y;
		this.mouseDownSignal.notify(evt);
	}

	onMouseUp(evt){
		this.mouseX = evt.x;
		this.mouseY = evt.y;
		this.mouseUpSignal.notify(evt);
	}

	onMouseMove(evt){
		this.mouseX = evt.x;
		this.mouseY = evt.y;
		this.mouseMoveSignal.notify(evt);
	}
}

class KeyEvents
{
	constructor(){
		window.addEventListener('keydown', evt => this.onKeyDown(evt));
		window.addEventListener('keyup', evt => this.onKeyUp(evt));
	}

	onKeyDown(evt){
		console.log(evt)
	}

	onKeyUp(evt){
		console.log(evt)
	}
}

exports.mouseEvents = new MouseEvents();
exports.keyboardEvents = new KeyEvents();
