// webgpu/core/GPUContext.js

class GPUContext
{
	constructor(renderPassEncoder){
		this.renderPassEncoder = renderPassEncoder;
		this.pipeline = null;
	}

	setPipeline(pipeline){
		if(this.pipeline == pipeline)return;
		this.pipeline = pipeline;
		pipeline.setPipeline(this.renderPassEncoder);
	}

	setViewport(x, y, width, height, minDepth=0, maxDepth=1){
		this.renderPassEncoder.setViewport(x, y, width, height, minDepth, maxDepth);
	}

	setViewportWH(width, height){
		this.renderPassEncoder.setViewport(0, 0, width, height, 0, 1);
	}

	setVertexBuffer(slot, buffer, offset, size){
		this.renderPassEncoder.setVertexBuffer(slot, buffer, offset, size);
	}

	setIndexBuffer(buffer, indexFormat, offset, size){
		this.renderPassEncoder.setIndexBuffer(buffer, indexFormat, offset, size);
	}

	setBindGroup(index, bindGroup){
		this.renderPassEncoder.setBindGroup(index, bindGroup);
	}

	draw(vertexCount, instanceCount=1, firstVertex=0, firstInstance=0){
		this.renderPassEncoder.draw(vertexCount, instanceCount, firstVertex, firstInstance);
	}

	drawIndexed(indexCount, instanceCount=1, firstIndex=0, baseVertex=0, firstInstance=0){
		this.renderPassEncoder.drawIndexed(indexCount, instanceCount, firstIndex, baseVertex, firstInstance);
	}
}

exports.GPUContext = GPUContext;
