// webgpu/core/GPUContext.js

class GPUContext
{
	constructor(renderPassEncoder){
		this.renderPassEncoder = renderPassEncoder;
		this.pipeline = null;
		this.resourcesQueue = [];
		this.index = null;
	}

	setPipeline(pipeline){
		if(this.pipeline == pipeline)return;
		pipeline.setPipeline(this.renderPassEncoder);
		if(this.pipeline == null){
			for(let resources of this.resourcesQueue){
				pipeline.setResources(this.renderPassEncoder, resources);
			}
		}
		this.pipeline = pipeline;
	}

	setResources(resources){
		const {pipeline} = this;
		if(pipeline){
			pipeline.setResources(this.renderPassEncoder, resources);
		}else{
			this.resourcesQueue.push(resources);
		}
		this.index = resources.index ?? this.index;
	}

	setViewportWH(width, height){
		this.renderPassEncoder.setViewport(0, 0, width, height, 0, 1);
	}

	drawIndexedAuto(instanceCount=1, firstInstance=0){
		this.renderPassEncoder.drawIndexed(this.index.length, instanceCount, 0, 0, firstInstance);
	}

	draw(vertexCount, instanceCount=1, firstVertex=0, firstInstance=0){
		this.renderPassEncoder.draw(vertexCount, instanceCount, firstVertex, firstInstance);
	}
/*
	drawIndexed(indexCount, instanceCount=1, firstIndex=0, baseVertex=0, firstInstance=0){
		this.renderPassEncoder.drawIndexed(indexCount, instanceCount, firstIndex, baseVertex, firstInstance);
	}

	setViewport(x, y, width, height, minDepth=0, maxDepth=1){
		this.renderPassEncoder.setViewport(x, y, width, height, minDepth, maxDepth);
	}

	setVertexBuffer(slot, buffer, offset, size){
		this.renderPassEncoder.setVertexBuffer(slot, buffer, offset, size);
	}

	setIndexBuffer(buffer, indexFormat, offset, size){
		this.renderPassEncoder.setIndexBuffer(buffer, indexFormat, offset, size);
	}

	setBindGroup(index, bindGroup){
		this.renderPassEncoder.setBindGroup(index, bindGroup);
	}

	setBlendConstant(color){
		this.renderPassEncoder.setBlendConstant(color);
	}

	setScissorRect(x, y, width, height){
		this.renderPassEncoder.setScissorRect(x, y, width, height);
	}

	setStencilReference(reference){
		this.renderPassEncoder.setStencilReference(reference);
	}
*/
}

exports.GPUContext = GPUContext;
