// js/acorn_print.js
'use strict';

const collectArgs = node => node.arguments.map(walk);
const getProperty = node => node.computed ? '[' + walk(node.property) + ']' : '.' + node.property.name;

const addTab = ((...args) => code => code.replace(...args))(/^/gm, ' '.repeat(4));

const handler = {
	ImportDeclaration(node){
		if(node.specifiers.length <= 0){
			return `import ${walk(node.source)}`;
		}
		let specifiers;
		let a = node.specifiers.filter(v => v.type != 'ImportSpecifier').map(walk);
		let b = node.specifiers.filter(v => v.type == 'ImportSpecifier').map(walk);
		if(a.length > 0 && b.length > 0){
			specifiers = a.join(', ') + ', {' + b.join(', ') + '}';
		}else if(a.length > 0){
			specifiers = a.join(', ');
		}else if(b.length > 0){
			specifiers = '{' + b.join(', ') + '}';
		}
		return `import ${specifiers} from ${walk(node.source)}`;
	},
	ImportDefaultSpecifier(node){
		return walk(node.local);
	},
	ImportNamespaceSpecifier(node){
		return '* as ' + walk(node.local);
	},
	ImportSpecifier(node){
		let name = walk(node.imported);
		let alias = walk(node.local);
		return name == alias ? name : `${name} as ${alias}`;
	},
	ExportNamedDeclaration(node){
		let result = 'export ';
		if(node.declaration){
			result += walk(node.declaration);
		}
		if(node.specifiers.length > 0){
			result += '{' + node.specifiers.map(walk).join(', ') + '}';
		}
		if(node.source){
			result += ' from ' + walk(node.source);
		}
		return result;
	},
	ExportDefaultDeclaration(node){
		return 'export derault ' + walk(node.declaration);
	},
	ExportSpecifier(node){
		let name = walk(node.local);
		let alias = walk(node.exported);
		return name == alias ? name : `${name} as ${alias}`;
	},
	ExportAllDeclaration(node){
		return 'export * from ' + walk(node.source);
	},
	ClassDeclaration(node){
		return `class ${walk(node.id)}${node.superClass ? ' extends ' + walk(node.superClass) : ''} {\n${walk(node.body)}\n}`;
	},
	ClassBody(node){
		return addTab(handler.Program(node));
	},
	MethodDefinition(node){
		let prefix = node.static ? 'static ' : '';
		let key = walk(node.key);
		if(node.computed){
			key = '[' + key + ']';
		}
		if(node.value.type == 'FunctionExpression'){
			return prefix + handler.FunctionExpression(node.value, key);
		}
		return prefix + walk(node.key) + ':' + walk(node.value);
	},
	Program(node){
		return node.body.map(walk).join('\n');
	},
	BlockStatement(node){
		return handler.Program(node);
	},
	Identifier(node){
		return node.name;
	},
	FunctionDeclaration(node){
		return handler.FunctionExpression(node);
	},
	ArrowFunctionExpression(node){
		const body = walk(node.body);
		let params;
		if(node.params.length == 1 && node.params[0].type == 'Identifier'){
			params = walk(node.params[0]);
		}else{
			params = `(${node.params.map(walk)})`;
		}
		const result = params + ` => `;
		if(node.expression){
			return result + body;
		}
		return result + `{\n${addTab(body)}\n}`;
	},
	FunctionExpression(node, method){
		const body = walk(node.body);
		return `${node.async ? 'async ': ''}${method || 'function'}${node.generator ? '*' : ' '}${node.id ? node.id.name : ''}(${node.params.map(walk)}){\n${addTab(body)}\n}`;
	},
	AwaitExpression(node){
		return `await ` + walk(node.argument);
	},
	YieldExpression(node){
		return `yield` + (node.delegate ? '*' : ' ') + walk(node.argument);
	},
	ExpressionStatement(node){
		return walk(node.expression) + ';';
	},
	CallExpression(node){
		return walk(node.callee) + '(' + collectArgs(node) + ')';
	},
	SpreadElement(node){
		return '...' + walk(node.argument);
	},
	RestElement(node){
		return '...' + walk(node.argument);
	},
	MemberExpression(node){
		return walk(node.object) + getProperty(node);
	},
	EmptyStatement(node){
		return '';
	},
	UnaryExpression(node){
		return node.operator + ' ' + walk(node.argument);
	},
	BinaryExpression(node){
		return walk(node.left) + ' ' + node.operator + ' ' + walk(node.right);
	},
	AssignmentExpression(node){
		return handler.BinaryExpression(node);
	},
	LogicalExpression(node){
		return handler.BinaryExpression(node);
	},
	Literal(node){
		return node.raw;
	},
	VariableDeclaration(node){
		return node.kind + ' ' + node.declarations.map(walk);
	},
	VariableDeclarator(node){
		if(node.init){
			return walk(node.id) + ' = ' + walk(node.init);
		}
		return walk(node.id);
	},
	ArrayPattern(node){
		return handler.ArrayExpression(node);
	},
	ObjectPattern(node){
		return handler.ObjectExpression(node);
	},
	AssignmentPattern(node){
		return walk(node.left) + ' = ' + walk(node.right);
	},
	ReturnStatement(node){
		if(node.argument){
			return 'return ' + walk(node.argument);
		}
		return 'return';
	},
	ConditionalExpression(node){
		return walk(node.test) + ' ? ' + walk(node.consequent) + ' : ' + walk(node.alternate);
	},
	IfStatement(node){
		let result = `if(${walk(node.test)}){\n` + addTab(walk(node.consequent)) + `\n}`;
		if(!node.alternate)return result;
		const {alternate} = node;
		if(alternate.type == 'IfStatement'){
			return result + `else ${walk(alternate)}`;
		}
		return result + `else{\n` + addTab(walk(alternate)) + '\n}';
	},
	WhileStatement(node){
		return `while(${walk(node.test)}){\n${addTab(walk(node.body))}\n}`;
	},
	DoWhileStatement(node){
		return `do{\n${addTab(walk(node.body))}\n}while(${walk(node.test)});`;
	},
	ForStatement(node){
		return `for(${walk(node.init)};${walk(node.test)};${walk(node.update)}){\n${addTab(walk(node.body))}\n}`;
	},
	ForInStatement(node){
		return `for(${walk(node.left)} in ${walk(node.right)}){\n${addTab(walk(node.body))}\n}`;
	},
	ForOfStatement(node){
		return `for(${walk(node.left)} of ${walk(node.right)}){\n${addTab(walk(node.body))}\n}`;
	},
	BreakStatement(node){
		return node.label ? 'break ' + node.label : 'break';
	},
	ContinueStatement(node){
		return node.label ? 'continue ' + node.label : 'continue';
	},
	LabeledStatement(node){
		return walk(node.label) + ':\n' + walk(node.body);
	},
	UpdateExpression(node){
		let result = walk(node.argument);
		if(node.prefix){
			return node.operator + result;
		}
		return result + node.operator;
	},
	ArrayExpression(node){
		return '[' + node.elements.map(walk) + ']';
	},
	ObjectExpression(node){
		return `{${node.properties.map(walk)}}`;
	},
	Property(node){
		let key = walk(node.key);
		if(node.shorthand){
			return key;
		}
		if(node.computed){
			key = `[${key}]`;
		}
		if(node.method){
			return handler.FunctionExpression(node.value, key);
		}
		return key + ':' + walk(node.value);
	},
	NewExpression(node){
		return 'new ' + handler.CallExpression(node);
	},
	DebuggerStatement(node){
		return 'debugger';
	},
	ThisExpression(node){
		return 'this';
	},
	Super(node){
		return 'super';
	},
	SequenceExpression(node){
		return '(' + node.expressions.map(walk) + ')';
	},
	SwitchStatement(node){
		const body = addTab(node.cases.map(walk).join('\n'));
		return `switch(${walk(node.discriminant)}){\n${body}\n}`;
	},
	SwitchCase(node){
		const body = addTab(node.consequent.map(walk).join('\n'));
		if(node.test){
			return `case ${walk(node.test)}:\n${body}`;
		}
		return `default:\n${body}`;
	},
	ThrowStatement(node){
		return 'throw ' + walk(node.argument);
	},
	TryStatement(node){
		let result = `try{\n${addTab(walk(node.block))}\n}`;
		if(node.handler){
			result += walk(node.handler);
		}
		if(node.finalizer){
			result += `finally{\n${addTab(walk(node.finalizer))}\n}`;
		}
		return result;
	},
	CatchClause(node){
		return `catch(${walk(node.param)}){\n${addTab(walk(node.body))}\n}`;
	},
	WithStatement(node){
		return `with(${walk(node.object)}){\n${addTab(walk(node.body))}\n}`;
	},
	TemplateLiteral(node){
		return '`' + walk(node.quasis[0]) + node.expressions.map((v, i) => '${' + walk(v) + '}' + walk(node.quasis[i+1])).join('') + '`';
	},
	TemplateElement(node){
		return node.value.raw;
	},
	TaggedTemplateExpression(node){
		return walk(node.tag) + walk(node.quasi);
	}
};

function walk(node){
	if(node == null)return '';
	if(!handler[node.type])console.error(node);
	return handler[node.type](node);
}

if(__filename == process.argv[1]){
	const acorn = require('./acorn.js');
	const fs = require('fs');
	console.log(walk(acorn.parse(fs.readFileSync(process.argv[2], 'utf8'))));
}

module.exports = walk;
