// opengl/managers/MouseManager.js
"use strict";

const WebGL = require('../WebGL');
const {MouseRenderTarget} = require("../rendertargets");
const {DrawMode} = require("../renderers");

const colorBlack = new Float32Array(4);

/* draw 2d first, if done, return else draw 3d */
class MouseManager
{
	constructor(){
		this.requestPick = false;
		this.objectList = [null];
		this.init();
	}

	init(){
		const {canvas} = WebGL;

		let mouseEventHandler = this.onMouseEvent.bind(this);
		for(let evtType of ["mousedown", "mouseup", "mousemove"]){
			canvas.addEventListener(evtType, mouseEventHandler);
		}

		this.renderTarget = new MouseRenderTarget(canvas.width, canvas.height);

		let buffer = new ArrayBuffer(4);
		this._pickData = new Uint8Array(buffer);
		this._pickValue = new Uint32Array(buffer);
	}

	update(view3d){
		if(!this.requestPick){
			return;
		}
		this.requestPick = false;
		const {mouseX, mouseY} = this;
		const {gl, height} = WebGL;

		this.renderTarget.active(gl.FRAMEBUFFER);
		
		this._pickBegin();

		gl.clearBufferfv(gl.COLOR, 0, colorBlack);
		gl.clearBufferfi(gl.DEPTH_STENCIL, 0, 1, 0);

		gl.enable(gl.STENCIL_TEST);

		gl.stencilFunc(gl.ALWAYS, 1, 0xFF);
		gl.stencilOp(gl.KEEP, gl.KEEP, gl.REPLACE);

		view3d.scene2d.draw(DrawMode.Pick);

		gl.stencilFunc(gl.NOTEQUAL, 1, 0xFF);
		gl.stencilOp(gl.KEEP, gl.KEEP, gl.KEEP);

		//view3d.scene3d.draw(DrawMode.Pick);
		const {camera} = view3d.scene3d;
		if(camera)camera.renderPicking();

		gl.disable(gl.STENCIL_TEST);

		gl.readPixels(mouseX, height - mouseY, 1, 1, gl.RGBA, gl.UNSIGNED_BYTE, this._pickData);//gl.FLOAT, new Float32Array
		
		gl.bindFramebuffer(gl.FRAMEBUFFER, null);

		let mouseTarget = this.objectList[this._pickValue[0]];
		if(this.mouseEvent.type == "mouseup")
			console.log(mouseTarget, this._pickValue[0], "click");
	}

	onMouseEvent(evt){
		this.requestPick = true;
		this.mouseEvent = evt;
		this.mouseX = evt.x;
		this.mouseY = evt.y;
	}

	_pickBegin(){
		this.objectList.length = 1;
	}

	pickRegister(target){
		let index = this.nextPickID;
		this.objectList[index] = target;
		return index;
	}

	get nextPickID(){
		return this.objectList.length;
	}
}

module.exports = MouseManager;
