// uniformbuffers/UniformBuffer.js
"use strict";

class UniformBuffer
{
	constructor(registerCount, baseDataCount=0, instanceCount=1){
		this.registerCount = registerCount;
		this.instanceCount = instanceCount;
		this.baseDataCount = baseDataCount;
		this.data = new ArrayBuffer(this.calcValueCount(instanceCount) << 2);
		this.floatView = new Float32Array(this.data);
		this.intView = new Int32Array(this.data);
	}

	active(gl, index=0){
		gl.bindBufferBase(gl.UNIFORM_BUFFER, index, this.buffer || this.createUniformBuffer(gl));
	}

	upload(gl, instanceCount=1){
		gl.bufferSubData(gl.UNIFORM_BUFFER, 0, this.floatView, 0, this.calcValueCount(instanceCount));
	}

	createUniformBuffer(gl){
		this.buffer = gl.createBuffer();
		gl.bindBuffer(gl.UNIFORM_BUFFER, this.buffer);
		gl.bufferData(gl.UNIFORM_BUFFER, this.data.byteLength, gl.DYNAMIC_DRAW);
		return this.buffer;
	}

	calcValueCount(instanceCount){
		let {baseDataCount, registerCount} = this;
		return baseDataCount + registerCount * instanceCount << 2;
	}

	setInts(index, ...values){
		this.intView.set(values, index);
	}

	setIntv(index, values){
		this.intView.set(values, index);
	}

	setFloats(index, ...values){
		this.floatView.set(values, index);
	}

	setFloatv(index, values){
		this.floatView.set(values, index);
	}
}

module.exports = UniformBuffer;
