// opengl/materials/TerrainMaterial.js
"use strict";

const Material = require("./Material");
const {RenderState} = require("../renderers");
const WebGL = require('../WebGL');
const {ProgramName} = require('../const');

class TerrainMaterial extends Material{
	constructor(texture){
		super(ProgramName.terrain);
		this.texture = texture;
		//this.renderState = RenderState.Texture;
	}
}

TerrainMaterial.renderState = RenderState.Texture;
TerrainMaterial.bindTexture = (gl, texture) => gl.bindTexture(gl.TEXTURE_2D_ARRAY, texture);
TerrainMaterial.onDrawBatch = function(drawUnitList, offset, instanceCount, batchIndex){
	let drawUnit = drawUnitList[offset];
	let {gl} = WebGL;
	let {vao} = drawUnit;
	//let program = gl.getParameter(gl.CURRENT_PROGRAM);
	//gl.uniform2i(gl.getUniformLocation(program, 'vertexCount'), vao.vertexCountX, vao.vertexCountY);
	//gl.uniform1f(gl.getUniformLocation(program, 'tileSize'), drawUnit.size);
	//gl.uniform1f(gl.getUniformLocation(program, 'uvScale'), drawUnit.uvScale);
	gl.uniform2i(this.address_vertexCount, vao.vertexCountX, vao.vertexCountY);
	gl.uniform1f(this.address_tileSize, drawUnit.size);
	gl.uniform1f(this.address_uvScale, drawUnit.uvScale);
}
TerrainMaterial.initUniformLocation = function(gl, program){
	this.address_vertexCount = gl.getUniformLocation(program, 'vertexCount');
	this.address_tileSize = gl.getUniformLocation(program, 'tileSize');
	this.address_uvScale = gl.getUniformLocation(program, 'uvScale');
}

module.exports = TerrainMaterial;
