// opengl/rendertargets/RenderTarget.js
"use strict";

const WebGL = require('../WebGL');

class RenderTarget{
	constructor(width, height){
		this.width = width;
		this.height = height;
	}

	active(){
		let {gl} = WebGL;
		gl.bindFramebuffer(gl.FRAMEBUFFER, this.frameBuffer || this.createFrameBuffer(gl));
	}

	dispose(){
		if(!this.frameBuffer)return;
		let {gl} = WebGL;
		this.onDispose(gl);
		gl.deleteFramebuffer(this.frameBuffer);
		this.frameBuffer = null;
	}

	createFrameBuffer(gl){
		this.frameBuffer = gl.createFramebuffer();
		gl.bindFramebuffer(gl.FRAMEBUFFER, this.frameBuffer);
		this.onCreate(gl);
		gl.bindFramebuffer(gl.FRAMEBUFFER, null);
		return this.frameBuffer;
	}

	attachTexture2D(attachment, texture){
		let {gl} = WebGL;
		gl.framebufferTexture2D(gl.FRAMEBUFFER, attachment, gl.TEXTURE_2D, texture, 0);
	}

	attachRenderbuffer(attachment, renderbuffer){
		let {gl} = WebGL;
		gl.framebufferRenderbuffer(gl.FRAMEBUFFER, attachment, gl.RENDERBUFFER, renderbuffer);
	}

	onCreate(gl){}
	onDispose(gl){}
}

module.exports = RenderTarget;
