// opengl/rendertargets/RenderTarget.js
"use strict";

const WebGL = require('../WebGL');

class RenderTarget{
	constructor(width, height){
		this.width = width;
		this.height = height;
	}

	active(){
		let {gl} = WebGL;
		gl.bindFramebuffer(gl.FRAMEBUFFER, this.frameBuffer || this.createFrameBuffer(gl));
	}

	dispose(){
		if(!this.frameBuffer)return;
		let {gl} = WebGL;
		this.onDispose(gl);
		gl.deleteFramebuffer(this.frameBuffer);
		this.frameBuffer = null;
	}

	createFrameBuffer(gl){
		this.frameBuffer = gl.createFramebuffer();
		gl.bindFramebuffer(gl.FRAMEBUFFER, this.frameBuffer);
		this.onCreate(gl);
		throwError(gl.checkFramebufferStatus(gl.FRAMEBUFFER));
		gl.bindFramebuffer(gl.FRAMEBUFFER, null);
		return this.frameBuffer;
	}

	attachTexture2D(attachment, texture){
		let {gl} = WebGL;
		gl.framebufferTexture2D(gl.FRAMEBUFFER, attachment, gl.TEXTURE_2D, texture, 0);
	}

	attachRenderbuffer(attachment, renderbuffer){
		let {gl} = WebGL;
		gl.framebufferRenderbuffer(gl.FRAMEBUFFER, attachment, gl.RENDERBUFFER, renderbuffer);
	}

	onCreate(gl){}
	onDispose(gl){}
}

const throwError = function(list){
	let codeList = list.map(v => WebGL2RenderingContext[v]);
	return status => status != codeList[0] && console.error(list[codeList.indexOf(status)]);
}([
	'FRAMEBUFFER_COMPLETE',
	'FRAMEBUFFER_INCOMPLETE_ATTACHMENT',
	'FRAMEBUFFER_INCOMPLETE_MISSING_ATTACHMENT',
	'FRAMEBUFFER_INCOMPLETE_DIMENSIONS',
	'FRAMEBUFFER_UNSUPPORTED',
	'FRAMEBUFFER_INCOMPLETE_MULTISAMPLE',
	'RENDERBUFFER_SAMPLES'
]);

module.exports = RenderTarget;
