// opengl/vertexarrays/MeshVertexArray.js

const WebGL = require('../WebGL');
const VertexArray = require('./VertexArray');

class MeshVertexArray extends VertexArray
{
	static new(subMesh){
		let vao = subMesh.__vao__;
		if(!vao){
			vao = new this(subMesh);
			subMesh.__vao__ = vao;
		}
		return vao;
	}

	constructor(subMesh){
		super(subMesh.cpuMode && WebGLRenderingContext.DYNAMIC_DRAW);
		this.subMesh = subMesh;
		if(subMesh.cpuMode){
			this.vertexData = subMesh.copyVertexDataForCPU();
		}else{
			this.maxCountPerDraw = Math.floor(1024 / (subMesh.boneCount << 1));//default value is 1
		}
	}

	onCreate(gl){
		const {vertexData, indexData} = this.subMesh;
		super.onCreate(gl, vertexData, indexData);
		this.subMesh.onCreateVAO(gl);
	}

	draw(instanceCount=1){
		let {gl} = WebGL;
		this.subMesh.onDraw(gl, instanceCount);
	}

	onActive(gl){
		this.subMesh.onActive(gl);
	}

	updateVertexDataForCPU(boneMatrixList, bindBufferFlag=true){
		let {subMesh, vertexData} = this;
		subMesh.updateVertexDataForCPU(vertexData, boneMatrixList);
		this.updateVertexBuffer(vertexData, bindBufferFlag);
	}
}

module.exports = MeshVertexArray;
