// opengl/vertexarrays/VertexArray.js
"use strict";

const WebGL = require('../WebGL');

class VertexArray
{
	active(){
		let {gl} = WebGL;
		gl.bindVertexArray(this.vao || this.createVertexArray(gl));
	}

	createVertexArray(gl){
		this.vao = gl.createVertexArray();
		gl.bindVertexArray(this.vao);
		this.onCreate(gl);
		gl.bindVertexArray(null);
		return this.vao;
	}

	onCreate(gl, vertexData, indexData){
		if(indexData){
			this.indexBuffer = gl.createBuffer();
			gl.bindBuffer(gl.ELEMENT_ARRAY_BUFFER, this.indexBuffer);
			gl.bufferData(gl.ELEMENT_ARRAY_BUFFER, indexData, gl.STATIC_DRAW);
		}
		if(vertexData){
			this.vertexBuffer = gl.createBuffer();
			gl.bindBuffer(gl.ARRAY_BUFFER, this.vertexBuffer);
			gl.bufferData(gl.ARRAY_BUFFER, vertexData, gl.STATIC_DRAW);
		}
	}

	dispose(){
		if(!this.vao)return;
		let {gl} = WebGL;
		this.onDispose(gl);
		gl.deleteVertexArray(this.vao);
		this.vao = null;
	}

	onDispose(gl){
		if(this.indexBuffer){
			gl.deleteBuffer(this.indexBuffer);
			this.indexBuffer = null;
		}
		if(this.vertexBuffer){
			gl.deleteBuffer(this.vertexBuffer);
			this.vertexBuffer = null;
		}
	}
}

module.exports = VertexArray;
