// webgpu/DrawUnitBucket.js
'use strict';

const ObjectCache = require('utils/ObjectCache');
const ObjectBucket = require('ds/ObjectBucket');

const mapObjCache = new ObjectCache(() => new Map(), v => v.clear());

const Vulkan = require('./Vulkan');

class DrawUnitBucket extends ObjectBucket
{
	constructor(){
		super();
		this.regKeyFn(getMaterialType, updateMaterialType, sortMaterialType);
		this.regKeyFn(getMesh, updateMesh);
	}

	newMap(){
		return mapObjCache.getOut();
	}

	clear(){
		super.clear();
		mapObjCache.putAllIn();
	}

	draw(){
		this.forEach(onMaterialDraw);
		return this;
	}
}

const onMaterialDraw = entityList => Vulkan.currentMaterial.draw(entityList);

const sortMaterialType = (fn => list => list.sort(fn))((a, b) => a.order - b.order);

function getMaterialType(drawUnit, key){
	return drawUnit.material.constructor;
}

function getMesh(drawUnit, key){
	return drawUnit.material.constructor.useMeshFlag ? drawUnit.entity.mesh : null;
}

function updateMaterialType(materialType){
	const {renderPassEncoder} = Vulkan;
	Vulkan.currentMaterial = materialType;
	renderPassEncoder.setPipeline(materialType.pipeline);
}

function updateMesh(mesh){
	Vulkan.currentMesh = mesh;
	if(!mesh)return;
	const {currentMaterial} = Vulkan;
	currentMaterial.setVertexBuffer();
	currentMaterial.setIndexBuffer();
}

module.exports = DrawUnitBucket;
