// webonly/serial/ESP32Serial.js

const ChromeSerial = require('webonly/chrome/ChromeSerial');
const {waitWhen} = require('utils/events');
const retry = require('utils/function/retry');
const zlib = require('zlib');
const crypto = require('crypto');

const codeText = require('./ESP32_bin');
const codeData = Buffer.from('CMD8Pw==', 'base64');

const {delay, ints_to_buffer, align_data, encode_c0db, checker_c0db} = require('./helper');

class Serial extends ChromeSerial
{
	async upload(data, offset=0x1000, progressFn=null){
		try{
			await this.sync(Buffer.alloc(36, 0x55));
			await this.write_stub();
			await this.write_flash(data, offset, progressFn);
			await this.check_flash_md5(data, offset);
		}finally{
			await this.setControlSignals({dtr:false, rts:true});
			await this.setControlSignals({dtr:false, rts:false});
		}
	}

	async sync(data){
		await this.setControlSignals({dtr:false, rts:true});
		await this.setControlSignals({dtr:true, rts:false});
		await delay(20);
		await retry(() => this.send_cmd(8, data, 20), 7);
		await delay(100);
	}

	recv(timeout=1000){
		return waitWhen(this, 'data', timeout, checker_c0db());
	}

	read_reg(address){
		return this.send_cmd(0xA, ints_to_buffer(address)).then(v => v[0]);
	}

	write_reg(address, value){
		return this.send_cmd(0x9, ints_to_buffer(address, value, 0xFFFFFFFF, 0));
	}

	async check_flash_md5(data, offset){
		const timeout = Math.max(3000, 8000 * data.length / 0x100000);//speed=8s/M
		const result = await this.send_cmd(0x13, ints_to_buffer(offset, data.length, 0, 0), timeout);
		const md5_old = crypto.createHash('md5').update(data).digest('hex');
		const md5_new = result[1].slice(0, 16).map(v => v.toString(16).padStart(2, '0')).join('');
		console.assert(md5_old === md5_new);
	}

	async change_bitrate(bitrate){
		await this.send_cmd(0xF, ints_to_buffer(bitrate, this.bitrate));
		await delay(1000)
		//await this.update(bitrate);
		//return
		await this.disconnect();
		await delay(50);
		await this.connect(this.port, bitrate);
	}

	async write_dataframe(cmd, frameSize, data, progressFn=null){
		const blockCount = Math.ceil(data.length / frameSize);
		for(let i=0; i<blockCount; ++i){
			const packet = data.slice(frameSize * i, frameSize * (i + 1));
			const payload = Buffer.concat([ints_to_buffer(packet.length, i, 0, 0), packet]);
			await this.send_cmd(cmd, payload);
			progressFn?.((frameSize * i + packet.length) / data.length);
		}
	}

	async write_stub(){
		const ESP_RAM_BLOCK = 0x1800;
		await this.send_cmd(5, ints_to_buffer(codeText.length, Math.ceil(codeText.length / ESP_RAM_BLOCK), ESP_RAM_BLOCK, 0x400be000));
		await this.write_dataframe(7, ESP_RAM_BLOCK, codeText);
		await this.send_cmd(5, ints_to_buffer(codeData.length, Math.ceil(codeData.length / ESP_RAM_BLOCK), ESP_RAM_BLOCK, 0x3ffdeba8));
		await this.write_dataframe(7, ESP_RAM_BLOCK, codeData);
		await this.send_cmd(6, ints_to_buffer(0, 0x400be598));
		console.assert(Buffer.from(await this.recv()).toString() == 'OHAI');
	}

	async write_flash(data, offset, progressFn=null){
		const FLASH_WRITE_SIZE = 0x4000;
		data = align_data(data, 4, 0xFF);
		const old_size = data.length;
		data = zlib.deflateSync(data, {level:9});
		await this.send_cmd(0x10, ints_to_buffer(old_size, Math.ceil(data.length / FLASH_WRITE_SIZE), FLASH_WRITE_SIZE, offset));
		await this.write_dataframe(0x11, FLASH_WRITE_SIZE, data, progressFn);
		await this.send_cmd(0x12, ints_to_buffer(1));//0==reboot
	}
}

Serial.prototype.send_cmd = function(){
	function checksum(data, offset){
		let result = 0xEF;
		for(let i=offset, n=data.length; i<n; ++i){
			result ^= data[i];
		}
		return result;
	}
	return async function(op, data, timeout=1000){
		const head = ints_to_buffer(op << 8 | data.length << 16, checksum(data, 16));
		await this.send_buffer(encode_c0db(Buffer.concat([head, data])));
		const result = await this.recv(timeout);
		console.assert(result.length === 8 + (result[2] | result[3] << 8));
		if(result[0] == 1 && result[1] == op){
			return [Buffer.from(result).readUInt32LE(4), result.slice(8)];
		}
		return Promise.reject();
	}
}();

module.exports = Serial;
