// vdom/signal.js

const {createSignal, createEffect} = function(){
	const effectStack = [];
	function createSignal(value){
		const listeners = new Set();
		function getFn(){
			const effect = effectStack[effectStack.length-1];
			if(effect)listeners.add(effect);
			return value;
		}
		function setFn(newValue){
			value = newValue;
			for(const effect of [...listeners]){
				effect.execute();
			}
		}
		return [getFn, setFn];
	}
	function createEffect(fn){
		const effect = {
			execute(){
				effectStack.push(effect);
				try{
					fn();
				}finally{
					effectStack.pop();
				}
			}
		}
		effect.execute();
	}
	return {createSignal, createEffect};
}();

function createMemo(fn){
	const [memo, setMemo] = createSignal();
	createEffect(() => setMemo(fn()));
	return memo;
}

exports.createSignal = createSignal;
exports.createEffect = createEffect;
exports.createMemo = createMemo;
/*
const [name, setName] = createSignal("a");
const [age, setAge] = createSignal(18);
const fullName = createMemo(() => "c-" + name());
createEffect(() => console.log(name(), fullName(), age()));
setName("b");
setAge(20);
*/
