// webonly/utils_gpu.js

const textureFormat = navigator.gpu.getPreferredCanvasFormat();//bgra8unorm
const textureUsage = GPUTextureUsage.TEXTURE_BINDING | GPUTextureUsage.COPY_DST | GPUTextureUsage.RENDER_ATTACHMENT;

async function requestDevice(){
	const adapter = await navigator.gpu.requestAdapter();
	const device = await adapter.requestDevice();
	return device;
}

function requestSwapChain(canvas, device){
	const swapChain = canvas.getContext('webgpu');
	swapChain.configure({
		device,
		format: textureFormat,
		usage: GPUTextureUsage.RENDER_ATTACHMENT,
		alphaMode: 'opaque',//premultiplied
		colorSpace: 'srgb',//display-p3
	});
	return swapChain;
}

function createDepthTexture(device, width, height){
	return device.createTexture({
		size: [width, height, 1],
		format: 'depth24plus-stencil8',
		usage: GPUTextureUsage.RENDER_ATTACHMENT
	});
}

function createTexture(device, width, height, depthOrArrayLayers=1, usage=textureUsage){
	return device.createTexture({
		dimension: '2d',
		format: textureFormat,
		size: [width, height, depthOrArrayLayers],
		usage,
	});
}

function createTextureWithImage(device, image){
	const texture = createTexture(device, image.width, image.height);
	device.queue.copyExternalImageToTexture(
		{source: image, origin:[0, 0], flipY:false},
		{texture, mipLevel: 0, origin: [0, 0, 0]},
		[image.width, image.height, 1]
	);
	return texture;
}

async function createTextureArray(device, imageList){
	const width = Math.max(...imageList.map(v => v.naturalWidth));
	const height = Math.max(...imageList.map(v => v.naturalHeight));
	const arrayLayerCount = imageList.length;
	const resizeOption = {resizeWidth: width, resizeHeight: height, resizeQuality: 'high'};
	const texture = createTexture(device, width, height, arrayLayerCount);
	for(let i=0; i<arrayLayerCount; ++i){
		device.queue.copyExternalImageToTexture(
			{source: await createImageBitmap(imageList[i], resizeOption)},
			{texture, origin: [0, 0, i]},
			[width, height, 1]
		);
	}
	return texture;
}

const createBuffer = (device, usage, size, mappedAtCreation=false) => device.createBuffer({usage, size, mappedAtCreation});

function createBufferWith(device, usage, data){
	const buffer = createBuffer(device, usage, data.byteLength, true);
	const mapping = buffer.getMappedRange();
	new data.constructor(mapping).set(data);
	buffer.unmap();
	return buffer;
}

function setViewport(renderPassEncoder, rect, minDepth=0, maxDepth=1){
	renderPassEncoder.setViewport(rect.x, rect.y, rect.width, rect.height, minDepth, maxDepth);
}

const calcBindGroupLayoutSign = function(){
	function* sign(desc){
		yield desc.binding;
		yield desc.visibility;
		if(desc.buffer){
			yield desc.buffer.type ?? 'uniform';//storage, read-only-storage
			yield desc.buffer.hasDynamicOffset ? 1 : 0;
			yield desc.buffer.minBindingSize ?? 0;
		}else if(desc.sampler){
			yield desc.sampler.type ?? 'filtering';//non-filtering, comparison
		}else if(desc.storageTexture){
			yield desc.storageTexture.access ?? 'write-only';//read-only, read-write
			yield desc.storageTexture.viewDimension ?? '2d';//1d, 3d, 2d-array, cube, cube-array
			yield desc.storageTexture.format;
		}else if(desc.texture){
			yield desc.texture.sampleType ?? 'float';//depth, sint, uint, unfilterable-float
			yield desc.texture.viewDimension ?? '2d';
			yield desc.texture.multisampled ? 1 : 0;
		}else if(desc.externalTexture){
		}
	}
	const sortFn = (a, b) => a.binding - b.binding;
	const mapFn = v => Array.from(sign(v)).join(',');
	return entries => entries.sort(sortFn).map(mapFn).join('&');
}();
const calcPipelineLayoutSign = bindGroupLayouts => bindGroupLayouts.map(calcBindGroupLayoutSign).join('|');

const calcSamplerSign = function(){
	function* sign(desc){
		yield desc.addressModeU ?? 'clamp-to-edge';//repeat, mirror-repeat
		yield desc.addressModeV ?? 'clamp-to-edge';
		yield desc.addressModeW ?? 'clamp-to-edge';
		yield desc.minFilter ?? 'nearest';//linear
		yield desc.magFilter ?? 'nearest';
		yield desc.mipmapFilter ?? 'nearest';
		yield desc.lodMinClamp ?? 0;
		yield desc.lodMaxClamp ?? 32;
		yield desc.maxAnisotropy ?? 1;//1-16
		yield desc.compare;//never, always, equal, not-equal, less, less-equal, greater, greater-equal
	}
	return desc => Array.from(sign(desc)).join(',');
}();

exports.textureFormat = textureFormat;
exports.requestDevice = requestDevice;
exports.requestSwapChain = requestSwapChain;
exports.createTexture = createTexture;
exports.createTextureWithImage = createTextureWithImage;
exports.createTextureArray = createTextureArray;
exports.createDepthTexture = createDepthTexture;
exports.createBuffer = createBuffer;
exports.createBufferWith = createBufferWith;
exports.setViewport = setViewport;
exports.calcBindGroupLayoutSign = calcBindGroupLayoutSign;
exports.calcPipelineLayoutSign = calcPipelineLayoutSign;
exports.calcSamplerSign = calcSamplerSign;
