// webonly/utils_gpu.js

function createTexture(device, image){
	const texture = device.createTexture({
		size: [image.width, image.height, 1],
		format: 'rgba8unorm',
		usage: GPUTextureUsage.TEXTURE_BINDING | GPUTextureUsage.COPY_DST | GPUTextureUsage.RENDER_ATTACHMENT,
	});
	device.queue.copyExternalImageToTexture(
		{source: image, origin:[0, 0], flipY:false},
		{texture, mipLevel: 0, origin: [0, 0, 0]},
		[image.width, image.height, 1]
	);
	return texture;
}

function createDepthTexture(device, width, height){
	return device.createTexture({
		size: [width, height, 1],
		format: 'depth24plus-stencil8',
		usage: GPUTextureUsage.RENDER_ATTACHMENT
	});
}

const createBuffer = (device, usage, size, mappedAtCreation=false) => device.createBuffer({usage, size, mappedAtCreation});

function createBufferWith(device, usage, data){
	const buffer = createBuffer(device, usage, data.byteLength, true);
	const mapping = buffer.getMappedRange();
	new data.constructor(mapping).set(data);
	buffer.unmap();
	return buffer;
}

function setViewport(renderPassEncoder, rect, minDepth=0, maxDepth=1){
	renderPassEncoder.setViewport(rect.x, rect.y, rect.width, rect.height, minDepth, maxDepth);
}

exports.createTexture = createTexture;
exports.createDepthTexture = createDepthTexture;
exports.createBuffer = createBuffer;
exports.createBufferWith = createBufferWith;
exports.setViewport = setViewport;
