// cameras/Camera3D.js
"use strict";

const DisplayObject3D = require("../DisplayObject3D");
const DrawUnitCollector = require("../DrawUnitCollector");
const Matrix3D = require("../Matrix3D");

class Camera3D extends DisplayObject3D
{
	constructor(lens){
		super();
		this.drawUnitCollector = new DrawUnitCollector();
		this.worldTransformInvert = new Matrix3D();
		this.lens = lens;
		//camera can render to back buffer or frame buffer
		this.randerTarget = null;
		//multi cameras render order control
		this.depth = 0;
	}

	init(){
		const {canvas, gl} = view3d;

		//store model world space XYZ
		this.geomDepthBuffer = gl.createTexture();
		gl.bindTexture(gl.TEXTURE_2D, this.geomDepthBuffer);
		gl.texImage2D(gl.TEXTURE_2D, 0, gl.DEPTH_COMPONENT32F, canvas.width, canvas.height, 0, gl.DEPTH_COMPONENT, gl.FLOAT, null);
		gl.texParameteri(gl.TEXTURE_2D, gl.TEXTURE_MIN_FILTER, gl.NEAREST);
		gl.texParameteri(gl.TEXTURE_2D, gl.TEXTURE_MAG_FILTER, gl.NEAREST);
		gl.texParameteri(gl.TEXTURE_2D, gl.TEXTURE_WRAP_S, gl.CLAMP_TO_EDGE);
		gl.texParameteri(gl.TEXTURE_2D, gl.TEXTURE_WRAP_T, gl.CLAMP_TO_EDGE);
		gl.bindTexture(gl.TEXTURE_2D, null);

		this.frameBuffer = gl.createFramebuffer();
		gl.bindFramebuffer(gl.FRAMEBUFFER, this.frameBuffer);
		gl.framebufferTexture2D(gl.FRAMEBUFFER, gl.DEPTH_ATTACHMENT, gl.TEXTURE_2D, this.geomDepthBuffer, 0);
		gl.bindFramebuffer(gl.FRAMEBUFFER, null);
	}

	render(root, gl, mode){
		let {drawUnitCollector} = this;
		drawUnitCollector.clear();
		root.collectDrawUnits(drawUnitCollector, gl, mode);
		drawUnitCollector.draw(gl, this);
	}

	draw(){
		const lightCount = this.getLightCount();
		if(lightCount <= 0){
			return;
		}

		for(let lightIndex=0; lightIndex<lightCount; ++lightIndex){
			let light = this.lightList[lightIndex];
			light.drawShadowMap();
		}

		gl.bindFramebuffer(gl.FRAMEBUFFER, this.frameBuffer);

		gl.clearColor(0, 0, 0, 0);
		gl.clear(gl.COLOR_BUFFER_BIT | gl.DEPTH_BUFFER_BIT);
		//draw

		gl.bindFramebuffer(gl.FRAMEBUFFER, null);

		gl.bindTexture(gl.TEXTURE_2D, this.geomColorBuffer);
		//blend MULTIPLY

		for(let lightIndex=0; lightIndex<lightCount; ++lightIndex){
			let light = this.lightList[lightIndex];
			light.drawLight();
		}
	}
}

module.exports = Camera3D;
