// webgpu/cameras/Camera2D.js
'use strict';

const Material2D = require('../materials/Material2D');
const Matrix2D = require('../math/Matrix2D');
const {walkTree} = require('utils/traverseTree');

class Camera2D
{
	constructor(lens){
		this.worldTransformInvert = new Matrix2D();
		this.lens = [0, 0, -1, 1];
		//this.layerFlagFilter = 0xFFFFFFFF;
		//this.viewportRect = new Float32Array([0, 0, 1, 1]);
		//this.projectionMatrix = new Array(16);
		this.drawUnitList = [];

		if(lens)this.setLens(lens);
	}

	setLens(value){
		const {width, height} = value;
		const {lens} = this;
		lens[0] = 2 / width;
		lens[1] = -2 / height;
	}
}

Camera2D.prototype.collectDrawUnits = function(){
	const fn = walkTree(function(item, root, forceDrawFlag, result){
		if(!item.visible)return true;
		if(item.filter && !(item == root && forceDrawFlag)){
			const {texture} = item.filter;
			const drawUnit = Object.create(item);
			drawUnit.texture = texture;
			drawUnit.width = texture.width;
			drawUnit.height = texture.height;
			result.push(drawUnit);
			return true;
		}else if(item.texture){
			result.push(item);
		}
	});
	return function(root, forceDrawFlag=false){
		fn(root, true, root, forceDrawFlag, this.drawUnitList);
	}
}();

Camera2D.prototype.draw = function(){
	const tempBuffer = new Float32Array(12);
	return function(){
		const {drawUnitList} = this;
		if(drawUnitList.length <= 0)return;
		tempBuffer.set(this.lens);
		this.worldTransformInvert.copyToArray(tempBuffer, 4);
		Material2D.setBindGroup(0, [{binding: 0, resource:{buffer:tempBuffer}}]);
		Material2D.draw(drawUnitList);
		drawUnitList.length = 0;
	};
}();

module.exports = Camera2D;
