// webgpu/cameras/Camera2D.js
'use strict';

const Vulkan = require('../Vulkan');
const Material2D = require('../materials/Material2D');
const Matrix2D = require('../math/Matrix2D');
const {walkTree} = require('utils/traverseTree');
const DrawUnitCollector2D = require('./DrawUnitCollector2D');
const RenderPass = require('../helper/RenderPass');
const {fetchView3D, fetchView2D} = require('../Vulkan/resource');
const {intersect} = require('../math/rectangle');
const {createWeakMapGetter} = require('utils/cache');

class Camera2D
{
	constructor(lens){
		this.worldTransformInvert = new Matrix2D();
		this.lens = [0, 0, -1, 1];
		//this.layerFlagFilter = 0xFFFFFFFF;
		//this.viewportRect = new Float32Array([0, 0, 1, 1]);
		//this.projectionMatrix = new Array(16);
		this.drawUnitCollector = new DrawUnitCollector2D();

		if(lens)this.setLens(lens);
	}

	setLens(value){
		const {width, height} = value;
		const {lens} = this;
		lens[0] = 2 / width;
		lens[1] = -2 / height;
	}
}

Camera2D.prototype.collectDrawUnits = function(){
	const fn = walkTree(function(item, root, forceDrawFlag, drawUnitCollector){
		if(!item.visible)return true;
		if(item.filter && !(item == root && forceDrawFlag)){
			const {texture} = item.filter;
			const drawUnit = Object.create(item);
			drawUnit.texture = texture;
			drawUnit.width = texture.width;
			drawUnit.height = texture.height;
			drawUnitCollector.addBitmap(drawUnit);
			drawUnitCollector.addFilter(item);
			return true;
		}else if(item.texture){
			drawUnitCollector.addBitmap(item);
			if(item.type == 'bitmap3d'){
				drawUnitCollector.addBitmap3D(item);
			}
		}
	});
	return function(root, forceDrawFlag=false){
		fn(root, true, root, forceDrawFlag, this.drawUnitCollector);
	}
}();

Camera2D.prototype.draw = function(){
	const tempBuffer = new Float32Array(12);
	return function(renderPassEncoder){
		const {bitmap3DList, filterList, bitmapList} = this.drawUnitCollector;
		if(bitmap3DList.length > 0){
			const commandEncoder = Vulkan.device.createCommandEncoder();
			for(let item of bitmap3DList)drawBitmap3D(commandEncoder, item);
			Vulkan.device.queue.submit([commandEncoder.finish()]);
		}
		if(filterList.length > 0){
			const commandEncoder = Vulkan.device.createCommandEncoder();
			for(let item of filterList)drawFilter2D(commandEncoder, item);
			Vulkan.device.queue.submit([commandEncoder.finish()]);
		}
		if(bitmapList.length > 0){
			tempBuffer.set(this.lens);
			this.worldTransformInvert.copyToArray(tempBuffer, 4);
			Material2D.setBindGroup(renderPassEncoder, 0, [{binding: 0, resource:{buffer:tempBuffer}}]);
			Material2D.draw(bitmapList, renderPassEncoder);
			this.drawUnitCollector.clear();
		}
	};
}();

const drawBitmap3D = function(){
	const renderPass3D = RenderPass({
		colorAttachments:[{
			loadOp:'clear',//load
			clearValue:[0,0,0,0],//when loadOp is clear
			storeOp:'store',//discard
			view:null
		}],
		depthStencilAttachment:{
			depthLoadOp:'clear',//load
			depthClearValue:1,//[0,1],when depthLoadOp is clear
			depthStoreOp:'discard',//store
			depthReadOnly:false,
			stencilLoadOp:'clear',//load
			stencilClearValue:0,
			stencilStoreOp:'discard',//store
			stencilReadOnly:false,
			view:null
		}
	});
	return (commandEncoder, item) => {
		const {colorView, depthView, width, height} = fetchView3D(item);
		item.texture.rawTexture = colorView;
		item.texture.setUV(width, height, {x:0,y:0,width:item.width,height:item.height}, false);
		item.camera.collectDrawUnits(item.root3d);
		for(let renderPassEncoder of renderPass3D(commandEncoder, colorView, depthView)){
			renderPassEncoder.setViewport(0, 0, item.width, item.height, 0, 1);
			item.camera.draw(renderPassEncoder);
		}
	}
}();

const drawFilter2D = function(){
	const renderPass2D = RenderPass({colorAttachments:[{loadOp:'clear',clearValue:[0,0,0,0],storeOp:'store'}]});
	const fetchCamera = createWeakMapGetter(() => new Camera2D());
	return (commandEncoder, item) => {
		const areaRect = item.getBoundingRect();
		const {texture} = item.filter;
		if(!areaRect)return texture.rawTexture = null;
		const {width, height, view0, view1, copyTexture} = fetchView2D(areaRect);
		const drawRect = {width:areaRect.width, height:areaRect.height, x:0, y:0};
		const frameRect = {width, height};

		const camera = fetchCamera(item);
		camera.setLens(frameRect);
		camera.worldTransformInvert.copyFrom(item.worldTransformInvert);
		camera.worldTransformInvert.translate(-areaRect.x, -areaRect.y);
		camera.collectDrawUnits(item, true);

		texture.rawTexture = view1;
		for(let renderPassEncoder of renderPass2D(commandEncoder, view0))camera.draw(renderPassEncoder);
		if(item.filter.justCopyFlag){
			const clipRect = intersect(areaRect, item.filter.clipRect);
			if(!clipRect)return;
			copyTexture(commandEncoder, clipRect.x - areaRect.x, clipRect.y - areaRect.y, clipRect.width, clipRect.height);
			texture.setUV(width, height, {x: 0, y: 0, width:clipRect.width, height:clipRect.height}, false);
			const {adjustX, adjustY} = item.filter;
			texture.xyuvAdd[0] = clipRect.x - (adjustX ? item.filter.clipRect.x : 0);
			texture.xyuvAdd[1] = clipRect.y - (adjustY ? item.filter.clipRect.y : 0);
		}else{
			texture.setUV(width, height, drawRect, false);
			texture.xyuvAdd[0] = areaRect.x;
			texture.xyuvAdd[1] = areaRect.y;
			for(let renderPassEncoder of renderPass2D(commandEncoder, view1)){
				const materialType = item.filter.constructor;
				materialType.setBindGroup(renderPassEncoder, 0, [{binding:0, resource:view0}]);
				materialType.draw([{frameRect, drawRect, material: item.filter}], renderPassEncoder);
			}
		}
	}
}();

module.exports = Camera2D;
