// opengl/Transform3D.js
"use strict";

const Transform = require("./Transform");
const {Matrix3D} = require("./math");

class Transform3D extends Transform{
	constructor(){
		super(Matrix3D);
	}

	isVisible(){
		return true;
	}

	get x(){return this._localMatrix.translation.x;}
	get y(){return this._localMatrix.translation.y;}
	get z(){return this._localMatrix.translation.z;}
	get rotation(){return this._localMatrix.rotation;}
	get translation(){return this._localMatrix.translation;}
	get scale(){return this._localMatrix.scale;}
	
	set x(value){
		this._localMatrix.translation.x = value;
		this.onTransformChanged();
	}
	
	set y(value){
		this._localMatrix.translation.y = value;
		this.onTransformChanged();
	}

	set z(value){
		this._localMatrix.translation.z = value;
		this.onTransformChanged();
	}
	
	set rotation(value){
		this._localMatrix.rotation.copyFrom(value);
		this.onTransformChanged();
	}

	set scale(value){
		this._localMatrix.scale = value;
		this.onTransformChanged();
	}
}

module.exports = Transform3D;
