// opengl/WebGL_Init.js
'use strict';

const WebGL = require('./WebGL');

const Render2D = require("./Render2D");
//const Render3D = require("./Render3D");

const {RenderStateStack, RenderState} = require("./renderers");
const {
	AssetManager,
	TextureManager,
	ProgramManager,
	ProfileManager,
	RenderTargetManager,
	MouseManager,
	KeyBoardManager
} = require("./managers");

function init(canvas, width, height){
	let gl = canvas.getContext("webgl2", {stencil:true});
	console.log(gl.getContextAttributes());

	this.width = width;
	this.height = height;
	this.canvas = canvas;
	this.gl = gl;

	this.mouseMgr = new MouseManager();
	this.keyboardMgr = new KeyBoardManager();
	this.renderTargetMgr = new RenderTargetManager();
	this.assetMgr = new AssetManager();
	this.textureMgr = new TextureManager();
	this.programMgr = new ProgramManager();
	this.profileMgr = new ProfileManager();
	
	this.render2d = new Render2D();
	//this.render3d = new Render3D();
	//this.renderQueue = new RenderQueue();
	this.renderStateStack = new RenderStateStack();
	this.renderState = new RenderState().load();

	gl.viewport(0, 0, width, height);

	this.render2d.viewport(width, height);
}

module.exports = (...args) => init.apply(WebGL, args);
