// opengl/const/KeyCode.js

exports.BACKSPACE = 8
exports.TAB = 9
exports.ENTER = 13
exports.ESC = 27
exports.SPACE = 32
exports.PAGE_UP = 33
exports.PAGE_DOWN = 34
exports.END = 35
exports.HOME = 36

exports.ARROW_LEFT = 37
exports.ARROW_UP = 38
exports.ARROW_RIGHT = 39
exports.ARROW_DOWN = 40

exports.NUM_0 = 48
exports.NUM_1 = 49
exports.NUM_2 = 50
exports.NUM_3 = 51
exports.NUM_4 = 52
exports.NUM_5 = 53
exports.NUM_6 = 54
exports.NUM_7 = 55
exports.NUM_8 = 56
exports.NUM_9 = 57

exports.A = 65
exports.B = 66
exports.C = 67
exports.D = 68
exports.E = 69
exports.F = 70
exports.G = 71
exports.H = 72
exports.I = 73
exports.J = 74
exports.K = 75
exports.L = 76
exports.M = 77
exports.N = 78
exports.O = 79
exports.P = 80
exports.Q = 81
exports.R = 82
exports.S = 83
exports.T = 84
exports.U = 85
exports.V = 86
exports.W = 87
exports.X = 88
exports.Y = 89
exports.Z = 90

exports.F1 = 112
exports.F2 = 113
exports.F3 = 114
exports.F4 = 115
exports.F5 = 116
exports.F6 = 117
exports.F7 = 118
exports.F8 = 119
exports.F9 = 120
exports.F10 = 121
exports.F11 = 122
exports.F12 = 123
