// utils/cache.js

function delayNew(newFn){
	let v;
	return function(){
		if(newFn){
			v = newFn.call(this);
			newFn = null;
		}
		return v;
	}
}

const createGetter = (newFn, keyFn, hit=Object.create(null)) => function(...args){
	const k = keyFn?.(...args) ?? args[0];
	if(k in hit)return hit[k];
	const v = newFn.apply(this, args);
	hit[k] = v;
	return v;
}

const createMapGetter = (newFn, keyFn, hit=new Map()) => function(...args){
	const k = keyFn?.(...args) ?? args[0];
	if(hit.has(k))return hit.get(k);
	const v = newFn.apply(this, args);
	hit.set(k, v);
	return v;
}

function createGetSet(newFn, setFlag=true, value, onSet){
	return [function(){
		if(setFlag){
			setFlag = false;
			value = newFn.call(this);
		}
		return value;
	}, function(){
		const oldFlag = setFlag;
		setFlag = true;
		onSet?.call(this, oldFlag);
	}];
}

function defineProperty(target, key, get, set){
	Object.defineProperty(target, key, {get, set, enumerable: true});
}

const defineMainSubProp = function(){
	const mainPropFn = (subPropList, value) => [() => value, function(newValue){
		if(value === newValue)return;
		value = newValue;
		for(const [getFn, setFn] of subPropList)setFn.call(this);
	}];
	return function(mainProp, subPropMap){
		const keyList = Object.keys(subPropMap);
		return function(target, value){
			const subPropList = keyList.map(k => createGetSet(subPropMap[k], false));
			defineProperty(target, mainProp, ...mainPropFn(subPropList, value));
			for(let i=0, n=keyList.length; i<n; ++i){
				defineProperty(target, keyList[i], subPropList[i][0]);
			}
		}
	}
}();

const defineSubMainProp = function(){
	function subPropFn(setFn, defaultValue){
		const needCheck = typeof defaultValue !== 'function';
		let value = needCheck ? defaultValue : defaultValue();
		return [() => value, function(newValue){
			if(needCheck && value === newValue)return;
			value = newValue;
			setFn.call(this);
		}];
	}
	return (mainProp, mainPropFn, subPropMap) => (target, value, onSet) => {
		const [getFn, setFn] = createGetSet(mainPropFn, false, value, onSet);
		defineProperty(target, mainProp, getFn);
		for(const name in subPropMap){
			defineProperty(target, name, ...subPropFn(setFn, subPropMap[name]));
		}
	}
}();

const createSingleValueWrapper = oldValue => newValue => (newValue === undefined) ? oldValue : (oldValue = newValue);

function createFetch(loadFn){
	const cache = new Map();
	return (key, promiseFlag=true) => {
		if(!cache.has(key)){
			const waitList = [];
			const info = {waitList};
			cache.set(key, info);
			loadFn(key).then(value => {
				info.value = value;
				info.waitList = null;
				waitList.forEach(f => f[0](value));
			}, value => {
				cache.delete(key);
				waitList.forEach(f => f[1](value));
			});
		}
		const {waitList, value} = cache.get(key);
		return promiseFlag ? waitList ? new Promise((...args) => waitList.push(args)) : Promise.resolve(value) : value;
	};
}

module.exports = {
	delayNew,
	createGetter,
	createMapGetter,
	createFetch,
	createSingleValueWrapper,
	createGetSet,
	defineMainSubProp,
	defineSubMainProp,
	createListGetter: (newFn, list=[]) => createGetter(newFn, null, list),
};
