// utils/random.js
"use strict";

function bool(){
	return Math.random() < 0.5;
}

function sign(){
	return bool() ? 1 : -1;
}

function rangeFloat(a, b){
	return a + (b - a) * Math.random();
}

function rangeInt(a, b){
	return Math.floor(rangeFloat(Math.ceil(a), Math.floor(b)));
}

function arrayItem(list, removeFlag=false){
	let index = Math.floor(list.length * Math.random());
	if(removeFlag){
		return list.splice(index, 1)[0];
	}
	return list[index];
}

function arrayIndexes(length, count=1, repeatFlag=true){
	let result = [];
	if(repeatFlag){
		while(result.length < count){
			result.push(Math.floor(length * Math.random()));
		}
	}else{
		let indexList = [];
		for(let i=0; i<length; ++i){
			indexList[i] = i;
		}
		if(count > length){
			count = length;
		}
		while(result.length < count){
			result.push(arrayItem(indexList, true));
		}
	}
	return result;
}

function arrayItems(list, count=1, repeatFlag=true){
	let indexList = arrayIndexes(list.length, count, repeatFlag);
	return indexList.map(index => list[index]);
}

module.exports = {
	bool,
	sign,
	rangeFloat,
	rangeInt,
	arrayItem,
	arrayIndexes,
	arrayItems
};
