// ecs/src/Engine.js
'use strict';

class Engine
{
	constructor(){
		Object.defineProperty(this, 'systemList', {value: new Set()});
		Object.defineProperty(this, 'entityList', {value: new Set()});
		Object.defineProperty(this, 'entityDict', {value: {}});
		Object.defineProperty(this, 'componentDict', {value: new Map()});
	}

	addSystem(system){
		this.systemList.add(system.call(this));
		return this;
	}

	addEntity(entity){
		let {entityDict, entityList, componentDict} = this;
		entityDict[entity.id] = entity;
		entityList.add(entity);
		entity.engine = this;
		componentDict.forEach(entity.addEntityToSet, entity);
		return this;
	}

	removeEntity(entity){
		let {entityDict, entityList, componentDict} = this;
		delete entityDict[entity.id];
		entityList.delete(entity);
		entity.engine = null;
		componentDict.forEach(entity.removeEntityFromSet, entity);
	}

	getEntity(id){
		return this.entityDict[id];
	}

	onComponentAdded(entity, componentType){
		for(let [componentTypes, entitySet] of this.componentDict){
			if(!componentTypes.includes(componentType))continue;
			entity.addEntityToSet(entitySet, componentTypes);
		}
	}

	onComponentRemoved(entity, componentType){
		for(let [componentTypes, entitySet] of this.componentDict){
			if(!componentTypes.includes(componentType))continue;
			entity.removeEntityFromSet(entitySet);
		}
	}

	findEntities(componentTypes){
		let {entityList, componentDict} = this;
		if(!componentTypes || componentTypes.length === 0){
			return entityList;
		}
		if(!componentDict.has(componentTypes)){
			let entitySet = new Set();
			for(let entity of entityList){
				entity.addEntityToSet(entitySet, componentTypes);
			}
			componentDict.set(componentTypes, entitySet);
		}
		return componentDict.get(componentTypes);
	}

	update(){
		for(let system of this.systemList){
			system.call(this);
		}
	}
}

module.exports = Engine;
