// webgpu/materials/BillboardMaterial.js
'use strict';

const Material = require('./Material');
const {classifyTexture} = require('./helper');

const pipelineDesc = {
	layout: ['0', '2_dynamic', '3'],
	vertex: {
		entryPoint: 'main',
		module: 'vertex_billboard',
	},
	fragment: {
		entryPoint: 'main',
		module: 'fragment_billboard',
		targets:[{
			format: 'bgra8unorm',
			blend: {
				color:{srcFactor: 'one', dstFactor: 'one-minus-src', operation: 'add'},
				alpha:{srcFactor: 'src-alpha', dstFactor: 'one-minus-src-alpha', operation: 'add'}
			}
		}]
	},
	depthStencil: {
		format: 'depth24plus-stencil8',
		depthWriteEnabled: false,
		depthCompare: 'less-equal',
	},
	primitive: {
		topology: 'triangle-strip',
		stripIndexFormat: 'uint16',
	}
};

class BillboardMaterial extends Material.Subclass(
	pipelineDesc,
	{transform:1}
){
	static useMeshFlag = 4;
	static *calcDrawIterator(entityList){
		for(const [texture, subEntityList] of classifyTexture(entityList)){
			this.setBindGroup(2, texture.createView());
			yield * super.calcDrawIterator(subEntityList);
		}
	}
	constructor(texture, width, height){
		super();
		this.texture = texture;
	}
}

module.exports = BillboardMaterial;
