// webgpu/materials/ColorMaterial.js
'use strict';

const Vulkan = require('../Vulkan');
const Material = require('./Material');

const pipelineDesc = {
	layout: ['0', '2_dynamic', '2_dynamic', '4'],
	vertex: {
		module: 'vertex_3d',
		buffers: [{
			arrayStride: 3 << 2,
			stepMode: 'vertex',
			attributes: [{shaderLocation:0, format:'float32x3'}]
		}, {
			arrayStride: 3 << 2,
			stepMode: 'vertex',
			attributes: [{shaderLocation:1, format:'float32x3'}]
		}, {
			arrayStride: 2 << 2,
			stepMode: 'vertex',
			attributes: [{shaderLocation:2, format:'float32x2'}]
		}, {
			arrayStride: 1 << 2,
			stepMode: 'vertex',
			attributes: [{shaderLocation:3, format:'uint32'}]
		}]
	},
	fragment: {
		module: 'fragment_3d_color',
		targets:[{
			format: 'bgra8unorm',
			blend: ['transparent']
		}]
	},
	depthStencil: {
		format: 'depth24plus-stencil8',
		depthWriteEnabled: true,
		depthCompare: 'less-equal',
	}
};

class ColorMaterial extends Material.Subclass(
	pipelineDesc,
	{transform:1,bone:2,color:3},
	['position', 'normal', 'uv', 'boneIndex']
){
	static onDraw(instanceCount, mesh){
		for(let i=0, n=mesh.subMeshList.length; i<n; ++i){
			const subMesh = mesh.subMeshList[i];
			this.setVertexBuffer(subMesh);
			Vulkan.renderPassEncoder.draw(subMesh.vertexCount, instanceCount);
		}
	}
	
	constructor(color){
		super();
		this.color = color;
	}
}

module.exports = ColorMaterial;
