// webgpu/materials/MirrorMaterial.js
'use strict';

const Vulkan = require('../Vulkan');
const TextureMaterial = require('./TextureMaterial');
const groupDraw = require('../helper/groupDraw');
const Camera3D = require('../cameras/Camera3D');
const Quaternion = require('../math/Quaternion');
const Lens = require('../cameras/Lens');
const Pipeline = require('./Pipeline');
const {worldMatrixList} = require('./helper/DefaultGroupEntries');

const pipelineQuad = new Pipeline({
	groups: [
		[{name:'vpMatrix', type:'MVP', visibility: GPUShaderStage.VERTEX, buffer:{}}],
		[
			{name:'baseTexture', texture:{}},
			{name:'g_sampler', sampler:{}, default:{magFilter:'linear', minFilter:'linear', addressModeU:'repeat', addressModeV:'repeat', addressModeW:'repeat'}},
			{name:'uvScale', type:'vec2<f32>', buffer:{}}
		]
	],
	vertex: {
		code: `
			#include<transform>
			struct Output {
				@builtin(position) position: vec4<f32>,
				@location(0) uv: vec2<f32>,
			}

			@vertex Output {
				var vertex: vec2<f32>;
				vertex.x = f32(vertexIndex >> 1);
				vertex.y = f32(vertexIndex & 1);

				var position = vertex * 2 - 1;
				position.y = -position.y;

				var output: Output;
				output.position = vec4<f32>(position, 0, 1);
				output.uv = vec2(1-vertex.x, vertex.y) * uvScale;
				return output;
			}
		`
	},
	fragment: {targets:[{}], code: `@fragment{return textureSample(baseTexture, g_sampler, uv);}`},
	depthStencil: {
		depthWriteEnabled: false,
		depthCompare: 'always',//always, never, equal, not-equal, less, less-equal, greater, greater-equal
		stencilFront: {compare: 'not-equal'},
		stencilBack: {compare: 'not-equal'},
	},
	primitive: {topology: 'triangle-strip'}
});

const pipeline = new Pipeline({
	groups: [
		[{name:'vpMatrix', type:'MVP', buffer:{}}],
		[worldMatrixList()]
	],
	vertex:{code:`
		#include<transform>
		@vertex{
			let mat = worldMatrixList[instanceIndex];
			return positionWorldToScreen(vec4<f32>(position, 1) * mat, vpMatrix);
		}`,
		buffers: [{name: 'position', arrayStride: 4 << 2, attributes: [{format:'float32x3'}]}],
	},
	fragment: {targets:[{writeMask: 0, blend: 'one add zero'}], code: `@fragment fn main(){}`},
	depthStencil: {
		stencilFront: {passOp: 'increment-clamp'},
		stencilBack: {passOp: 'increment-clamp'},
	},
	primitive: {topology: 'triangle-strip'}
});

class MirrorMaterial extends TextureMaterial.Subclass(pipeline){
	static maxCountPerDraw = 1;
	static onDraw(drawUnitList, renderPassEncoder){
		const {position} = drawUnitList[0].entity.mesh.subMeshList[0].attributes;
		pipeline.setResources(renderPassEncoder, {
			'worldMatrixList': {buffer:drawUnitList},
			position
		});
		renderPassEncoder.draw(4);

		const material = drawUnitList[0].material;
		uv_scale[0] = Vulkan.canvas.width / material.texture.width;
		uv_scale[1] = Vulkan.canvas.height / material.texture.height;

		pipelineQuad.setPipeline(renderPassEncoder);
		pipelineQuad.setResources(renderPassEncoder, {
			'baseTexture': material.texture,
			'uvScale': {buffer:uv_scale}
		});
		renderPassEncoder.draw(4);
	}

	constructor(texture){
		super();
		let camera = new Camera3D(new Lens.Perspective(Math.PI / 6, 900 / 1200, -1500, 5000, -2000));
		camera.rotation = new Quaternion().fromEulerAngles(120 * Math.PI / 180, 0, -135 * Math.PI / 180);
		this.camera = camera;
	}

	get pick(){}
	get depth(){}
}

const uv_scale = new Float32Array(2);

module.exports = MirrorMaterial;
