// webgpu/materials/TerrainMaterial.js
'use strict';

const Vulkan = require('../Vulkan');
const Material = require('./Material');
const {attributes, blend} = require('./helper');

const pipelineDesc = {
	layout: ['0', '5'],
	vertex: {
		module: 'vertex_terrain',
		buffers: [{
			arrayStride: 12,
			attributes: attributes('uint8x2', [4, 'unorm8x2'], [8, 'unorm8x4'])
		}]
	},
	fragment: {
		module: 'fragment_terrain',
		targets:[{
			format: 'bgra8unorm',
			blend: blend('transparent')
		}]
	},
	depthStencil: {
		format: 'depth24plus-stencil8',
		depthWriteEnabled: true,
		depthCompare: 'less-equal',
	}
};

class TerrainMaterial extends Material.Subclass(
	pipelineDesc,
){
	static onDraw(entityList){
		const {renderPassEncoder, currentMesh} = Vulkan;
		if(!this.bindGroup){
			this.bindGroup = Vulkan.createBindGroup(
				Vulkan.layout_5, [
				entityList[0].material.texture
			]);
		}
		renderPassEncoder.setBindGroup(1, this.bindGroup);
		renderPassEncoder.drawIndexed(currentMesh.indexCount, 1, 0, 0, 0);
	}
	constructor(texture){
		super();
		this.texture = texture;
	}
}

module.exports = TerrainMaterial;
