// webgpu/materials/TerrainMaterial.js
'use strict';

const Material = require('./Material');
const {classifyTexture} = require('./helper');

const pipelineDesc = {
	layout: ['0', '5', '2_dynamic'],
	vertex: {
		entryPoint: 'main',
		module: 'vertex_terrain',
		buffers: [{
			name: 'layer0',
			arrayStride: 12,
			stepMode: 'vertex',
			attributes: [
				{shaderLocation: 0, offset:0, format:'uint8x2'},
				{shaderLocation: 1, offset:4, format:'unorm8x2'},
				{shaderLocation: 2, offset:8, format:'unorm8x4'}
			]
		}]
	},
	fragment: {
		entryPoint: 'main',
		module: 'fragment_terrain',
		targets:[{
			format: 'bgra8unorm',
			blend: {
				color:{srcFactor: 'src-alpha', dstFactor: 'one-minus-src-alpha', operation: 'add'},
				alpha:{srcFactor: 'src-alpha', dstFactor: 'one-minus-src-alpha', operation: 'add'}
			}
		}]
	},
	depthStencil: {
		format: 'depth24plus-stencil8',
		depthWriteEnabled: true,
		depthCompare: 'less-equal',
	}
};

class TerrainMaterial extends Material.Subclass(
	pipelineDesc,
	{transform:2}
){
	static *calcDrawIterator(entityList, mesh){
		for(const [texture, subEntityList] of classifyTexture(entityList)){
			this.setBindGroup(1, texture);
			yield * super.calcDrawIterator(subEntityList);
		}
	}
	constructor(texture){
		super();
		this.texture = texture;
	}
}

module.exports = TerrainMaterial;
