// webgpu/materials/TextureMaterial.js
'use strict';

const Vulkan = require('../Vulkan');
const Material = require('./Material');

const {delayNew} = require('utils/cache');
const transformSlot = require('./slots/transformSlot');
const boneSlot = require('./slots/boneSlot');

class TextureMaterial extends Material
{
	static get pipeline(){
		return Vulkan.pipelineDict.pipeline;
	}
	static bindGroupSlots = delayNew(() => [transformSlot(), boneSlot()]);
	static onDraw(entityList, infoList, boneInfoList){
		const {renderPassEncoder, device, currentMesh} = Vulkan;
		for(const subMesh of currentMesh.subMeshList){
			if(!subMesh.texture)continue;
			let texture = Vulkan.fetchTexture(subMesh.texture);
			if(!texture)continue;
			renderPassEncoder.setBindGroup(3, device.createBindGroup({
				layout: Vulkan.bindGroupLayoutDict.bindGroupLayout3,
				entries: [{
					binding: 0,
					resource: texture.createView()
				}]
			}));
			const {vertexCount, vertexOffset} = subMesh;
			renderPassEncoder.setBindGroup(1, transformSlot().getBindGroup(infoList[0][0]), [infoList[0][1] << 8]);
			renderPassEncoder.setBindGroup(2, boneSlot().getBindGroup(boneInfoList[0][0]), [boneInfoList[0][1] << 12]);
			renderPassEncoder.draw(vertexCount, boneInfoList[0][2], vertexOffset, 0);
			continue;
			for(const [index, offset, count] of infoList){
				renderPassEncoder.setBindGroup(1, transformSlot().getBindGroup(index), [0]);
				renderPassEncoder.setBindGroup(2, boneSlot().getBindGroup(index), [0]);
				renderPassEncoder.draw(vertexCount, count, vertexOffset, offset);//here is a bug, driver ignored offset arg!
			}
		}
	}
	constructor(){
		super();
	}
}

module.exports = TextureMaterial;
