// webgpu/materials/WingTextureMaterial.js
'use strict';

const Vulkan = require('../Vulkan');
const TextureMaterial = require('./TextureMaterial');

const pipelineDesc = {
	layout: ['0', '2_dynamic', '2_dynamic', '3'],
	vertex: {
		entryPoint: 'main',
		module: 'vertex_3d',
		buffers: [{
			name: 'position',
			arrayStride: 3 << 2,
			stepMode: 'vertex',
			attributes: [{shaderLocation:0, offset: 0, format:'float32x3'}]
		}, {
			name: 'normal',
			arrayStride: 3 << 2,
			stepMode: 'vertex',
			attributes: [{shaderLocation:1, offset: 0, format:'float32x3'}]
		}, {
			name: 'uv',
			arrayStride: 2 << 2,
			stepMode: 'vertex',
			attributes: [{shaderLocation:2, offset: 0, format:'float32x2'}]
		}, {
			name: 'boneIndex',
			arrayStride: 1 << 2,
			stepMode: 'vertex',
			attributes: [{shaderLocation:3, offset: 0, format:'uint32'}]
		}]
	},
	fragment: {
		entryPoint: 'main',
		module: 'fragment_3d',
		targets:[{
			format: 'bgra8unorm',
			blend: {
				color:{srcFactor: 'one', dstFactor: 'one-minus-src', operation: 'add'},
				alpha:{srcFactor: 'src-alpha', dstFactor: 'one-minus-src-alpha', operation: 'add'}
			}
		}]
	},
	depthStencil: {
		format: 'depth24plus-stencil8',
		depthWriteEnabled: false,
		depthCompare: 'less-equal',
	}
};

class WingTextureMaterial extends TextureMaterial.Subclass(
	pipelineDesc,
	{transform:1, bone:2}
){
}

module.exports = WingTextureMaterial;
