// webgpu/materials/WingTextureMaterial.js
'use strict';

const Vulkan = require('../Vulkan');
const TextureMaterial = require('./TextureMaterial');

const pipelineDesc = {
	layout: ['0', '2_dynamic', '2_dynamic', '3'],
	vertex: {
		module: 'vertex_3d',
		buffers: [{
			name: 'position',
			arrayStride: 3 << 2,
			stepMode: 'vertex',
			attributes: [{shaderLocation:0, format:'float32x3'}]
		}, {
			name: 'normal',
			arrayStride: 3 << 2,
			stepMode: 'vertex',
			attributes: [{shaderLocation:1, format:'float32x3'}]
		}, {
			name: 'uv',
			arrayStride: 2 << 2,
			stepMode: 'vertex',
			attributes: [{shaderLocation:2, format:'float32x2'}]
		}, {
			name: 'boneIndex',
			arrayStride: 1 << 2,
			stepMode: 'vertex',
			attributes: [{shaderLocation:3, format:'uint32'}]
		}]
	},
	fragment: {
		module: 'fragment_3d',
		targets:[{
			format: 'bgra8unorm',
			blend: [['one', 'one-minus-src'], 'transparent']
		}]
	},
	depthStencil: {
		format: 'depth24plus-stencil8',
		depthWriteEnabled: false,
		depthCompare: 'less-equal',
	}
};

class WingTextureMaterial extends TextureMaterial.Subclass(
	pipelineDesc,
	{transform:1, bone:2}
){
}

module.exports = WingTextureMaterial;
