// framework/mvc.js
"use strict";

function handleMsg(dict, msg){
	for(let target of dict.values()){
		if(msg.isProcessCanceled())break;
		safeCall(target, msg.name, msg);
	}
}

function safeCall(target, name, ...args){
	if(target[name] instanceof Function)
		return target[name](...args);
}

function isBaseClass(v, base){
	if(!v.prototype)
		return v instanceof base;
	for(;v;v=Object.getPrototypeOf(v))
		if(v === base)return true;
}

function filterModel(v){
	if(!Array.isArray(v))
		return isBaseClass(v, Model);
	if(v.length == 2)
		return isBaseClass(v[1], Model)
	if(v.length == 1)
		return isBaseClass(v[0], Model)
}
function filterService(v){
	if(!Array.isArray(v))
		return isBaseClass(v, Service);
	if(v.length == 3)
		return true;
	if(v.length == 1)
		return isBaseClass(v[0], Service);
	if(v.length == 2){
		return (typeof v[1] == "boolean") || isBaseClass(v[1], Service);
	}
	throw new Error(v);
}
const filterController = v => isBaseClass(v, Controller);
const filterView = v => isBaseClass(v, View);

function collectMSVC(target, filter, fnName){
	let msvc = safeCall(target, "collectMSVC");
	if(Array.isArray(msvc) && msvc.length > 0)
		return msvc.filter(filter);
	return safeCall(target, fnName);
}

const Injector = require("./ioc");

class Application
{
	constructor(){
		Object.defineProperties(this, {
			injector: {value: new Injector()},
			moduleDict: {value: new Map()}
		});
		this.injector.mapValue(Application, this, null, null);
		this.injector.mapValue(Injector, this.injector, null, null);
	}

	regModule(module){
		console.assert(!this.moduleDict.has(module.constructor));
		this.moduleDict.set(module.constructor, module);
		module.injector.parent = this.injector;
	}

	startup(){
		let moduleList = Array.from(this.moduleDict.values());
		for(let module of moduleList) module.initAllModels();
		for(let module of moduleList) module.initAllServices();
		for(let module of moduleList) module.initAllViews();
		for(let module of moduleList) module.initAllControllers();
		for(let module of moduleList) safeCall(module, "onStartup");
	}
}

class Module
{
	constructor(){
		Object.defineProperties(this, {
			injector: {value: new Injector()},
			controllerDict: {value: new Map()},
			viewSet: {value: new Set()}
		});
		this.injector.mapValue(Module, this, null, null);
		this.injector.mapValue(Injector, this.injector, null, null);
	}

	notify(msgName, msgData=null){
		let msg = new Msg(msgName, msgData);
		handleMsg(this.controllerDict, msg);
		handleMsg(this.viewSet, msg);
		return !msg.isDefaultPrevented();
	}

	regService(serviceInterface, serviceClass, asLocal=false){
		let injector = asLocal ? this.injector : this.injector.parent;
		injector.mapSingleton(serviceInterface, serviceClass, null, this.injector);
	}

	regModel(model, modelType=null){
		this.injector.mapValue(modelType || model.constructor, model);
	}

	delModel(modelType){
		this.injector.ummap(modelType);
	}

	regView(view){
		console.assert(!this.viewSet.has(view));
		this.viewSet.add(view);
		this.injector.injectInto(view);
		safeCall(view, "onReg");
	}

	delView(view){
		safeCall(view, "onDel");
		this.viewSet.delete(view);
	}

	regController(controllerType){
		console.assert(!this.controllerDict.has(controllerType));
		let controller = new controllerType();
		this.controllerDict.set(controllerType, controller);
		this.injector.injectInto(controller);
		safeCall(controller, "onReg");
	}

	delController(controllerType){
		let controller = this.controllerDict.get(controllerType);
		safeCall(controller, "onDel");
		this.controllerDict.delete(controllerType);
	}

	initAllModels(){
		let set = collectMSVC(this, filterModel, "collectModels");
		if(!set)return;
		for(let v of set){
			if(Array.isArray(v)){
				if(v.length == 2){
					this.regModel(v[1], v[0]);
					continue;
				}else if(v.length == 1){
					v = v[0];
				}else{
					throw new Error(v);
				}
			}
			if(v.prototype){
				this.regModel(new v());
			}else{
				this.regModel(v);
			}
		}
	}
	initAllServices(){
		let set = collectMSVC(this, filterService, "collectServices");
		if(!set)return;
		for(let v of set){
			if(!Array.isArray(v)){
				this.regService(v, v);
			}else if(v.length == 3){
				this.regService(...v);
			}else if(v.length == 2){
				if(typeof v[1] == "boolean"){
					this.regService(v[0], ...v);
				}else{
					this.regService(...v);
				}
			}else if(v.length == 1){
				this.regService(v[0], v[0]);
			}else{
				throw new Error(v);
			}
		}
	}
	initAllViews(){
		let set = collectMSVC(this, filterView, "collectViews");
		if(!set)return;
		for(let v of set){
			this.regView(v);
		}
	}
	initAllControllers(){
		let set = collectMSVC(this, filterController, "collectControllers");
		if(!set)return;
		for(let v of set){
			this.regController(v);
		}
	}
}

class Notifier
{
	notify(msgName, msgData=null){
		return this.module.notify(msgName, msgData);
	}
}
Notifier[Injector.Tag] = {module: Module};

class Msg
{
	constructor(name, data){
		this.name = name;
		this.data = data;
		this.defaultPreventedFlag = false;
		this.processCanceledFlag = false;
	}

	cancelProcess(){
		this.processCanceledFlag = true;
	}

	isProcessCanceled(){
		return this.processCanceledFlag;
	}

	preventDefault(){
		this.defaultPreventedFlag = true;
	}

	isDefaultPrevented(){
		return this.defaultPreventedFlag;
	}
}

function createMsgNames(...nameList){
	return Object.defineProperties(
		Object.create(null),
		nameList.reduce(
			(dict, name) => (dict[name] = {value:Symbol(name),enumerable:true}, dict),
			Object.create(null)
		)
	);
}

class Model extends Notifier {}
class Service extends Notifier {}
class Controller extends Notifier {}
class View extends Notifier {}

module.exports = {
	Application,
	Module,
	Model,
	Service,
	View,
	Controller,
	Msg,
	createMsgNames
};
