// opengl/renderers/RenderStateStack.js
'use strict';

const WebGL = require('../WebGL');

class RenderStateStack
{
	constructor(){
		this.stateStack = new StateStack();
	}

	save(...keys){
		let {gl} = WebGL;
		let {stateStack} = this;
		for(let key of keys){
			stateStack.push(key, gl.getParameter(key));
		}
	}

	load(...keys){
		let {gl} = WebGL;
		let {stateStack} = this;
		for(let key of keys){
			const value = stateStack.pop(key);
			if(gl.getParameter(key) == value)return;
			switch(key){
			case gl.DRAW_FRAMEBUFFER_BINDING:
				gl.bindFramebuffer(gl.DRAW_FRAMEBUFFER, value);
				break;
			case gl.READ_FRAMEBUFFER_BINDING:
				gl.bindFramebuffer(gl.READ_FRAMEBUFFER, value);
				break;
			case gl.FRAMEBUFFER_BINDING:
				gl.bindFramebuffer(gl.FRAMEBUFFER, value);
				break;
			case gl.CURRENT_PROGRAM:
				gl.useProgram(value);
				break;
			case gl.VIEWPORT:
				gl.viewport(...value);
				break;
			}
		}
	}
}

class StateStack
{
	constructor(){
		this.stateDict = new Map();
	}

	push(key, value){
		let {stateDict} = this;
		if(stateDict.has(key)){
			stateDict.get(key).push(value);
		}else{
			stateDict.set(key, [value]);
		}
	}

	pop(key){
		return this.stateDict.get(key).pop();
	}
}

module.exports = RenderStateStack;
