// opengl/renderers/RenderStateEnum.js
"use strict";

const RenderState = require("./RenderState");

RenderState.Texture = Object.assign(new RenderState(), {
	depthTestEnabled: true,
	depthMask: true,
	depthFunc: WebGLRenderingContext.LEQUAL,
	blendEnabled: true,
	blendFuncSource: WebGLRenderingContext.SRC_ALPHA,
	blendFuncDest: WebGLRenderingContext.ONE_MINUS_SRC_ALPHA
});

RenderState.TextureHitTest = Object.assign(new RenderState(), {
	depthTestEnabled: true,
	depthMask: true,
	depthFunc: WebGLRenderingContext.LEQUAL,
	blendEnabled: false
});

RenderState.Shadow = Object.assign(new RenderState(), {
	depthTestEnabled: false,
	blendEnabled: true,
	blendFuncSource: WebGLRenderingContext.DST_COLOR,
	blendFuncDest: WebGLRenderingContext.ZERO
});

RenderState.Normal2D = Object.assign(new RenderState(), {
	depthTestEnabled: false,
	blendEnabled: true,
	blendFuncSource: WebGLRenderingContext.SRC_ALPHA,
	blendFuncDest: WebGLRenderingContext.ONE_MINUS_SRC_ALPHA
});

RenderState.Pick2D = Object.assign(new RenderState(), {
	depthTestEnabled: false,
	blendEnabled: false
});

module.exports = RenderState;
