// vertexarrays/MeshVertexArray.js
"use strict";

const VertexArray = require("./VertexArray");

class MeshVertexArray extends VertexArray{
	constructor(subMesh, formatList){
		super();
		Object.assign(this, subMesh);
		this.formatList = formatList;
		this.indexCount = this.indexData.byteLength >> 1;
		this.indexMode = this.indexCount > 0;
	}

	onCreate(gl){
		this.vertexBuffer = gl.createBuffer();
		gl.bindBuffer(gl.ARRAY_BUFFER, this.vertexBuffer);
		gl.bufferData(gl.ARRAY_BUFFER, this.vertexData, gl.STATIC_DRAW);

		gl.vertexAttribPointer(0, 3, gl.FLOAT, false, this.byteSizePerVertex, 0);
		gl.vertexAttribPointer(1, 2, gl.FLOAT, false, this.byteSizePerVertex, 24);
		gl.vertexAttribIPointer(2, 1, gl.SHORT, this.byteSizePerVertex, 32);
		gl.vertexAttrib4f(3, 1, 0, 0, 0);

		gl.enableVertexAttribArray(0);
		gl.enableVertexAttribArray(1);
		gl.enableVertexAttribArray(2);
		gl.disableVertexAttribArray(3);

		if(this.indexMode){
			this.indexBuffer = gl.createBuffer();
			gl.bindBuffer(gl.ELEMENT_ARRAY_BUFFER, this.indexBuffer);
			gl.bufferData(gl.ELEMENT_ARRAY_BUFFER, this.indexData, gl.STATIC_DRAW);
		}
	}

	onDispose(gl){
		gl.deleteBuffer(this.vertexBuffer);
		this.vertexBuffer = null;
		if(this.indexMode){
			gl.deleteBuffer(this.indexBuffer);
			this.indexBuffer = null;
		}
	}

	draw(gl){
		if(this.indexMode){
			gl.drawElements(gl.TRIANGLES, this.indexCount, gl.UNSIGNED_SHORT, 0);
		}else{
			gl.drawArrays(gl.TRIANGLES, 0, this.vertexCount);
		}
	}
}

module.exports = MeshVertexArray;
