// bluetooth/generators/ArduinoOfflineCode.js
'use strict';

const db = {board:'elf'}
const board_elf_mini = 'mini';
const board_elf_2560 = 'elf 2560';
const pinyin = require('utils/text/pinyin');

const identity = v => v;
//*
function hexToRgb(hex){
	var result = /^#?([a-f\d]{2})([a-f\d]{2})([a-f\d]{2})$/i.exec(hex);
	return result ? {
		R: parseInt(result[1], 16),
		G: parseInt(result[2], 16),
		B: parseInt(result[3], 16)
	} : {R:0,G:0,B:0};
}

function castMatrix(matrix, w, h){
	let bytes = [];
	for(let j=0; j<w; ++j){
		bytes[j] = 0;
		for(let i=0; i<h; ++i){
			if(matrix.charAt(i*w+j) === '1'){
				bytes[j] |= 1 << i;
			}
		}
	}
	return bytes.map(v => '0x' + v.toString(16)).join(', ');
}

const INT_2_BOOL = ['false','true'];
const INT_2_HIGH_LOW = ['LOW', 'HIGH'];
const IR_CODE_2_NAME = {
	0x45:'IR_CONTROLLER_A',
	0x46:'IR_CONTROLLER_B',
	0x47:'IR_CONTROLLER_C',
	0x44:'IR_CONTROLLER_D',
	0x40:'IR_CONTROLLER_UP',
	0x43:'IR_CONTROLLER_E',
	0x07:'IR_CONTROLLER_LEFT',
	0x15:'IR_CONTROLLER_OK',
	0x09:'IR_CONTROLLER_RIGHT',
	0x19:'IR_CONTROLLER_DOWN',
	0x0D:'IR_CONTROLLER_F',
	0x16:'IR_CONTROLLER_0',
	0x0C:'IR_CONTROLLER_1',
	0x18:'IR_CONTROLLER_2',
	0x5E:'IR_CONTROLLER_3',
	0x08:'IR_CONTROLLER_4',
	0x1C:'IR_CONTROLLER_5',
	0x5A:'IR_CONTROLLER_6',
	0x42:'IR_CONTROLLER_7',
	0x52:'IR_CONTROLLER_8',
	0x4A:'IR_CONTROLLER_9'
};

//*/
module.exports = function(Arduino){
	const {ORDER_ATOMIC, ORDER_HIGH, ORDER_UNARY, ORDER_ADDITIVE, ORDER_RELATIONAL, ORDER_NONE, END} = Arduino;
	/*function option_handler(block){
		let value = block.inputList[0].fieldRow[0].text_;
		return [value, Arduino.ORDER_ATOMIC];
	}*/

	function castToStr(block, key, mapFn=identity){
		let valueBlock = block.getInputTargetBlock(key);
		return valueBlock.type == 'text' ? JSON.stringify(mapFn(valueBlock.getFieldValue('TEXT'))) : 'String(' + Arduino.valueToCode(block, key, ORDER_NONE) + ').c_str()';
	}

	const register_option = name => Arduino[name] = block => [block.getFieldValue(name.toUpperCase()), Arduino.ORDER_ATOMIC];

	register_option('speed');
	register_option('angle');
	register_option('rgb');
	register_option('note');
	register_option('beat');
	register_option('matrix@21*7');
	register_option('matrix@14*5');
	register_option('speech-recognition-index');
	register_option('vibration-speed');
	register_option('rj11-rgb-index');
	register_option('rgb-ring-index');
	register_option('elf-core-rgb-index');
	//register_option('colour_picker', 'COLOUR');
	//register_option('math_integer', 'NUM');
	Arduino["matrix@14*5"] =
	Arduino["matrix@21*7"] = block => [block.getFieldValue('MATRIX'), Arduino.ORDER_ATOMIC];

	Arduino.board_port =
	Arduino.led_index =
	Arduino.buzzer_note =
	Arduino.buzzer_beat =
	Arduino.sensor_port =
	Arduino.servo_port =
	Arduino.motor_speed =
	Arduino.dc_motor_index = block => [block.getFieldValue('VALUE'), Arduino.ORDER_ATOMIC];
	Arduino.logic_boolean = block => [block.getFieldValue('BOOL') == 'TRUE' ? 1 : 0, Arduino.ORDER_ATOMIC];
	Arduino.ir_code = block => [IR_CODE_2_NAME[block.getFieldValue('VALUE')], Arduino.ORDER_ATOMIC];

	Arduino.weeemake_generate_offline_code = block => '';
	Arduino.when_start = block => '';
	Arduino.weeemake_set_var_type = (list => function(block){
		let raw_name = block.getField('VARIABLE').text_;
		let name = this.adjustVarName(raw_name);
		const VAR_TYPE = block.getFieldValue('VAR_TYPE');
		this.varType[name] = list[VAR_TYPE];
		return '';
	})(['double', 'long']);
	Arduino.weeemake_weeebot_board_rgb = function(block){
		const COLOR = Arduino.valueToCode(block, 'COLOR', ORDER_NONE);

		let key = `rgb_led_board`;

		Arduino.addVarDef(`WeRGBLed ${key}(OnBoard_RGB)`, key);

		let result = [
			`${key}.setColor(1, ${COLOR})`,
			`${key}.show()`
		];

		return result.map(v => Arduino.tab() + v + END).join('');
	}

	Arduino.weeemake_weeebot_board_rgb3 = function(block){
		const R = Arduino.valueToCode(block, 'R', ORDER_NONE);
		const G = Arduino.valueToCode(block, 'G', ORDER_NONE);
		const B = Arduino.valueToCode(block, 'B', ORDER_NONE);

		let key = `rgb_led_board`;

		Arduino.addVarDef(`WeRGBLed ${key}(OnBoard_RGB)`, key);

		let result = [
			`${key}.setColor(1, ${R}, ${G}, ${B})`,
			`${key}.show()`
		];

		return result.map(v => Arduino.tab() + v + END).join('');
	}

	Arduino.weeemake_weeebot_board_button = function(block){
		Arduino.addSetupCode('pinMode(OnBoard_Button, INPUT)', 'OnBoard_Button');
		return ['!digitalRead(OnBoard_Button)', ORDER_UNARY];
	}

	const genRGB = (key, port) => function(block){
		const PIXEL = Arduino.valueToCode(block, 'PIXEL', ORDER_NONE);
		const COLOR = Arduino.valueToCode(block, 'COLOR', ORDER_NONE);

		Arduino.addVarDef(`WeRGBLed ${key}(${port})`, key);

		let result = [
			`${key}.setColor(${PIXEL}, ${COLOR})`,
			`${key}.show()`
		];

		return result.map(v => Arduino.tab() + v + END).join('');
	};

	const genRGB3 = (key, port) => function(block){
		const PIXEL = Arduino.valueToCode(block, 'PIXEL', ORDER_NONE);
		const R = Arduino.valueToCode(block, 'R', ORDER_NONE);
		const G = Arduino.valueToCode(block, 'G', ORDER_NONE);
		const B = Arduino.valueToCode(block, 'B', ORDER_NONE);

		Arduino.addVarDef(`WeRGBLed ${key}(${port})`, key);

		let result = [
			`${key}.setColor(${PIXEL}, ${R}, ${G}, ${B})`,
			`${key}.show()`
		];

		return result.map(v => Arduino.tab() + v + END).join('');
	};

	Arduino.weeemake_weeebot_pin_ring_rgb = genRGB('rgb_led_ring', 'PORT_0');
	Arduino.weeemake_weeebot_pin_ring_rgb3 = genRGB3('rgb_led_ring', 'PORT_0');
	Arduino.weeemake_elf_core_board_rgb = genRGB('rgb_led_board', 'OnBoard_RGB');
	Arduino.weeemake_elf_core_board_rgb3 = genRGB3('rgb_led_board', 'OnBoard_RGB');

	Arduino.weeemake_soil =
	Arduino.weeemake_weeebot_pin_sound =
	Arduino.weeemake_weeebot_pin_light = function(block){
		//const BOARD_PORT = Arduino.valueToCode(block, 'BOARD_PORT', ORDER_NONE);
		const BOARD_PORT = Arduino.valueToCode(block, 'BOARD_PORT', ORDER_NONE);

		Arduino.addSetupCode(`pinMode(${BOARD_PORT}, INPUT)`, BOARD_PORT);
		return [`analogRead(${BOARD_PORT})`, ORDER_HIGH];
	}

	Arduino.weeemake_weeebot_pin_ir = function(block){
		const BOARD_PORT = Arduino.valueToCode(block, 'BOARD_PORT', ORDER_NONE);
		//let IR_CODE = block.getFieldValue('IR_CODE');
		let IR_CODE = Arduino.valueToCode(block, 'IR_CODE', ORDER_NONE)

		Arduino.addVarDef(`WeInfraredReceiver ir(${BOARD_PORT})`, 'ir');
		Arduino.addSetupCode('ir.begin()', 'ir');
		Arduino.addLoopCode('ir.loop()', 'ir');
		Arduino.irFlag = true;

		return [`ir.isKeyPressed(${IR_CODE})`, ORDER_HIGH];
	}

	Arduino.weeemake_weeebotMini_board_light = function(block){
		Arduino.addSetupCode(`pinMode(OnBoard_Light, INPUT)`, 'OnBoard_Light');
		return [`analogRead(OnBoard_Light)`, ORDER_HIGH];
	}

	Arduino.weeemake_weeebotMini_board_sound = function(block){
		Arduino.addSetupCode(`pinMode(OnBoard_Sound, INPUT)`, 'OnBoard_Sound');
		return [`analogRead(OnBoard_Sound)`, ORDER_HIGH];
	}

	Arduino.weeemake_weeebotMini_board_ir = function(block){
		let IR_CODE = Arduino.valueToCode(block, 'IR_CODE', ORDER_NONE)

		Arduino.addVarDef(`WeInfraredReceiver ir(OnBoard_IR)`, 'ir');
		Arduino.addSetupCode('ir.begin()', 'ir');
		Arduino.addLoopCode('ir.loop()', 'ir');
		Arduino.irFlag = true;

		return [`ir.isKeyPressed(${IR_CODE})`, ORDER_HIGH];
	}

	Arduino.weeemake_weeebotMini_board_back_led = (dict => function(block){
		let BACK_LED_PORT = block.getFieldValue('BACK_LED_PORT');
		const ON_OFF = Arduino.valueToCode(block, 'ON_OFF', ORDER_NONE);

		BACK_LED_PORT = dict[BACK_LED_PORT];

		Arduino.addSetupCode(`pinMode(${BACK_LED_PORT}, OUTPUT)`, BACK_LED_PORT);
		return Arduino.tab() + `digitalWrite(${BACK_LED_PORT}, ${ON_OFF})` + END;
	})({'4':'MINI_LEFT_YELLOW','3':'MINI_LEFT_RED','14':'MINI_RIGHT_RED','13':'MINI_RIGHT_YELLOW'});

	Arduino.weeemake_single_led = function(block){
		const SENSOR_PORT = Arduino.valueToCode(block, 'SENSOR_PORT', ORDER_NONE);
		const ON_OFF = Arduino.valueToCode(block, 'ON_OFF', ORDER_NONE);

		Arduino.addSetupCode(`pinMode(${SENSOR_PORT}, OUTPUT)`, SENSOR_PORT);
		return Arduino.tab() + `digitalWrite(${SENSOR_PORT}, ${ON_OFF})` + END;
	}

	Arduino.weeemake_single_line_follower = function(block){
		const BOARD_PORT = Arduino.valueToCode(block, 'BOARD_PORT', ORDER_NONE);
		Arduino.addSetupCode(`pinMode(${BOARD_PORT}, INPUT)`, BOARD_PORT);
		return [`1023 - analogRead(${BOARD_PORT})`, ORDER_ADDITIVE];
	}

	Arduino.elf_328p_mpin = 
	Arduino.elf_mini_mpin = 
	Arduino.elf_uno_mpin = 
	function(block){
		return [block.getFieldValue('value'), ORDER_HIGH]
	}

	Arduino.weeemake_dc_motor = function(block){
		const pin = Arduino.valueToCode(block, 'pin', ORDER_NONE).split(',');
		const speed = Arduino.valueToCode(block, 'speed', ORDER_NONE);
		Arduino.addSetupCode(`pinMode(${pin[0]}, OUTPUT)`);
		Arduino.addSetupCode(`pinMode(${pin[1]}, OUTPUT)`);
		if(block.getInputTargetBlock('pin').type == 'elf_328p_mpin'){
			return [`{`,
			`\tint16_t speed = ${speed};`,
			`\tanalogWrite(${pin[0]}, speed >= 0 ? speed : -speed);`,
			`\tdigitalWrite(${pin[1]}, speed >= 0 ? 0 : 1);`,
			`}`].map(v => Arduino.tab() + v + '\n').join('');
		}
		return [`{`,
		`\tint16_t speed = ${speed};`,
		`\tanalogWrite(${pin[0]}, speed >= 0 ? 0 : -speed);`,
		`\tanalogWrite(${pin[1]}, speed >= 0 ? speed : 0);`,
		`}`].map(v => Arduino.tab() + v + '\n').join('');
	}

	Arduino.weeemake_robot_move = function(block){
		//1,2,3,4 => F,B,L,R
		const MOVE_DIRECTION = block.getFieldValue('MOVE_DIRECTION');
		let SPEED = Arduino.valueToCode(block, 'SPEED', ORDER_NONE);

		let needSpeedVar = !/^\w+$/.test(SPEED);
		let result = [];
		Arduino.addVarDef('WeDCMotor dc_1(M1)', 'dc_1');
		Arduino.addVarDef('WeDCMotor dc_2(M2)', 'dc_2');
		let l_sign = (MOVE_DIRECTION === '1' || MOVE_DIRECTION === '4') ? '-' : '';
		let r_sign = (MOVE_DIRECTION === '1' || MOVE_DIRECTION === '3') ? '' : '-';
		if(needSpeedVar){
			Arduino.addVarDef('int speed', 'speed');
			result.push(`speed = ${SPEED}`);
			SPEED = 'speed';
		}
		result.push(`dc_1.run(${r_sign}${SPEED})`);
		result.push(`dc_2.run(${l_sign}${SPEED})`);
		return result.map(v => Arduino.tab() + v + END).join('');
	}

	Arduino.weeemake_vehicle_move = function(block){
		let LSPEED;
		let RSPEED = Arduino.valueToCode(block, 'RSPEED', ORDER_NONE);

		if(block.getInputTargetBlock('LSPEED').type != 'speed'){
			LSPEED = '-' + Arduino.valueToCode(block, 'LSPEED', ORDER_UNARY);
		}else{
			LSPEED = - Arduino.valueToCode(block, 'LSPEED', ORDER_NONE);
		}
		
		let result = [];
		Arduino.addVarDef('WeDCMotor dc_1(M1)', 'dc_1');
		Arduino.addVarDef('WeDCMotor dc_2(M2)', 'dc_2');

		result.push(`dc_1.run(${RSPEED})`);
		result.push(`dc_2.run(${LSPEED})`);
		return result.map(v => Arduino.tab() + v + END).join('');
	}

	Arduino.weeemake_robot_stop = function(block){
		Arduino.addVarDef('WeDCMotor dc_1(M1)', 'dc_1');
		Arduino.addVarDef('WeDCMotor dc_2(M2)', 'dc_2');
		let result = [];
		result.push(`dc_1.stop()`);
		result.push(`dc_2.stop()`);
		return result.map(v => Arduino.tab() + v + END).join('');
	}

	Arduino.weeemake_dc_130_motor = function(block){
		//const SENSOR_PORT = Arduino.valueToCode(block, 'SENSOR_PORT', ORDER_NONE);
		const SENSOR_PORT = Arduino.valueToCode(block, 'SENSOR_PORT', ORDER_NONE);
		const SPEED = Arduino.valueToCode(block, 'SPEED', ORDER_NONE);

		let key = `dc130_` + SENSOR_PORT.split('_')[1];

		Arduino.addVarDef(`We130DCMotor ${key}(${SENSOR_PORT})`, key);

		return Arduino.tab() + `${key}.run(${SPEED})` + END;
	}

	Arduino.weeemake_encoder_motor_set_origin = function(block){
		const MOTOR_PORT = block.getFieldValue('MOTOR_PORT');

		let key = `encoder_` + MOTOR_PORT.split('_')[1];

		Arduino.addVarDef(`WeEncoderMotor ${key}(${MOTOR_PORT})`, key);

		return Arduino.tab() + `${key}.setPositionOrigin()` + END;
	}

	Arduino.weeemake_encoder_motor_power = function(block){
		const MOTOR_PORT = block.getFieldValue('MOTOR_PORT');
		const SPEED = Arduino.valueToCode(block, 'SPEED', ORDER_NONE);

		let key = `encoder_` + MOTOR_PORT.split('_')[1];

		Arduino.addVarDef(`WeEncoderMotor ${key}(${MOTOR_PORT})`, key);

		return Arduino.tab() + `${key}.run(${SPEED})` + END;
	}

	Arduino.weeemake_encoder_motor_speed = function(block){
		const MOTOR_PORT = block.getFieldValue('MOTOR_PORT');
		const SPEED = Arduino.valueToCode(block, 'SPEED', ORDER_NONE);

		let key = `encoder_` + MOTOR_PORT.split('_')[1];

		Arduino.addVarDef(`WeEncoderMotor ${key}(${MOTOR_PORT})`, key);

		return Arduino.tab() + `${key}.runSpeed_188(${SPEED})` + END;
	}

	Arduino.weeemake_encoder_motor_move = function(block){
		const MOTOR_PORT = block.getFieldValue('MOTOR_PORT');
		const SPEED = Arduino.valueToCode(block, 'SPEED', ORDER_NONE);

		let key = `encoder_` + MOTOR_PORT.split('_')[1];

		Arduino.addVarDef(`WeEncoderMotor ${key}(${MOTOR_PORT})`, key);

		return Arduino.tab() + `${key}.move(150, ${SPEED})` + END;
	}

	Arduino.weeemake_servo = function(block){
		const SERVO_PORT = Arduino.valueToCode(block, 'SERVO_PORT', ORDER_NONE);
		const ANGLE = Arduino.valueToCode(block, 'ANGLE', ORDER_NONE);

		let key = `servo_` + SERVO_PORT.split('_').pop();

		if(db.board === board_elf_mini){
			Arduino.addVarDef(`WeServo ${key}(${SERVO_PORT})`, key);
		}else{
			Arduino.addVarDef(`Servo ${key}`, key);
			Arduino.addSetupCode(`${key}.attach(${SERVO_PORT})`, key);
		}

		return Arduino.tab() + `${key}.write(${ANGLE})` + END;
	}

	Arduino.weeemake_servo_360 = function(block){
		const SERVO_PORT = Arduino.valueToCode(block, 'SERVO_PORT', ORDER_NONE);
		const ANGLE = Arduino.valueToCode(block, 'ANGLE', ORDER_NONE);

		let key = `servo_` + SERVO_PORT.split('_').pop();

		Arduino.addVarDef(`WeServo360 ${key}(${SERVO_PORT})`, key);

		return Arduino.tab() + `${key}.write(${ANGLE})` + END;
	}

	Arduino.weeemake_buzzer = function(block){
		const NOTE = Arduino.valueToCode(block, 'NOTE', ORDER_NONE);
		const BEAT = Arduino.valueToCode(block, 'BEAT', ORDER_NONE);

		Arduino.addVarDef(`WeBuzzer buzzer(OnBoard_Buzzer)`, 'buzzer');

		return Arduino.tab() + `buzzer.tone(${NOTE}, ${BEAT})` + END;
	}

	Arduino.weeemake_rgb_strip = function(block){
		const SENSOR_PORT = Arduino.valueToCode(block, 'SENSOR_PORT', ORDER_NONE);
		const PIXEL = Arduino.valueToCode(block, 'PIXEL', ORDER_NONE);
		const COLOR = Arduino.valueToCode(block, 'COLOR', ORDER_NONE);

		let key = `rgb_led_` + SENSOR_PORT.split('_')[1];

		Arduino.addVarDef(`WeRGBLed ${key}(${SENSOR_PORT})`, key);

		let result = [
			`${key}.setColor(${PIXEL}, ${COLOR})`,
			`${key}.show()`
		];

		return result.map(v => Arduino.tab() + v + END).join('');
	}

	Arduino.weeemake_rgb3_strip = function(block){
		const SENSOR_PORT = Arduino.valueToCode(block, 'SENSOR_PORT', ORDER_NONE);
		const PIXEL = Arduino.valueToCode(block, 'PIXEL', ORDER_NONE);
		const R = Arduino.valueToCode(block, 'R', ORDER_NONE);
		const G = Arduino.valueToCode(block, 'G', ORDER_NONE);
		const B = Arduino.valueToCode(block, 'B', ORDER_NONE);

		let key = `rgb_led_` + SENSOR_PORT.split('_')[1];

		Arduino.addVarDef(`WeRGBLed ${key}(${SENSOR_PORT})`, key);

		let result = [
			`${key}.setColor(${PIXEL}, ${R}, ${G}, ${B})`,
			`${key}.show()`
		];

		return result.map(v => Arduino.tab() + v + END).join('');
	}

	Arduino.weeemake_rgb_RJ11 = function(block){
		const SENSOR_PORT = Arduino.valueToCode(block, 'SENSOR_PORT', ORDER_NONE);
		//const RJ11_RGB_INDEX = block.getFieldValue('RJ11_RGB_INDEX');
		const INDEX = Arduino.valueToCode(block, 'INDEX', ORDER_NONE);
		const COLOR = Arduino.valueToCode(block, 'COLOR', ORDER_NONE);

		let key = `rgb_led_RJ_` + SENSOR_PORT.split('_')[1];

		Arduino.addVarDef(`WeRGBLED_RJ ${key}(${SENSOR_PORT})`, key);

		let result = [
			`${key}.setColor(${INDEX}, ${COLOR})`,
			`${key}.show()`
		];

		return result.map(v => Arduino.tab() + v + END).join('');
	}

	Arduino.weeemake_rgb3_RJ11 = function(block){
		const SENSOR_PORT = Arduino.valueToCode(block, 'SENSOR_PORT', ORDER_NONE);
		//const RJ11_RGB_INDEX = block.getFieldValue('RJ11_RGB_INDEX');
		//const PIXEL = Arduino.valueToCode(block, 'PIXEL', ORDER_NONE);
		const INDEX = Arduino.valueToCode(block, 'INDEX', ORDER_NONE);
		const R = Arduino.valueToCode(block, 'R', ORDER_NONE);
		const G = Arduino.valueToCode(block, 'G', ORDER_NONE);
		const B = Arduino.valueToCode(block, 'B', ORDER_NONE);

		let key = `rgb_led_RJ_` + SENSOR_PORT.split('_')[1];

		Arduino.addVarDef(`WeRGBLED_RJ ${key}(${SENSOR_PORT})`, key);

		let result = [
			`${key}.setColor(${INDEX}, ${R}, ${G}, ${B})`,
			`${key}.show()`
		];

		return result.map(v => Arduino.tab() + v + END).join('');
	}

	Arduino.weeemake_line_follower = function(block){
		const SENSOR_PORT = Arduino.valueToCode(block, 'SENSOR_PORT', ORDER_NONE);
		const LINE_FOLLOWER_INDEX = block.getFieldValue('LINE_FOLLOWER_INDEX');

		let key = `linefollower_` + SENSOR_PORT.split('_')[1];

		Arduino.addVarDef(`WeLineFollower ${key}(${SENSOR_PORT})`, key);

		return [`${key}.startRead(${LINE_FOLLOWER_INDEX})`, ORDER_HIGH];
	}

	Arduino.weeemake_ultrasonic = function(block){
		const SENSOR_PORT = Arduino.valueToCode(block, 'SENSOR_PORT', ORDER_NONE);

		let key = `ultrasonic_` + SENSOR_PORT.split('_')[1];

		Arduino.addVarDef(`WeUltrasonicSensor ${key}(${SENSOR_PORT})`, key);

		return [`${key}.distanceCm()`, ORDER_HIGH];
	}

	Arduino.weeemake_ultrasonic_led = function(block){
		const SENSOR_PORT = Arduino.valueToCode(block, 'SENSOR_PORT', ORDER_NONE);
		//const ULTRASONIC_LED_INDEX = Arduino.valueToCode(block, 'ULTRASONIC_LED_INDEX', ORDER_NONE);
		const ULTRASONIC_LED_INDEX = Arduino.valueToCode(block, 'ULTRASONIC_LED_INDEX', ORDER_NONE);
		const ON_OFF = Arduino.valueToCode(block, 'ON_OFF', ORDER_NONE);

		let key = `ultrasonic_` + SENSOR_PORT.split('_')[1];

		Arduino.addVarDef(`WeUltrasonicSensor ${key}(${SENSOR_PORT})`, key);

		return Arduino.tab() + `${key}.setLed(${ULTRASONIC_LED_INDEX}, ${ON_OFF})` + END;
	}

	Arduino.weeemake_ultrasonic_rgb = function(block){
		const SENSOR_PORT = Arduino.valueToCode(block, 'SENSOR_PORT', ORDER_NONE);
		const ULTRASONIC_LED_INDEX = Arduino.valueToCode(block, 'ULTRASONIC_LED_INDEX', ORDER_NONE);
		const COLOR = Arduino.valueToCode(block, 'COLOR', ORDER_NONE);

		let key = `ultrasonic_` + SENSOR_PORT.split('_')[1];

		Arduino.addVarDef(`WeUltrasonicSensor ${key}(${SENSOR_PORT})`, key);

		return Arduino.tab() + `${key}.setColor(${ULTRASONIC_LED_INDEX}, ${COLOR})` + END;
	}

	Arduino.weeemake_ultrasonic_rgb3 = function(block){
		const SENSOR_PORT = Arduino.valueToCode(block, 'SENSOR_PORT', ORDER_NONE);
		const ULTRASONIC_LED_INDEX = Arduino.valueToCode(block, 'ULTRASONIC_LED_INDEX', ORDER_NONE);
		const R = Arduino.valueToCode(block, 'R', ORDER_NONE);
		const G = Arduino.valueToCode(block, 'G', ORDER_NONE);
		const B = Arduino.valueToCode(block, 'B', ORDER_NONE);

		let key = `ultrasonic_` + SENSOR_PORT.split('_')[1];

		Arduino.addVarDef(`WeUltrasonicSensor ${key}(${SENSOR_PORT})`, key);

		return Arduino.tab() + `${key}.setColor(${ULTRASONIC_LED_INDEX}, ${R}, ${G}, ${B})` + END;
	}

	Arduino.weeemake_ir_avoid = function(block){
		const SENSOR_PORT = Arduino.valueToCode(block, 'SENSOR_PORT', ORDER_NONE);

		let key = `ir_avoid_` + SENSOR_PORT.split('_')[1];

		Arduino.addVarDef(`WeIRAvoidSensor ${key}(${SENSOR_PORT})`, key);

		return [`${key}.isObstacle()`, ORDER_HIGH];
	}

	Arduino.weeemake_ir_avoid_led = function(block){
		const SENSOR_PORT = Arduino.valueToCode(block, 'SENSOR_PORT', ORDER_NONE);
		const ULTRASONIC_LED_INDEX = Arduino.valueToCode(block, 'ULTRASONIC_LED_INDEX', ORDER_NONE);
		const ON_OFF = Arduino.valueToCode(block, 'ON_OFF', ORDER_NONE);

		let key = `ir_avoid_` + SENSOR_PORT.split('_')[1];

		Arduino.addVarDef(`WeIRAvoidSensor ${key}(${SENSOR_PORT})`, key);

		return Arduino.tab() + `${key}.setLed(${ULTRASONIC_LED_INDEX}, ${ON_OFF})` + END;
	}

	Arduino.weeemake_ir_avoid_rgb = function(block){
		const SENSOR_PORT = Arduino.valueToCode(block, 'SENSOR_PORT', ORDER_NONE);
		const ULTRASONIC_LED_INDEX = Arduino.valueToCode(block, 'ULTRASONIC_LED_INDEX', ORDER_NONE);
		const COLOR = Arduino.valueToCode(block, 'COLOR', ORDER_NONE);

		let key = `ir_avoid_` + SENSOR_PORT.split('_')[1];

		Arduino.addVarDef(`WeIRAvoidSensor ${key}(${SENSOR_PORT})`, key);

		return Arduino.tab() + `${key}.setColor(${ULTRASONIC_LED_INDEX}, ${COLOR})` + END;
	}

	Arduino.weeemake_ir_avoid_rgb3 = function(block){
		const SENSOR_PORT = Arduino.valueToCode(block, 'SENSOR_PORT', ORDER_NONE);
		const ULTRASONIC_LED_INDEX = Arduino.valueToCode(block, 'ULTRASONIC_LED_INDEX', ORDER_NONE);
		const R = Arduino.valueToCode(block, 'R', ORDER_NONE);
		const G = Arduino.valueToCode(block, 'G', ORDER_NONE);
		const B = Arduino.valueToCode(block, 'B', ORDER_NONE);

		let key = `ir_avoid_` + SENSOR_PORT.split('_')[1];

		Arduino.addVarDef(`WeIRAvoidSensor ${key}(${SENSOR_PORT})`, key);

		return Arduino.tab() + `${key}.setColor(${ULTRASONIC_LED_INDEX}, ${R}, ${G}, ${B})` + END;
	}

	Arduino.weeemake_led_matrix_number = function(block){
		const LED_MATRIX_TYPE = block.getFieldValue('LED_MATRIX_TYPE').replace('x', '_');
		const SENSOR_PORT = Arduino.valueToCode(block, 'SENSOR_PORT', ORDER_NONE);
		const NUM = Arduino.valueToCode(block, 'NUM', ORDER_NONE);

		let key = `led_matrix_` + SENSOR_PORT.split('_')[1];

		Arduino.addVarDef(`WeLEDPanelModuleMatrix${LED_MATRIX_TYPE} ${key}(${SENSOR_PORT})`, key);

		return Arduino.tab() + `${key}.showNum(${NUM})` + END;
	}

	Arduino.weeemake_led_matrix_time = function(block){
		const LED_MATRIX_TYPE = block.getFieldValue('LED_MATRIX_TYPE').replace('x', '_');
		const SENSOR_PORT = Arduino.valueToCode(block, 'SENSOR_PORT', ORDER_NONE);
		const HOUR = Arduino.valueToCode(block, 'HOUR', ORDER_NONE);
		const SHOW_COLON = Arduino.valueToCode(block, 'SHOW_COLON', ORDER_NONE);
		const MINUTE = Arduino.valueToCode(block, 'MINUTE', ORDER_NONE);

		let key = `led_matrix_` + SENSOR_PORT.split('_')[1];

		Arduino.addVarDef(`WeLEDPanelModuleMatrix${LED_MATRIX_TYPE} ${key}(${SENSOR_PORT})`, key);

		return Arduino.tab() + `${key}.showClock(${HOUR}, ${MINUTE}, ${SHOW_COLON})` + END;
	}

	Arduino.weeemake_led_matrix_string = function(block){
		const LED_MATRIX_TYPE = block.getFieldValue('LED_MATRIX_TYPE').replace('x', '_');
		const SENSOR_PORT = Arduino.valueToCode(block, 'SENSOR_PORT', ORDER_NONE);
		const X = Arduino.valueToCode(block, 'X', ORDER_NONE);
		const Y = Arduino.valueToCode(block, 'Y', ORDER_NONE);
		const STR = Arduino.valueToCode(block, 'STR', ORDER_NONE);

		let key = `led_matrix_` + SENSOR_PORT.split('_')[1];

		Arduino.addVarDef(`WeLEDPanelModuleMatrix${LED_MATRIX_TYPE} ${key}(${SENSOR_PORT})`, key);

		return Arduino.tab() + `${key}.showChar(${X}, ${Y}, ${STR})` + END;
	}

	Arduino.weeemake_led_matrix_bitmap = function(block){
		const SENSOR_PORT = Arduino.valueToCode(block, 'SENSOR_PORT', ORDER_NONE);
		const X = Arduino.valueToCode(block, 'X', ORDER_NONE);
		const Y = Arduino.valueToCode(block, 'Y', ORDER_NONE);
		const MATRIX = Arduino.valueToCode(block, 'MATRIX', ORDER_NONE);
		const IS_BIG_MATRIX = block.getInputTargetBlock('MATRIX').type.indexOf("21*7") > 0;

		let key = `led_matrix_` + SENSOR_PORT.split('_')[1];

		Arduino.addVarDef(`WeLEDPanelModuleMatrix${IS_BIG_MATRIX ? '7_21' : '5_14'} ${key}(${SENSOR_PORT})`, key);
		let data = IS_BIG_MATRIX ? castMatrix(MATRIX, 21, 7) : castMatrix(MATRIX, 14, 5);
		return Arduino.tab() + `${key}.showBitmap2(${X}, ${Y}, ${data})` + END;
	}

	Arduino.weeemake_led_matrix_bitmap_21x7 = function(block){
		const SENSOR_PORT = Arduino.valueToCode(block, 'SENSOR_PORT', ORDER_NONE);
		const X = Arduino.valueToCode(block, 'X', ORDER_NONE);
		const Y = Arduino.valueToCode(block, 'Y', ORDER_NONE);
		const MATRIX = Arduino.valueToCode(block, 'MATRIX', ORDER_NONE);

		let key = `led_matrix_` + SENSOR_PORT.split('_')[1];

		Arduino.addVarDef(`WeLEDPanelModuleMatrix7_21 ${key}(${SENSOR_PORT})`, key);

		return Arduino.tab() + `${key}.showBitmap2(${X}, ${Y}, ${castMatrix(MATRIX, 21, 7)})` + END;
	}

	Arduino.weeemake_led_matrix_bitmap_14x5 = function(block){
		const SENSOR_PORT = Arduino.valueToCode(block, 'SENSOR_PORT', ORDER_NONE);
		const X = Arduino.valueToCode(block, 'X', ORDER_NONE);
		const Y = Arduino.valueToCode(block, 'Y', ORDER_NONE);
		const MATRIX = Arduino.valueToCode(block, 'MATRIX', ORDER_NONE);

		let key = `led_matrix_` + SENSOR_PORT.split('_')[1];

		Arduino.addVarDef(`WeLEDPanelModuleMatrix5_14 ${key}(${SENSOR_PORT})`, key);

		return Arduino.tab() + `${key}.showBitmap2(${X}, ${Y}, ${castMatrix(MATRIX, 14, 5)})` + END;
	}
/*
	Arduino.weeemake_led_matrix_pixel_show = function(block){
		const LED_MATRIX_TYPE = block.getFieldValue('LED_MATRIX_TYPE').replace('x', '_');
		const SENSOR_PORT = Arduino.valueToCode(block, 'SENSOR_PORT', ORDER_NONE);
		const X = Arduino.valueToCode(block, 'X', ORDER_NONE);
		const Y = Arduino.valueToCode(block, 'Y', ORDER_NONE);

		let key = `led_matrix_` + SENSOR_PORT.split('_')[1];

		Arduino.addVarDef(`WeLEDPanelModuleMatrix${LED_MATRIX_TYPE} ${key}(${SENSOR_PORT})`, key);

		return Arduino.tab() + `${key}.turnOnDot(${X}, ${Y})` + END;
	}

	Arduino.weeemake_led_matrix_pixel_hide = function(block){
		const LED_MATRIX_TYPE = block.getFieldValue('LED_MATRIX_TYPE').replace('x', '_');
		const SENSOR_PORT = Arduino.valueToCode(block, 'SENSOR_PORT', ORDER_NONE);
		const X = Arduino.valueToCode(block, 'X', ORDER_NONE);
		const Y = Arduino.valueToCode(block, 'Y', ORDER_NONE);

		let key = `led_matrix_` + SENSOR_PORT.split('_')[1];

		Arduino.addVarDef(`WeLEDPanelModuleMatrix${LED_MATRIX_TYPE} ${key}(${SENSOR_PORT})`, key);

		return Arduino.tab() + `${key}.turnOffDot(${X}, ${Y})` + END;
	}
*/
	Arduino.weeemake_led_matrix_pixel = function(block){
		const LED_MATRIX_TYPE = block.getFieldValue('LED_MATRIX_TYPE').replace('x', '_');
		const SENSOR_PORT = Arduino.valueToCode(block, 'SENSOR_PORT', ORDER_NONE);
		const SHOW_HIDE = Arduino.valueToCode(block, 'SHOW_HIDE', ORDER_NONE);
		const X = Arduino.valueToCode(block, 'X', ORDER_NONE);
		const Y = Arduino.valueToCode(block, 'Y', ORDER_NONE);

		let key = `led_matrix_` + SENSOR_PORT.split('_')[1];

		Arduino.addVarDef(`WeLEDPanelModuleMatrix${LED_MATRIX_TYPE} ${key}(${SENSOR_PORT})`, key);

		return Arduino.tab() + `${key}.setDot(${X}, ${Y}, ${SHOW_HIDE})` + END;
	};

	Arduino.weeemake_led_matrix_clear = function(block){
		const LED_MATRIX_TYPE = block.getFieldValue('LED_MATRIX_TYPE').replace('x', '_');
		const SENSOR_PORT = Arduino.valueToCode(block, 'SENSOR_PORT', ORDER_NONE);

		let key = `led_matrix_` + SENSOR_PORT.split('_')[1];

		Arduino.addVarDef(`WeLEDPanelModuleMatrix${LED_MATRIX_TYPE} ${key}(${SENSOR_PORT})`, key);

		return Arduino.tab() + `${key}.clearScreen()` + END;
	}

	Arduino.weeemake_humiture = (list => function(block){
		const SENSOR_PORT = Arduino.valueToCode(block, 'SENSOR_PORT', ORDER_NONE);
		const HUMITURE_TYPE = block.getFieldValue('HUMITURE_TYPE');

		let key = `humiture_` + SENSOR_PORT.split('_')[1];

		Arduino.addVarDef(`WeHumiture ${key}(${SENSOR_PORT})`, key);
		Arduino.addLoopCode(`${key}.startRead()`, key);

		return [`${key}.${list[HUMITURE_TYPE]}()`, ORDER_HIGH];
	})(['getHumidity', 'getTemperature']);

	Arduino.weeemake_seven_segment = function(block){
		const SENSOR_PORT = Arduino.valueToCode(block, 'SENSOR_PORT', ORDER_NONE);
		const NUM = Arduino.valueToCode(block, 'NUM', ORDER_NONE);

		let key = `seven_segment_` + SENSOR_PORT.split('_')[1];

		Arduino.addVarDef(`We7SegmentDisplay ${key}(${SENSOR_PORT})`, key);

		return Arduino.tab() + `${key}.showNumber(${NUM})` + END;
	}
/*
	Arduino.weeemake_sliding_potentiometer = function(block){
		const SENSOR_PORT = Arduino.valueToCode(block, 'SENSOR_PORT', ORDER_NONE);

		let key = `potentiometer_` + SENSOR_PORT.split('_')[1];

		Arduino.addVarDef(`WeSlidingPotentiometer ${key}(${SENSOR_PORT})`, key);

		return [`${key}.readAnalog()`, ORDER_HIGH];
	}
*/
	Arduino.weeemake_potentiometer = function(block){
		const SENSOR_PORT = Arduino.valueToCode(block, 'SENSOR_PORT', ORDER_NONE);

		let key = `potentiometer_` + SENSOR_PORT.split('_')[1];

		Arduino.addVarDef(`WePotentiometer ${key}(${SENSOR_PORT})`, key);

		return [`${key}.readAnalog()`, ORDER_HIGH];
	}

	Arduino.weeemake_gas_sensor = function(block){
		const SENSOR_PORT = Arduino.valueToCode(block, 'SENSOR_PORT', ORDER_NONE);

		let key = `gas_sensor_` + SENSOR_PORT.split('_')[1];

		Arduino.addVarDef(`WeGasSensor ${key}(${SENSOR_PORT})`, key);

		return [`${key}.readAnalog()`, ORDER_HIGH];
	}

	Arduino.weeemake_led_button_light = function(block){
		const SENSOR_PORT = Arduino.valueToCode(block, 'SENSOR_PORT', ORDER_NONE);
		const BUTTON_INDEX = block.getFieldValue('BUTTON_INDEX');
		const ON_OFF = Arduino.valueToCode(block, 'ON_OFF', ORDER_NONE);

		let key = `button4led_` + SENSOR_PORT.split('_')[1];

		Arduino.addVarDef(`We4LEDButton ${key}(${SENSOR_PORT})`, key);

		return Arduino.tab() + `${key}.setLed(${BUTTON_INDEX}, ${ON_OFF})` + END;
	}

	Arduino.weeemake_relay = function(block){
		const SENSOR_PORT = Arduino.valueToCode(block, 'SENSOR_PORT', ORDER_NONE);
		const ON_OFF = Arduino.valueToCode(block, 'ON_OFF', ORDER_NONE);

		let key = `relay_` + SENSOR_PORT.split('_')[1];

		Arduino.addVarDef(`WeRelay ${key}(${SENSOR_PORT})`, key);

		return Arduino.tab() + `${key}.setNC(${ON_OFF})` + END;
	}

	Arduino.weeemake_water_atomizer = function(block){
		const SENSOR_PORT = Arduino.valueToCode(block, 'SENSOR_PORT', ORDER_NONE);
		const ON_OFF = Arduino.valueToCode(block, 'ON_OFF', ORDER_NONE);

		let key = `water_atomizer_` + SENSOR_PORT.split('_')[1];

		Arduino.addVarDef(`WeWaterAtomizer ${key}(${SENSOR_PORT})`, key);
		
		return Arduino.tab() + `${key}.setRun(${ON_OFF})` + END;
	}

	Arduino.weeemake_color_sensor_white_balance = function(block){
		const SENSOR_PORT = Arduino.valueToCode(block, 'SENSOR_PORT', ORDER_NONE);

		let key = `color_sensor_` + SENSOR_PORT.split('_')[1];

		Arduino.addVarDef(`WeColorSensor ${key}(${SENSOR_PORT})`, key);
		
		return Arduino.tab() + `${key}.whitebalance()` + END;
	}

	Arduino.weeemake_color_sensor_light = function(block){
		const SENSOR_PORT = Arduino.valueToCode(block, 'SENSOR_PORT', ORDER_NONE);
		const ON_OFF = Arduino.valueToCode(block, 'ON_OFF', ORDER_NONE);

		let key = `color_sensor_` + SENSOR_PORT.split('_')[1];

		Arduino.addVarDef(`WeColorSensor ${key}(${SENSOR_PORT})`, key);
		
		return Arduino.tab() + `${key}.setLight(${ON_OFF})` + END;
	}

	Arduino.weeemake_mp3_do = (list => function(block){
		const SENSOR_PORT = Arduino.valueToCode(block, 'SENSOR_PORT', ORDER_NONE);
		const MP3_FUNCTION = block.getFieldValue('MP3_FUNCTION');

		let key = `mp3_` + SENSOR_PORT.split('_')[1];

		Arduino.addVarDef(`WeMP3 ${key}(${SENSOR_PORT})`, key);
		
		return Arduino.tab() + `${key}.${list[MP3_FUNCTION]}()` + END;
	})(['play', 'pause', 'prevMusic', 'nextMusic']);
/*
	Arduino.weeemake_mp3_play = function(block){
		const SENSOR_PORT = Arduino.valueToCode(block, 'SENSOR_PORT', ORDER_NONE);

		let key = `mp3_` + SENSOR_PORT.split('_')[1];

		Arduino.addVarDef(`WeMP3 ${key}(${SENSOR_PORT})`, key);
		
		return Arduino.tab() + `${key}.play()` + END;
	}

	Arduino.weeemake_mp3_pause = function(block){
		const SENSOR_PORT = Arduino.valueToCode(block, 'SENSOR_PORT', ORDER_NONE);

		let key = `mp3_` + SENSOR_PORT.split('_')[1];

		Arduino.addVarDef(`WeMP3 ${key}(${SENSOR_PORT})`, key);
		
		return Arduino.tab() + `${key}.pause()` + END;
	}

	Arduino.weeemake_mp3_prev_music = function(block){
		const SENSOR_PORT = Arduino.valueToCode(block, 'SENSOR_PORT', ORDER_NONE);

		let key = `mp3_` + SENSOR_PORT.split('_')[1];

		Arduino.addVarDef(`WeMP3 ${key}(${SENSOR_PORT})`, key);
		
		return Arduino.tab() + `${key}.prevMusic()` + END;
	}

	Arduino.weeemake_mp3_next_music = function(block){
		const SENSOR_PORT = Arduino.valueToCode(block, 'SENSOR_PORT', ORDER_NONE);

		let key = `mp3_` + SENSOR_PORT.split('_')[1];

		Arduino.addVarDef(`WeMP3 ${key}(${SENSOR_PORT})`, key);
		
		return Arduino.tab() + `${key}.nextMusic()` + END;
	}
*/
	Arduino.weeemake_mp3_set_music = function(block){
		const SENSOR_PORT = Arduino.valueToCode(block, 'SENSOR_PORT', ORDER_NONE);
		const NUM = Arduino.valueToCode(block, 'NUM', ORDER_NONE);

		let key = `mp3_` + SENSOR_PORT.split('_')[1];

		Arduino.addVarDef(`WeMP3 ${key}(${SENSOR_PORT})`, key);
		
		return Arduino.tab() + `${key}.appointMusic(${NUM})` + END;
	}

	Arduino.weeemake_mp3_set_volume = function(block){
		const SENSOR_PORT = Arduino.valueToCode(block, 'SENSOR_PORT', ORDER_NONE);
		const NUM = Arduino.valueToCode(block, 'NUM', ORDER_NONE);

		let key = `mp3_` + SENSOR_PORT.split('_')[1];

		Arduino.addVarDef(`WeMP3 ${key}(${SENSOR_PORT})`, key);
		
		return Arduino.tab() + `${key}.appointVolume(${NUM})` + END;
	}

	Arduino.weeemake_mp3_set_device = function(block){
		const SENSOR_PORT = Arduino.valueToCode(block, 'SENSOR_PORT', ORDER_NONE);
		const MP3_DEVICE_TYPE = block.getFieldValue('MP3_DEVICE_TYPE');

		let key = `mp3_` + SENSOR_PORT.split('_')[1];

		Arduino.addVarDef(`WeMP3 ${key}(${SENSOR_PORT})`, key);
		
		return Arduino.tab() + `${key}.appointDevice(${MP3_DEVICE_TYPE})` + END;
	}

	Arduino.weeemake_mp3_is_over = function(block){
		const SENSOR_PORT = Arduino.valueToCode(block, 'SENSOR_PORT', ORDER_NONE);

		let key = `mp3_` + SENSOR_PORT.split('_')[1];

		Arduino.addVarDef(`WeMP3 ${key}(${SENSOR_PORT})`, key);

		return [`${key}.isOver()`, ORDER_HIGH];
	}

	Arduino.weeemake_oled_set_size = function(block){
		const SENSOR_PORT = Arduino.valueToCode(block, 'SENSOR_PORT', ORDER_NONE);
		const OLED_SIZE = block.getFieldValue('OLED_SIZE');

		let key = `oled_` + SENSOR_PORT.split('_')[1];

		Arduino.addVarDef(`WeOLED ${key}(${SENSOR_PORT})`, key);

		return Arduino.tab() + `${key}.setSize(${OLED_SIZE})` + END;
	}

	Arduino.weeemake_oled_show_string = function(block){
		const SENSOR_PORT = Arduino.valueToCode(block, 'SENSOR_PORT', ORDER_NONE);
		const X = Arduino.valueToCode(block, 'X', ORDER_NONE);
		const Y = Arduino.valueToCode(block, 'Y', ORDER_NONE);
		const STR = Arduino.valueToCode(block, 'STR', ORDER_NONE);

		let key = `oled_` + SENSOR_PORT.split('_')[1];

		Arduino.addVarDef(`WeOLED ${key}(${SENSOR_PORT})`, key);

		return Arduino.tab() + `${key}.showString(${X}, ${Y}, ${STR})` + END;
	}

	Arduino.weeemake_oled_show_number = function(block){
		const SENSOR_PORT = Arduino.valueToCode(block, 'SENSOR_PORT', ORDER_NONE);
		const X = Arduino.valueToCode(block, 'X', ORDER_NONE);
		const Y = Arduino.valueToCode(block, 'Y', ORDER_NONE);
		const NUM = Arduino.valueToCode(block, 'NUM', ORDER_NONE);

		let key = `oled_` + SENSOR_PORT.split('_')[1];

		Arduino.addVarDef(`WeOLED ${key}(${SENSOR_PORT})`, key);

		return Arduino.tab() + `${key}.showNum(${X}, ${Y}, ${NUM})` + END;
	}

	Arduino.weeemake_oled_clear_screen = function(block){
		const SENSOR_PORT = Arduino.valueToCode(block, 'SENSOR_PORT', ORDER_NONE);

		let key = `oled_` + SENSOR_PORT.split('_')[1];

		Arduino.addVarDef(`WeOLED ${key}(${SENSOR_PORT})`, key);

		return Arduino.tab() + `${key}.clearScreen()` + END;
	}

	Arduino.weeemake_color_sensor = (list => function(block){
		const SENSOR_PORT = Arduino.valueToCode(block, 'SENSOR_PORT', ORDER_NONE);
		const COLOR_TYPE = block.getFieldValue('COLOR_TYPE');

		let key = `color_sensor_` + SENSOR_PORT.split('_')[1];

		Arduino.addVarDef(`WeColorSensor ${key}(${SENSOR_PORT})`, key);
		Arduino.addLoopCode(`${key}.readColorData()`, key);

		return [`${key}.${list[COLOR_TYPE]}()`, ORDER_HIGH];
	})(['showColorData', 'showRedData', 'showGreenData', 'showBlueData']);

	Arduino.weeemake_flame_sensor = function(block){
		const SENSOR_PORT = Arduino.valueToCode(block, 'SENSOR_PORT', ORDER_NONE);
		const FLAME_INDEX = block.getFieldValue('FLAME_INDEX');

		let key = `flame_sensor_` + SENSOR_PORT.split('_')[1];

		Arduino.addVarDef(`WeFlameSensor ${key}(${SENSOR_PORT})`, key);
		Arduino.addLoopCode(`${key}.readData()`, key);

		return [`${key}.showSensor${FLAME_INDEX}()`, ORDER_HIGH];
	}

	Arduino.weeemake_joystick = (list => function(block){
		const SENSOR_PORT = Arduino.valueToCode(block, 'SENSOR_PORT', ORDER_NONE);
		const AXIS2 = block.getFieldValue('AXIS2');

		let key = `joystick_` + SENSOR_PORT.split('_')[1];

		Arduino.addVarDef(`WeJoystick ${key}(${SENSOR_PORT})`, key);
		Arduino.addLoopCode(`${key}.readData()`, key);

		return [`${key}.${list[AXIS2]}()`, ORDER_HIGH];
	})(['showX', 'showY']);

	Arduino.weeemake_compass = (list => function(block){
		const SENSOR_PORT = Arduino.valueToCode(block, 'SENSOR_PORT', ORDER_NONE);
		const AXIS3 = block.getFieldValue('AXIS3');

		let key = `compass_` + SENSOR_PORT.split('_')[1];

		Arduino.addVarDef(`WeCompassSensor ${key}(${SENSOR_PORT})`, key);
		Arduino.addLoopCode(`${key}.update()`, key);

		return [`${key}.${list[AXIS3]}()`, ORDER_HIGH];
	})(['getHeadS', 'getHeadY', 'getHeadZ']);

	Arduino.weeemake_gyro = (list => function(block){
		const SENSOR_PORT = Arduino.valueToCode(block, 'SENSOR_PORT', ORDER_NONE);
		const GYRO_VALUE_TYPE = block.getFieldValue('GYRO_VALUE_TYPE');
		const AXIS3 = block.getFieldValue('AXIS3');

		let key = `gyro_` + SENSOR_PORT.split('_')[1];

		Arduino.addVarDef(`WeGyroSensor ${key}(${SENSOR_PORT})`, key);
		Arduino.addSetupCode(`${key}.begin()`, key);
		Arduino.addLoopCode(`${key}.update()`, key);

		return [`${key}.${list[GYRO_VALUE_TYPE][AXIS3]}()`, ORDER_HIGH];
	})([['getAngleX', 'getAngleY', 'getAngleZ'],['getGyroX', 'getGyroY', 'getGyroZ']]);

	Arduino.weeemake_touch = function(block){
		const SENSOR_PORT = Arduino.valueToCode(block, 'SENSOR_PORT', ORDER_NONE);

		let key = `touch_sensor_` + SENSOR_PORT.split('_')[1];

		Arduino.addVarDef(`WeTouchSensor ${key}(${SENSOR_PORT})`, key);

		return [`${key}.touched()`, ORDER_HIGH];
	}

	Arduino.weeemake_led_button = function(block){
		const SENSOR_PORT = Arduino.valueToCode(block, 'SENSOR_PORT', ORDER_NONE);
		const BUTTON_INDEX = block.getFieldValue('BUTTON_INDEX');

		let key = `button4led_` + SENSOR_PORT.split('_')[1];

		Arduino.addVarDef(`We4LEDButton ${key}(${SENSOR_PORT})`, key);

		return [`${key}.readValue() == ${1 << (BUTTON_INDEX - 1)}`, ORDER_RELATIONAL];
	}

	Arduino.weeemake_pir = function(block){
		const SENSOR_PORT = Arduino.valueToCode(block, 'SENSOR_PORT', ORDER_NONE);

		let key = `pir_` + SENSOR_PORT.split('_')[1];

		Arduino.addVarDef(`WePIRSensor ${key}(${SENSOR_PORT})`, key);

		return [`${key}.readSensor()`, ORDER_HIGH];
	}

	Arduino.weeemake_tilt = function(block){
		const SENSOR_PORT = Arduino.valueToCode(block, 'SENSOR_PORT', ORDER_NONE);
		const TILT_DIR = block.getFieldValue('TILT_DIR');

		let key = `tilt_` + SENSOR_PORT.split('_')[1];
		let valueKey = key + '_value';

		Arduino.addVarDef(`WeTiltSwitch ${key}(${SENSOR_PORT})`, key);
		Arduino.addVarDef(`int ${valueKey}`, valueKey);
		Arduino.addLoopCode(`${valueKey} = ${key}.readSensor()`);

		return [`${valueKey} == ${TILT_DIR}`, ORDER_RELATIONAL];
	}

	Arduino.weeemake_limit_switch = function(block){
		const SENSOR_PORT = Arduino.valueToCode(block, 'SENSOR_PORT', ORDER_NONE);

		Arduino.addSetupCode(`pinMode(${SENSOR_PORT}, INPUT)`, SENSOR_PORT);
		return [`digitalRead(${SENSOR_PORT})`, ORDER_HIGH];
	}

	Arduino.weeemake_temperature = function(block){
		const SENSOR_PORT = Arduino.valueToCode(block, 'SENSOR_PORT', ORDER_NONE);

		let key = `temperature_` + SENSOR_PORT.split('_')[1];

		Arduino.addVarDef(`WeTemperature ${key}(${SENSOR_PORT})`, key);

		return [`${key}.temperature()`, ORDER_HIGH];
	}

	Arduino.weeemake_barometer = (list => function(block){
		const SENSOR_PORT = Arduino.valueToCode(block, 'SENSOR_PORT', ORDER_NONE);
		const BAROMETER_TYPE = block.getFieldValue('BAROMETER_TYPE');

		let key = `barometer_` + SENSOR_PORT.split('_')[1];

		Arduino.addVarDef(`WeBarometerSensor ${key}(${SENSOR_PORT})`, key);

		return [`${key}.${list[BAROMETER_TYPE]}()`, ORDER_HIGH];
	})(['readHeight', 'readPressure', 'readTemp']);

	Arduino.weeemake_PM25 = (list => function(block){
		const SENSOR_PORT = Arduino.valueToCode(block, 'SENSOR_PORT', ORDER_NONE);
		const PM25_TYPE = block.getFieldValue('PM25_TYPE');

		let key = `PM25_` + SENSOR_PORT.split('_')[1];

		Arduino.addVarDef(`WePM25Sensor ${key}(${SENSOR_PORT})`, key);

		return [`${key}.${list[PM25_TYPE]}()`, ORDER_HIGH];
	})([
	'readPm1_0Concentration', 'readPm2_5Concentration', 'readPm10Concentration',
	'read0_3NumIn100ml', 'read0_5NumIn100ml', 'read1_0NumIn100ml', 'read2_5NumIn100ml',
	'read5_0NumIn100ml', 'read10NumIn100ml']);

	Arduino.weeemake_funny_touch = function(block){
		const SENSOR_PORT = Arduino.valueToCode(block, 'SENSOR_PORT', ORDER_NONE);
		const FUNNY_TOUCH_INDEX = block.getFieldValue('FUNNY_TOUCH_INDEX');

		let index = (1 << FUNNY_TOUCH_INDEX - 1).toString(16);
		let key = `funnyTouch_` + SENSOR_PORT.split('_')[1];

		Arduino.addVarDef(`WeFunnyTouchSensor ${key}(${SENSOR_PORT})`, key);

		return [`(${key}.readValue() & 0x${index}) > 0`, ORDER_RELATIONAL];
	}

	Arduino.weeemake_funny_touch_value = function(block){
		const SENSOR_PORT = Arduino.valueToCode(block, 'SENSOR_PORT', ORDER_NONE);

		let key = `funnyTouch_` + SENSOR_PORT.split('_')[1];

		Arduino.addVarDef(`WeFunnyTouchSensor ${key}(${SENSOR_PORT})`, key);

		return [`${key}.readValue()`, ORDER_HIGH];
	}

	Arduino.weeemake_uv = function(block){
		const SENSOR_PORT = Arduino.valueToCode(block, 'SENSOR_PORT', ORDER_NONE);

		let key = `uv_` + SENSOR_PORT.split('_')[1];

		Arduino.addVarDef(`WeUVSensor ${key}(${SENSOR_PORT})`, key);

		return [`${key}.readAnalog()`, ORDER_HIGH];
	}

	Arduino.weeemake_raindrop = function(block){
		const SENSOR_PORT = Arduino.valueToCode(block, 'SENSOR_PORT', ORDER_NONE);

		let key = `water_` + SENSOR_PORT.split('_')[1];

		Arduino.addVarDef(`WeWaterSensor ${key}(${SENSOR_PORT})`, key);

		return [`${key}.readAnalog()`, ORDER_HIGH];
	}

	function addUsbHost(){
		const key = `usb_host`;
		const args = db.board === board_elf_2560 ? '51, 50' : 'A3, A2';
		
		Arduino.addInclude('#include"SoftwareSerial.h"');
		Arduino.addVarDef(`WeUSBHost ${key}(${args})`, key);
		Arduino.addSetupCode(`${key}.init(USB2_0)`, key);
		Arduino.addLoopCode(`${key}.loop()`, key);

		return key;
	}

	Arduino.weeemake_usb_host = (list => function(block){
		const USB_HOST_KEY = block.getFieldValue('USB_HOST_KEY');

		let key = addUsbHost();

		return [`${key}.ButtonPressed(${list[USB_HOST_KEY]})`, ORDER_HIGH];
	})([null, 'WeJOYSTICK_START', 'WeJOYSTICK_SELECT', 'WeJOYSTICK_MODE', 'WeJOYSTICK_BUTTON_L', 'WeJOYSTICK_BUTTON_R',
	'WeJOYSTICK_RIGHT_L', 'WeJOYSTICK_LEFT_L', 'WeJOYSTICK_UP_L', 'WeJOYSTICK_DOWN_L', 'WeJOYSTICK_UP_R',
	'WeJOYSTICK_RIGHT_R', 'WeJOYSTICK_DOWN_R', 'WeJOYSTICK_LEFT_R', 'WeJOYSTICK_L1', 'WeJOYSTICK_R1', 'WeJOYSTICK_L2', 'WeJOYSTICK_R2']);

	Arduino.weeemake_usb_host_joystick = function(block){
		const LEFT_RIGHT = block.getFieldValue('LEFT_RIGHT');
		const AXIS2 = block.getFieldValue('AXIS2');

		let key = addUsbHost();

		let a = LEFT_RIGHT == '0' ? 'L' : 'R';
		let b = AXIS2 == '0' ? 'x' : 'y';
		let method = 'joystick' + a + b;

		return [`${key}.${method}()`, ORDER_HIGH];
	}

	Arduino.weeemake_digitalRead = function(block){
		const DPIN = block.getFieldValue('DPIN');

		Arduino.addSetupCode(`pinMode(${DPIN}, INPUT)`, DPIN);
		return [`digitalRead(${DPIN})`, ORDER_HIGH];
	}

	Arduino.weeemake_digitalWrite = function(block){
		const DPIN = block.getFieldValue('DPIN');
		const HIGH_LOW = block.getFieldValue('HIGH_LOW');

		Arduino.addSetupCode(`pinMode(${DPIN}, OUTPUT)`, DPIN);
		return Arduino.tab() + `digitalWrite(${DPIN}, ${INT_2_HIGH_LOW[HIGH_LOW]})` + END;
	}

	Arduino.weeemake_analogRead = function(block){
		const APIN = block.getFieldValue('APIN');

		Arduino.addSetupCode(`pinMode(${APIN}, INPUT)`, APIN);
		return [`analogRead(${APIN})`, ORDER_HIGH];
	}

	Arduino.weeemake_analogWrite = function(block){
		const PWM_PIN = block.getFieldValue('PWM_PIN');
		const VALUE = this.valueToCode(block, 'VALUE', ORDER_NONE);

		Arduino.addSetupCode(`pinMode(${PWM_PIN}, OUTPUT)`, PWM_PIN);
		return Arduino.tab() + `analogWrite(${PWM_PIN}, ${VALUE})` + END;
	}

	Arduino.weeemake_ascii2str = function(block){
		const VALUE = this.valueToCode(block, 'VALUE', ORDER_NONE);
		return [`String(char(${VALUE}))`, ORDER_HIGH];
	}

	Arduino.weeemake_str2ascii = function(block){
		const VALUE = this.valueToCode(block, 'VALUE', ORDER_NONE);
		return [`int(${VALUE}[0])`, ORDER_HIGH];
	}

	Arduino.weeemake_speech_recognition_set_keyword = function(block){
		const SENSOR_PORT = Arduino.valueToCode(block, 'SENSOR_PORT', ORDER_NONE);
		const VALUE = castToStr(block, 'VALUE', v => pinyin(v).trim());

		let key = `speechRec_` + SENSOR_PORT.split('_')[1];

		Arduino.addVarDef(`WeSpeechRecognition ${key}(${SENSOR_PORT})`, key);

		return Arduino.tab() + `${key}.setKeyword(${VALUE})` + END;
	}

	Arduino.weeemake_speech_recognition_set_password = function(block){
		const SENSOR_PORT = Arduino.valueToCode(block, 'SENSOR_PORT', ORDER_NONE);
		const INDEX = block.getFieldValue('INDEX');
		const VALUE = castToStr(block, 'VALUE', v => pinyin(v).trim());

		let key = `speechRec_` + SENSOR_PORT.split('_')[1];

		Arduino.addVarDef(`WeSpeechRecognition ${key}(${SENSOR_PORT})`, key);

		return Arduino.tab() + `${key}.setPassword(${INDEX}, ${VALUE})` + END;
	}

	Arduino.weeemake_speech_recognition_set_mode = function(block){
		const SENSOR_PORT = Arduino.valueToCode(block, 'SENSOR_PORT', ORDER_NONE);
		const SPEECH_RECOGNITION_MODE = block.getFieldValue('SPEECH_RECOGNITION_MODE');

		let key = `speechRec_` + SENSOR_PORT.split('_')[1];

		Arduino.addVarDef(`WeSpeechRecognition ${key}(${SENSOR_PORT})`, key);

		return Arduino.tab() + `${key}.setTriggerMode(${SPEECH_RECOGNITION_MODE})` + END;
	}

	Arduino.weeemake_speech_recognition_read = function(block){
		const SENSOR_PORT = Arduino.valueToCode(block, 'SENSOR_PORT', ORDER_NONE);

		let key = `speechRec_` + SENSOR_PORT.split('_')[1];
		let valueKey = key + '_value';

		Arduino.addVarDef(`WeSpeechRecognition ${key}(${SENSOR_PORT})`, key);
		Arduino.addVarDef(`int ${valueKey}`, valueKey);
		Arduino.addLoopCode(`${valueKey} = ${key}.read()`);

		return [valueKey, ORDER_HIGH];
	}

	Arduino.weeemake_gesture_recognition_read = function(block){
		const SENSOR_PORT = Arduino.valueToCode(block, 'SENSOR_PORT', ORDER_NONE);
		const GESTURE_RECOGNITION_TYPE = block.getFieldValue('GESTURE_RECOGNITION_TYPE');

		let key = `gestureSensor_` + SENSOR_PORT.split('_')[1];
		let valueKey = key + '_value';

		Arduino.addVarDef(`WeGestureSensor ${key}(${SENSOR_PORT})`, key);
		Arduino.addVarDef(`int ${valueKey}`, valueKey);
		Arduino.addLoopCode(`${valueKey} = ${key}.read()`);

		return [`${valueKey} == ${GESTURE_RECOGNITION_TYPE}`, ORDER_RELATIONAL];
	}

	Arduino.weeemake_gesture_recognition_value = function(block){
		const SENSOR_PORT = Arduino.valueToCode(block, 'SENSOR_PORT', ORDER_NONE);
		//const GESTURE_RECOGNITION_TYPE = block.getFieldValue('GESTURE_RECOGNITION_TYPE');

		let key = `gestureSensor_` + SENSOR_PORT.split('_')[1];
		let valueKey = key + '_value';

		Arduino.addVarDef(`WeGestureSensor ${key}(${SENSOR_PORT})`, key);
		Arduino.addVarDef(`int ${valueKey}`, valueKey);
		Arduino.addLoopCode(`${valueKey} = ${key}.read()`);

		return [valueKey, ORDER_HIGH];
	}

	Arduino.weeemake_gesture_recognition_compare = function(block){
		const VALUE = this.valueToCode(block, 'VALUE', ORDER_RELATIONAL);
		const GESTURE_RECOGNITION_TYPE = block.getFieldValue('GESTURE_RECOGNITION_TYPE');

		//let key = `gestureSensor_` + SENSOR_PORT.split('_')[1];
		//let valueKey = key + '_value';

		//Arduino.addVarDef(`WeGestureSensor ${key}(${SENSOR_PORT})`, key);
		//Arduino.addVarDef(`int ${valueKey}`, valueKey);
		//Arduino.addLoopCode(`${valueKey} = ${key}.read()`);

		return [`${VALUE} == ${GESTURE_RECOGNITION_TYPE}`, ORDER_RELATIONAL];
	}

	Arduino.weeemake_image_recognition_set_mode = (nameList => function(block){
		const SENSOR_PORT = Arduino.valueToCode(block, 'SENSOR_PORT', ORDER_NONE);
		const IMAGE_RECOGNITION_MODE = block.getFieldValue('IMAGE_RECOGNITION_MODE');

		let key = `imageRec_` + SENSOR_PORT.split('_')[1];

		Arduino.addVarDef(`WeImageRecognition ${key}(${SENSOR_PORT})`, key);

		return Arduino.tab() + `${key}.${nameList[IMAGE_RECOGNITION_MODE]}()` + END;
	})(['setAutoTrackingMode', 'setLineFollowerMode']);

	Arduino.weeemake_image_recognition_color_position = function(block){
		const SENSOR_PORT = Arduino.valueToCode(block, 'SENSOR_PORT', ORDER_NONE);
		const IMAGE_RECOGNITION_TYPE = block.getFieldValue('IMAGE_RECOGNITION_TYPE');

		let key = `imageRec_` + SENSOR_PORT.split('_')[1];

		Arduino.addVarDef(`WeImageRecognition ${key}(${SENSOR_PORT})`, key);

		return [`${key}.getColorPosition(${IMAGE_RECOGNITION_TYPE})`, ORDER_HIGH];
	}
/*
	Arduino.weeemake_image_recognition_auto_tracking = function(block){
		const SENSOR_PORT = Arduino.valueToCode(block, 'SENSOR_PORT', ORDER_NONE);

		let key = `imageRec_` + SENSOR_PORT.split('_')[1];

		Arduino.addVarDef(`WeImageRecognition ${key}(${SENSOR_PORT})`, key);

		return [`${key}.getAutoPosition()`, ORDER_HIGH];
	}

	Arduino.weeemake_image_recognition_line_follow = function(block){
		const SENSOR_PORT = Arduino.valueToCode(block, 'SENSOR_PORT', ORDER_NONE);

		let key = `imageRec_` + SENSOR_PORT.split('_')[1];

		Arduino.addVarDef(`WeImageRecognition ${key}(${SENSOR_PORT})`, key);

		return [`${key}.getLineFollowerAngle()`, ORDER_HIGH];
	}
*/
	Arduino.weeemake_image_recognition_update = (list => function(block){
		const SENSOR_PORT = Arduino.valueToCode(block, 'SENSOR_PORT', ORDER_NONE);
		const IMAGE_RECOGNITION_MODE = block.getFieldValue('IMAGE_RECOGNITION_MODE');

		let key = `imageRec_` + SENSOR_PORT.split('_')[1];

		Arduino.addVarDef(`WeImageRecognition ${key}(${SENSOR_PORT})`, key);

		return [`${key}.${list[IMAGE_RECOGNITION_MODE]}()`, ORDER_HIGH];
	})(['getAutoPosition', 'getLineFollowerAngle'])

	Arduino.weeemake_image_recognition_value = (nameList => function(block){
		const SENSOR_PORT = Arduino.valueToCode(block, 'SENSOR_PORT', ORDER_NONE);
		const IMAGE_RECOGNITION_VALUE = block.getFieldValue('IMAGE_RECOGNITION_VALUE');

		let key = `imageRec_` + SENSOR_PORT.split('_')[1];

		Arduino.addVarDef(`WeImageRecognition ${key}(${SENSOR_PORT})`, key);

		return [`${key}.${nameList[IMAGE_RECOGNITION_VALUE]}`, ORDER_HIGH];
	})(['centerX', 'centerY', 'pixels']);

	Arduino.weeemake_image_recognition_line_follow_angle = function(block){
		const SENSOR_PORT = Arduino.valueToCode(block, 'SENSOR_PORT', ORDER_NONE);

		let key = `imageRec_` + SENSOR_PORT.split('_')[1];

		Arduino.addVarDef(`WeImageRecognition ${key}(${SENSOR_PORT})`, key);

		return [`${key}.angle`, ORDER_HIGH];
	}

	Arduino.weeemake_image_recognition_set_fast_mode = function(block){
		const SENSOR_PORT = Arduino.valueToCode(block, 'SENSOR_PORT', ORDER_NONE);
		const ON_OFF = block.getFieldValue('ON_OFF') == 'TRUE' ? 1 : 0;

		let key = `imageRec_` + SENSOR_PORT.split('_')[1];

		Arduino.addVarDef(`WeImageRecognition ${key}(${SENSOR_PORT})`, key);

		return Arduino.tab() + `${key}.fastMode(${ON_OFF})` + END;
	}

	Arduino.weeemake_image_recognition_face_position = function(block){
		const SENSOR_PORT = Arduino.valueToCode(block, 'SENSOR_PORT', ORDER_NONE);

		let key = `imageRec_` + SENSOR_PORT.split('_')[1];

		Arduino.addVarDef(`WeImageRecognition ${key}(${SENSOR_PORT})`, key);

		return [`${key}.getFacePositon()`, ORDER_HIGH];
	}

	Arduino.weeemake_vibration_motor = function(block){
		const SENSOR_PORT = Arduino.valueToCode(block, 'SENSOR_PORT', ORDER_NONE);
		const SPEED = this.valueToCode(block, 'SPEED', ORDER_NONE);

		let key = `vibrationMotor_` + SENSOR_PORT.split('_')[1];

		Arduino.addVarDef(`WeVibrationMotor ${key}(${SENSOR_PORT})`, key);

		let code;

		if(SPEED === '100') code = 'run()'
		else if(SPEED === '0') code = 'stop()'
		else code = `run(${SPEED})`

		return Arduino.tab() + `${key}.${code}` + END;
	}

	Arduino.weeemake_led_single_line_follower = function(block){
		const SENSOR_PORT = Arduino.valueToCode(block, 'SENSOR_PORT', ORDER_NONE);

		let key = `ledLineFollower_` + SENSOR_PORT.split('_')[1];

		Arduino.addVarDef(`WeLEDLineFollower ${key}(${SENSOR_PORT})`, key);

		return [`${key}.readSensor()`, ORDER_HIGH];
	}

	Arduino.weeemake_led_single_line_follower_led = function(block){
		const SENSOR_PORT = Arduino.valueToCode(block, 'SENSOR_PORT', ORDER_NONE);
		const ON_OFF = Arduino.valueToCode(block, 'ON_OFF', ORDER_NONE);

		let key = `ledLineFollower_` + SENSOR_PORT.split('_')[1];

		Arduino.addVarDef(`WeLEDLineFollower ${key}(${SENSOR_PORT})`, key);

		return Arduino.tab() + `${key}.showLED(${ON_OFF})` + END;
	};

	Arduino.weeemake_elf_core_board_ir_sender = function(block){
		let IR_CODE = block.getFieldValue('IR_CODE');
		const GROUP = this.valueToCode(block, 'GROUP', ORDER_NONE);

		let key = `ir_sender`;

		Arduino.addVarDef(`WeInfraredSender ${key}(OnBoard_IR_T)`, key);

		return Arduino.tab() + `${key}.send_nec_message(${GROUP}, ${IR_CODE})` + END;
	}

	Arduino.weeemake_rj11_light = function(block){
		const SENSOR_PORT = Arduino.valueToCode(block, 'SENSOR_PORT', ORDER_NONE);

		let key = `lightSensorRJ11_` + SENSOR_PORT.split('_')[1];

		Arduino.addVarDef(`WeLightSensor_RJ ${key}(${SENSOR_PORT})`, key);

		return [`${key}.read()`, ORDER_HIGH];
	}

	Arduino.weeemake_rj11_sound = function(block){
		const SENSOR_PORT = Arduino.valueToCode(block, 'SENSOR_PORT', ORDER_NONE);

		let key = `soundSensorRJ11_` + SENSOR_PORT.split('_')[1];

		Arduino.addVarDef(`WeSoundSensor_RJ ${key}(${SENSOR_PORT})`, key);

		return [`${key}.read()`, ORDER_HIGH];
	}

	Arduino.weeemake_adapter_digitalWrite = function(block){
		const SENSOR_PORT = Arduino.valueToCode(block, 'SENSOR_PORT', ORDER_NONE);
		const ADAPTER_INDEX = block.getFieldValue('ADAPTER_INDEX');
		const HIGH_LOW = Arduino.valueToCode(block, 'HIGH_LOW', ORDER_NONE);

		let key = `adapter_` + SENSOR_PORT.split('_').pop();

		Arduino.addVarDef(`WeAdapter ${key}(${SENSOR_PORT})`, key);

		return Arduino.tab() + `${key}.digitalWrite(${ADAPTER_INDEX}, ${HIGH_LOW})` + END;
	}

	Arduino.weeemake_adapter_servo = function(block){
		const SENSOR_PORT = Arduino.valueToCode(block, 'SENSOR_PORT', ORDER_NONE);
		const ADAPTER_INDEX = block.getFieldValue('ADAPTER_INDEX');
		const ANGLE = Arduino.valueToCode(block, 'ANGLE', ORDER_NONE);

		let key = `adapter_` + SENSOR_PORT.split('_').pop();

		Arduino.addVarDef(`WeAdapter ${key}(${SENSOR_PORT})`, key);

		return Arduino.tab() + `${key}.write(${ADAPTER_INDEX}, ${ANGLE})` + END;
	}

	Arduino.weeemake_adapter_rgb = function(block){
		const SENSOR_PORT = Arduino.valueToCode(block, 'SENSOR_PORT', ORDER_NONE);
		const ADAPTER_INDEX = block.getFieldValue('ADAPTER_INDEX');
		const PIXEL = Arduino.valueToCode(block, 'PIXEL', ORDER_NONE);
		const COLOUR = Arduino.valueToCode(block, 'COLOUR', ORDER_NONE);

		let key = `adapter_` + SENSOR_PORT.split('_').pop();

		Arduino.addVarDef(`WeAdapter ${key}(${SENSOR_PORT})`, key);

		return Arduino.tab() + `${key}.RGBshow(${ADAPTER_INDEX}, ${PIXEL}, ${COLOUR})` + END;
	}

	Arduino.weeemake_adapter_digitalRead = function(block){
		const SENSOR_PORT = Arduino.valueToCode(block, 'SENSOR_PORT', ORDER_NONE);
		const ADAPTER_INDEX = block.getFieldValue('ADAPTER_INDEX');

		let key = `adapter_` + SENSOR_PORT.split('_').pop();

		Arduino.addVarDef(`WeAdapter ${key}(${SENSOR_PORT})`, key);

		return [`${key}.digitalRead(${ADAPTER_INDEX})`, ORDER_HIGH];
	}

	Arduino.weeemake_adapter_analogRead = function(block){
		const SENSOR_PORT = Arduino.valueToCode(block, 'SENSOR_PORT', ORDER_NONE);
		const ADAPTER_INDEX = block.getFieldValue('ADAPTER_INDEX');

		let key = `adapter_` + SENSOR_PORT.split('_').pop();

		Arduino.addVarDef(`WeAdapter ${key}(${SENSOR_PORT})`, key);

		return [`${key}.analogRead(${ADAPTER_INDEX})`, ORDER_HIGH];
	}

	Arduino.weeemake_adapter_temperature = function(block){
		const SENSOR_PORT = Arduino.valueToCode(block, 'SENSOR_PORT', ORDER_NONE);
		const ADAPTER_INDEX = block.getFieldValue('ADAPTER_INDEX');

		let key = `adapter_` + SENSOR_PORT.split('_').pop();

		Arduino.addVarDef(`WeAdapter ${key}(${SENSOR_PORT})`, key);

		return [`${key}.readTemperature(${ADAPTER_INDEX})`, ORDER_HIGH];
	}

	function addBluetoothController(){
		const key = `bluetooth_controller`;
		
		Arduino.addVarDef(`WeBluetoothController ${key}`, key);
		Arduino.addSetupCode(`Serial.begin(115200)`);
		Arduino.addLoopCode(`${key}.loop()`, key);

		return key;
	}

	const bluetooth_controller_keys = [null,
		'WeJOYSTICK_LY',
		'WeJOYSTICK_LX',
		'WeJOYSTICK_RY',
		'WeJOYSTICK_RX',
		'WeBUTTON_ZR',
		'WeBUTTON_R',
		'WeBUTTON_ZL',
		'WeBUTTON_L',
		'WeBUTTON_HOME',
		'WeBUTTON_BL',
		'WeBUTTON_Y',
		'WeBUTTON_B',
		'WeBUTTON_A',
		'WeBUTTON_X',
		'WeBUTTON_PLUS',
		'WeBUTTON_MODE',
		'WeBUTTON_UP',
		'WeBUTTON_DOWN',
		'WeBUTTON_LEFT',
		'WeBUTTON_RIGHT',
		'WeBUTTON_MINUS',
		'WeBUTTON_BR'
	];

	Arduino.weeemake_bluetooth_controller_button = function(block){
		const BLUETOOTH_CONTROLLER_KEY = block.getFieldValue('BLUETOOTH_CONTROLLER_KEY');

		let key = addBluetoothController();

		return [`${key}.buttonPressed(${bluetooth_controller_keys[BLUETOOTH_CONTROLLER_KEY]})`, ORDER_HIGH];
	}

	Arduino.weeemake_bluetooth_controller_joystick = function(block){
		const LEFT_RIGHT = block.getFieldValue('LEFT_RIGHT');
		const AXIS2 = block.getFieldValue('AXIS2');

		let key = addBluetoothController();

		let a = LEFT_RIGHT == '0' ? 0 : 2;
		let b = AXIS2 == '0' ? 2 : 1;

		return [`${key}.readAnalog(${bluetooth_controller_keys[a+b]})`, ORDER_HIGH];
	};

	Arduino.weeemake_wifi_set_info = function(block){
		const SENSOR_PORT = Arduino.valueToCode(block, 'SENSOR_PORT', ORDER_NONE);
		//const SSID = castToStr(block, 'SSID');
		//const PWD = castToStr(block, 'PWD');
		const SSID = JSON.stringify(block.getFieldValue('SSID'));
		const PWD = JSON.stringify(block.getFieldValue('PWD'));

		let key = `wifi_` + SENSOR_PORT.split('_').pop();

		Arduino.addVarDef(`WeWifi ${key}(${SENSOR_PORT})`, key);

		return Arduino.tab() + `${key}.setWIFIinfo(${SSID}, ${PWD})` + END;
	}

	Arduino.weeemake_wifi_init_mode = function(block){
		const SENSOR_PORT = Arduino.valueToCode(block, 'SENSOR_PORT', ORDER_NONE);
		const MODE = block.getFieldValue('MODE');

		const key = `wifi_` + SENSOR_PORT.split('_').pop();

		Arduino.addVarDef(`WeWifi ${key}(${SENSOR_PORT})`, key);

		let code;

		if(MODE == 'sta'){
			code = 'initSTA()'
		}else if(MODE == 'ap'){
			const CHANNEL = block.getFieldValue('CHANNEL');
			code = `initAP(${CHANNEL})`
		}else{
			const IP = JSON.stringify(block.getFieldValue('IP'));
			const PORT = block.getFieldValue('PORT');
			const method = MODE == 'tcp' ? 'initTCP' : 'initSlave';
			code = method + `(${IP}, ${PORT})`;
		}

		return Arduino.tab() + `${key}.${code}` + END;
	}
/*
	Arduino.weeemake_wifi_init_slave = function(block){
		const SENSOR_PORT = Arduino.valueToCode(block, 'SENSOR_PORT', ORDER_NONE);
		const IP = castToStr(block, 'IP');
		const PORT = this.valueToCode(block, 'PORT', ORDER_NONE);

		let key = `wifi_` + SENSOR_PORT.split('_').pop();

		Arduino.addVarDef(`WeWifi ${key}(${SENSOR_PORT})`, key);

		return Arduino.tab() + `${key}.initSlave(${IP}, ${PORT})` + END;
	}

	Arduino.weeemake_wifi_init_tcp = function(block){
		const SENSOR_PORT = Arduino.valueToCode(block, 'SENSOR_PORT', ORDER_NONE);
		const IP = castToStr(block, 'IP');
		const PORT = this.valueToCode(block, 'PORT', ORDER_NONE);

		let key = `wifi_` + SENSOR_PORT.split('_').pop();

		Arduino.addVarDef(`WeWifi ${key}(${SENSOR_PORT})`, key);

		return Arduino.tab() + `${key}.initTCP(${IP}, ${PORT})` + END;
	}

	Arduino.weeemake_wifi_init_ap = function(block){
		const SENSOR_PORT = Arduino.valueToCode(block, 'SENSOR_PORT', ORDER_NONE);
		const CHANNEL = this.valueToCode(block, 'CHANNEL', ORDER_NONE);

		let key = `wifi_` + SENSOR_PORT.split('_').pop();

		Arduino.addVarDef(`WeWifi ${key}(${SENSOR_PORT})`, key);

		return Arduino.tab() + `${key}.initAP(${CHANNEL})` + END;
	}

	Arduino.weeemake_wifi_init_sta = function(block){
		const SENSOR_PORT = Arduino.valueToCode(block, 'SENSOR_PORT', ORDER_NONE);

		let key = `wifi_` + SENSOR_PORT.split('_').pop();

		Arduino.addVarDef(`WeWifi ${key}(${SENSOR_PORT})`, key);

		return Arduino.tab() + `${key}.initSTA()` + END;
	}
*/
	Arduino.weeemake_wifi_write = function(block){
		const SENSOR_PORT = Arduino.valueToCode(block, 'SENSOR_PORT', ORDER_NONE);
		const VALUE = castToStr(block, 'VALUE');
		const ID = this.valueToCode(block, 'ID', ORDER_NONE);

		let key = `wifi_` + SENSOR_PORT.split('_').pop();

		Arduino.addVarDef(`WeWifi ${key}(${SENSOR_PORT})`, key);

		return Arduino.tab() + `${key}.write(${VALUE}, ${ID})` + END;
	}

	Arduino.weeemake_wifi_read = function(block){
		const SENSOR_PORT = Arduino.valueToCode(block, 'SENSOR_PORT', ORDER_NONE);

		let key = `wifi_` + SENSOR_PORT.split('_').pop();

		Arduino.addVarDef(`WeWifi ${key}(${SENSOR_PORT})`, key);

		return [`${key}.readString()`, ORDER_HIGH];
	}

	Arduino.weeemake_wifi_read_ip = function(block){
		const SENSOR_PORT = Arduino.valueToCode(block, 'SENSOR_PORT', ORDER_NONE);

		let key = `wifi_` + SENSOR_PORT.split('_').pop();

		Arduino.addVarDef(`WeWifi ${key}(${SENSOR_PORT})`, key);

		return [`${key}.readIP()`, ORDER_HIGH];
	}

	Arduino.weeemake_wifi_value = (list => function(block){
		const SENSOR_PORT = Arduino.valueToCode(block, 'SENSOR_PORT', ORDER_NONE);
		const WIFI_VALUE_NAME = block.getFieldValue('WIFI_VALUE_NAME');

		let key = `wifi_` + SENSOR_PORT.split('_').pop();

		Arduino.addVarDef(`WeWifi ${key}(${SENSOR_PORT})`, key);

		return [`${key}.${list[WIFI_VALUE_NAME]}`, ORDER_HIGH];
	})(['IP_addr', 'link_IP', 'rx_buffer']);

	Arduino.weeemake_wifi_check_recv_data = function(block){
		const SENSOR_PORT = Arduino.valueToCode(block, 'SENSOR_PORT', ORDER_NONE);
		const DATA = castToStr(block, 'DATA');

		let key = `wifi_` + SENSOR_PORT.split('_').pop();

		Arduino.addVarDef(`WeWifi ${key}(${SENSOR_PORT})`, key);

		return [`strcmp(${key}.rx_buffer, ${DATA}) == 0`, ORDER_RELATIONAL];
	}

	Arduino.weeemake_image_recognition_set_color = function(block){
		const SENSOR_PORT = Arduino.valueToCode(block, 'SENSOR_PORT', ORDER_NONE);
		const IMAGE_RECOGNITION_COLOR_INDEX = block.getFieldValue('IMAGE_RECOGNITION_COLOR_INDEX');
		const minL = this.valueToCode(block, 'minL', ORDER_NONE);
		const maxL = this.valueToCode(block, 'maxL', ORDER_NONE);
		const minA = this.valueToCode(block, 'minA', ORDER_NONE);
		const maxA = this.valueToCode(block, 'maxA', ORDER_NONE);
		const minB = this.valueToCode(block, 'minB', ORDER_NONE);
		const maxB = this.valueToCode(block, 'maxB', ORDER_NONE);

		let key = `imageRec_` + SENSOR_PORT.split('_')[1];

		Arduino.addVarDef(`WeImageRecognition ${key}(${SENSOR_PORT})`, key);

		return Arduino.tab() + `${key}.setLabColor${IMAGE_RECOGNITION_COLOR_INDEX}(${minL}, ${maxL}, ${minA}, ${maxA}, ${minB}, ${maxB})` + END;
	}

	Arduino.weeemake_image_recognition_set_threshold = function(block){
		const SENSOR_PORT = Arduino.valueToCode(block, 'SENSOR_PORT', ORDER_NONE);
		const VALUE = this.valueToCode(block, 'VALUE', ORDER_NONE);

		let key = `imageRec_` + SENSOR_PORT.split('_')[1];

		Arduino.addVarDef(`WeImageRecognition ${key}(${SENSOR_PORT})`, key);

		return Arduino.tab() + `${key}.setPixelsThreshold(${VALUE})` + END;
	}

	Arduino.weeemake_image_recognition_set_traffic_signs_mode = function(block){
		const SENSOR_PORT = Arduino.valueToCode(block, 'SENSOR_PORT', ORDER_NONE);

		let key = `imageRec_` + SENSOR_PORT.split('_')[1];

		Arduino.addVarDef(`WeImageRecognition ${key}(${SENSOR_PORT})`, key);

		return Arduino.tab() + `${key}.setTrafficSignsMode()` + END;
	}

	Arduino.weeemake_image_recognition_get_traffic_signs_value = function(block){
		const SENSOR_PORT = Arduino.valueToCode(block, 'SENSOR_PORT', ORDER_NONE);
		const TRAFFIC_SIGN = block.getFieldValue('TRAFFIC_SIGN');

		let key = `imageRec_` + SENSOR_PORT.split('_')[1];
		let valueKey = key + '_value';

		Arduino.addVarDef(`WeImageRecognition ${key}(${SENSOR_PORT})`, key);
		Arduino.addVarDef(`int ${valueKey}`, valueKey);

		Arduino.addLoopCode(`${valueKey} = ${key}.getTrafficSigns()`);

		return [`${valueKey} == ${TRAFFIC_SIGN}`, ORDER_RELATIONAL];
	}

	Arduino.weeemake_ir_send = function(block){
		const SENSOR_PORT = Arduino.valueToCode(block, 'SENSOR_PORT', ORDER_NONE);
		const DATA = block.getInputTargetBlock('DATA').getFieldValue('TEXT') || '0';

		let len = DATA.split(',').length;

		let key = `ir_sender_` + SENSOR_PORT.split('_')[1];

		Arduino.addVarDef(`WeInfraredSender ${key}(${SENSOR_PORT})`, key);
		let tab = Arduino.tab();

		return `${tab}{\n\t${tab}unsigned int data[] = {${DATA}};\n\t${tab}${key}.send_ir_raw(data, ${len});\n${tab}}\n`;
	}
}
